// @backstage/plugin-scaffolder-react/dist/next/components/TemplateOutputs/TextOutputs.esm.js
import { useApp } from '@backstage/core-plugin-api';
import Button from '@material-ui/core/Button';
import DescriptionIcon from '@material-ui/icons/Description';
import React from 'react';

const TextOutputs = (props) => {
  const {
    output: { text = [] },
    index,
    setIndex
  } = props;
  const app = useApp();
  const iconResolver = (key) => app.getSystemIcon(key) ?? DescriptionIcon;
  return /* @__PURE__ */ React.createElement(React.Fragment, null, text.filter(({ content }) => content !== void 0).map(({ title, icon }, i) => {
    const Icon = iconResolver(icon);
    return /* @__PURE__ */ React.createElement(
      Button,
      {
        key: i,
        startIcon: /* @__PURE__ */ React.createElement(Icon, null),
        component: "div",
        color: "primary",
        onClick: () => {
          if (index !== i) {
            setIndex?.(i);
          }
        },
        variant: index === i ? "outlined" : void 0
      },
      title
    );
  }));
};

export { TextOutputs };
