// @backstage/plugin-scaffolder-common/dist/alpha.cjs.js
'use strict';

var permissions = require('./permissions.cjs.js');



exports.RESOURCE_TYPE_SCAFFOLDER_ACTION = permissions.RESOURCE_TYPE_SCAFFOLDER_ACTION;
exports.RESOURCE_TYPE_SCAFFOLDER_TEMPLATE = permissions.RESOURCE_TYPE_SCAFFOLDER_TEMPLATE;
exports.actionExecutePermission = permissions.actionExecutePermission;
exports.scaffolderActionPermissions = permissions.scaffolderActionPermissions;
exports.scaffolderPermissions = permissions.scaffolderPermissions;
exports.scaffolderTaskPermissions = permissions.scaffolderTaskPermissions;
exports.scaffolderTemplatePermissions = permissions.scaffolderTemplatePermissions;
exports.taskCancelPermission = permissions.taskCancelPermission;
exports.taskCreatePermission = permissions.taskCreatePermission;
exports.taskReadPermission = permissions.taskReadPermission;
exports.templateManagementPermission = permissions.templateManagementPermission;
exports.templateParameterReadPermission = permissions.templateParameterReadPermission;
exports.templateStepReadPermission = permissions.templateStepReadPermission;
