// @backstage/core-components/dist/components/CopyTextButton/CopyTextButton.esm.js
import { useApi, errorApiRef } from '@backstage/core-plugin-api';
import IconButton from '@material-ui/core/IconButton';
import Tooltip from '@material-ui/core/Tooltip';
import CopyIcon from '@material-ui/icons/FileCopy';
import React__default, { useState, useEffect } from 'react';
import useCopyToClipboard from 'react-use/esm/useCopyToClipboard';
import { coreComponentsTranslationRef } from '../../translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

function CopyTextButton(props) {
  const { t } = useTranslationRef(coreComponentsTranslationRef);
  const {
    text,
    tooltipDelay = 1e3,
    tooltipText = t("copyTextButton.tooltipText"),
    "aria-label": ariaLabel = "Copy text"
  } = props;
  const errorApi = useApi(errorApiRef);
  const [open, setOpen] = useState(false);
  const [{ error }, copyToClipboard] = useCopyToClipboard();
  useEffect(() => {
    if (error) {
      errorApi.post(error);
    }
  }, [error, errorApi]);
  const handleCopyClick = (e) => {
    e.stopPropagation();
    setOpen(true);
    copyToClipboard(text);
  };
  return /* @__PURE__ */ React__default.createElement(React__default.Fragment, null, /* @__PURE__ */ React__default.createElement(
    Tooltip,
    {
      id: "copy-test-tooltip",
      title: tooltipText,
      placement: "top",
      leaveDelay: tooltipDelay,
      onClose: () => setOpen(false),
      open
    },
    /* @__PURE__ */ React__default.createElement(IconButton, { onClick: handleCopyClick, "aria-label": ariaLabel }, /* @__PURE__ */ React__default.createElement(CopyIcon, null))
  ));
}

export { CopyTextButton };
