// @backstage/plugin-techdocs/dist/home/components/Tables/actions.esm.js
import React from 'react';
import ShareIcon from '@material-ui/icons/Share';
import { FavoriteToggleIcon } from '@backstage/core-components';

const actionFactories = {
  createCopyDocsUrlAction(copyToClipboard) {
    return (row) => {
      return {
        icon: () => /* @__PURE__ */ React.createElement(ShareIcon, { fontSize: "small" }),
        tooltip: "Click to copy documentation link to clipboard",
        onClick: () => copyToClipboard(`${window.location.origin}${row.resolved.docsUrl}`)
      };
    };
  },
  createStarEntityAction(isStarredEntity, toggleStarredEntity) {
    return (row) => {
      const entity = row.entity;
      const isStarred = isStarredEntity(entity);
      return {
        cellStyle: { paddingLeft: "1em" },
        icon: () => /* @__PURE__ */ React.createElement(FavoriteToggleIcon, { isFavorite: isStarred }),
        tooltip: isStarred ? "Remove from favorites" : "Add to favorites",
        onClick: () => toggleStarredEntity(entity)
      };
    };
  }
};

export { actionFactories };
