// @backstage/plugin-techdocs/dist/home/components/Tables/CursorPaginatedDocsTable.esm.js
import React from 'react';
import { Table } from '@backstage/core-components';

function CursorPaginatedDocsTable(props) {
  const {
    actions,
    columns,
    data,
    next,
    prev,
    title,
    isLoading,
    options,
    ...restProps
  } = props;
  return /* @__PURE__ */ React.createElement(
    Table,
    {
      title: isLoading ? "" : title,
      columns,
      data,
      options: {
        paginationPosition: "both",
        ...options,
        // These settings are configured to force server side pagination
        pageSizeOptions: [],
        showFirstLastPageButtons: false,
        pageSize: Number.MAX_SAFE_INTEGER,
        emptyRowsWhenPaging: false,
        actionsColumnIndex: -1
      },
      onPageChange: (page) => {
        if (page > 0) {
          next?.();
        } else {
          prev?.();
        }
      },
      page: prev ? 1 : 0,
      totalCount: next ? Number.MAX_VALUE : Number.MAX_SAFE_INTEGER,
      localization: { pagination: { labelDisplayedRows: "" } },
      isLoading,
      ...restProps
    }
  );
}

export { CursorPaginatedDocsTable };
