// @backstage/plugin-catalog/dist/components/HasSystemsCard/HasSystemsCard.esm.js
import { RELATION_HAS_PART } from '@backstage/catalog-model';
import React from 'react';
import { RelatedEntitiesCard } from '../RelatedEntitiesCard/RelatedEntitiesCard.esm.js';
import { systemEntityColumns, asSystemEntities, systemEntityHelpLink } from '../RelatedEntitiesCard/presets.esm.js';
import { catalogTranslationRef } from '../../alpha/translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

function HasSystemsCard(props) {
  const { t } = useTranslationRef(catalogTranslationRef);
  const {
    variant = "gridItem",
    title = t("hasSystemsCard.title"),
    columns = systemEntityColumns,
    tableOptions = {}
  } = props;
  return /* @__PURE__ */ React.createElement(
    RelatedEntitiesCard,
    {
      variant,
      title,
      entityKind: "System",
      relationType: RELATION_HAS_PART,
      columns,
      asRenderableEntities: asSystemEntities,
      emptyMessage: t("hasSystemsCard.emptyMessage"),
      emptyHelpLink: systemEntityHelpLink,
      tableOptions
    }
  );
}

export { HasSystemsCard };
