// @backstage/plugin-catalog/dist/components/HasResourcesCard/HasResourcesCard.esm.js
import { RELATION_HAS_PART } from '@backstage/catalog-model';
import React from 'react';
import { RelatedEntitiesCard } from '../RelatedEntitiesCard/RelatedEntitiesCard.esm.js';
import { resourceEntityColumns, asResourceEntities, resourceEntityHelpLink } from '../RelatedEntitiesCard/presets.esm.js';
import { catalogTranslationRef } from '../../alpha/translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

function HasResourcesCard(props) {
  const { t } = useTranslationRef(catalogTranslationRef);
  const {
    variant = "gridItem",
    title = t("hasResourcesCard.title"),
    columns = resourceEntityColumns,
    tableOptions = {}
  } = props;
  return /* @__PURE__ */ React.createElement(
    RelatedEntitiesCard,
    {
      variant,
      title,
      entityKind: "Resource",
      relationType: RELATION_HAS_PART,
      columns,
      asRenderableEntities: asResourceEntities,
      emptyMessage: t("hasResourcesCard.emptyMessage"),
      emptyHelpLink: resourceEntityHelpLink,
      tableOptions
    }
  );
}

export { HasResourcesCard };
