// @backstage/backend-defaults/dist/entrypoints/auth/plugin/keys/createPluginKeySource.cjs.js
'use strict';

var DatabasePluginKeySource = require('./DatabasePluginKeySource.cjs.js');
var StaticConfigPluginKeySource = require('./StaticConfigPluginKeySource.cjs.js');

const CONFIG_ROOT_KEY = "backend.auth.pluginKeyStore";
async function createPluginKeySource(options) {
  const keyStoreConfig = options.config.getOptionalConfig(CONFIG_ROOT_KEY);
  const type = keyStoreConfig?.getOptionalString("type") ?? "database";
  if (!keyStoreConfig || type === "database") {
    return DatabasePluginKeySource.DatabasePluginKeySource.create({
      database: options.database,
      logger: options.logger,
      keyDuration: options.keyDuration,
      algorithm: options.algorithm
    });
  } else if (type === "static") {
    return StaticConfigPluginKeySource.StaticConfigPluginKeySource.create({
      sourceConfig: keyStoreConfig,
      keyDuration: options.keyDuration
    });
  }
  throw new Error(
    `Unsupported config value ${CONFIG_ROOT_KEY}.type '${type}'; expected one of 'database', 'static'`
  );
}

exports.createPluginKeySource = createPluginKeySource;
