// @backstage/plugin-auth-backend-module-onelogin-provider/dist/module.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var pluginAuthNode = require('@backstage/plugin-auth-node');
var authenticator = require('./authenticator.cjs.js');
var resolvers = require('./resolvers.cjs.js');

const authModuleOneLoginProvider = backendPluginApi.createBackendModule({
  pluginId: "auth",
  moduleId: "onelogin-provider",
  register(reg) {
    reg.registerInit({
      deps: {
        providers: pluginAuthNode.authProvidersExtensionPoint
      },
      async init({ providers }) {
        providers.registerProvider({
          providerId: "onelogin",
          factory: pluginAuthNode.createOAuthProviderFactory({
            authenticator: authenticator.oneLoginAuthenticator,
            signInResolverFactories: {
              ...resolvers.oneLoginSignInResolvers,
              ...pluginAuthNode.commonSignInResolvers
            }
          })
        });
      }
    });
  }
});

exports.authModuleOneLoginProvider = authModuleOneLoginProvider;
