// @backstage/plugin-api-docs/dist/components/GraphQlDefinitionWidget/GraphQlDefinition.esm.js
import { makeStyles } from '@material-ui/core/styles';
import { EditorContextProvider, SchemaContextProvider, ExplorerContextProvider, DocExplorer } from '@graphiql/react';
import 'graphiql/graphiql.css';
import { buildSchema } from 'graphql';
import React from 'react';

const useStyles = makeStyles({
  root: {
    height: "100%",
    display: "flex",
    flexFlow: "column nowrap"
  },
  graphiQlWrapper: {
    flex: 1,
    "@global": {
      ".graphiql-container": {
        boxSizing: "initial",
        height: "100%",
        minHeight: "600px",
        flex: "1 1 auto"
      },
      ".graphiql-sidebar": {
        width: "100%"
      }
    }
  }
});
const GraphQlDefinition = ({ definition }) => {
  const classes = useStyles();
  const schema = buildSchema(definition);
  return /* @__PURE__ */ React.createElement("div", { className: classes.root }, /* @__PURE__ */ React.createElement("div", { className: classes.graphiQlWrapper }, /* @__PURE__ */ React.createElement(EditorContextProvider, null, /* @__PURE__ */ React.createElement(
    SchemaContextProvider,
    {
      schema,
      fetcher: () => Promise.resolve(null)
    },
    /* @__PURE__ */ React.createElement("div", { className: "graphiql-container" }, /* @__PURE__ */ React.createElement("div", { className: "graphiql-sidebar" }, /* @__PURE__ */ React.createElement("div", { className: "graphiql-sidebar-section" }, /* @__PURE__ */ React.createElement(ExplorerContextProvider, null, /* @__PURE__ */ React.createElement(DocExplorer, null)))))
  ))));
};

export { GraphQlDefinition };
