// @backstage/plugin-catalog/dist/components/HasSubdomainsCard/HasSubdomainsCard.esm.js
import { RELATION_HAS_PART } from '@backstage/catalog-model';
import React from 'react';
import { RelatedEntitiesCard } from '../RelatedEntitiesCard/RelatedEntitiesCard.esm.js';
import { domainEntityColumns, asDomainEntities, domainEntityHelpLink } from '../RelatedEntitiesCard/presets.esm.js';
import { catalogTranslationRef } from '../../alpha/translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

function HasSubdomainsCard(props) {
  const { t } = useTranslationRef(catalogTranslationRef);
  const {
    variant = "gridItem",
    title = t("hasSubdomainsCard.title"),
    columns = domainEntityColumns,
    tableOptions = {}
  } = props;
  return /* @__PURE__ */ React.createElement(
    RelatedEntitiesCard,
    {
      variant,
      title,
      entityKind: "Domain",
      relationType: RELATION_HAS_PART,
      columns,
      asRenderableEntities: asDomainEntities,
      emptyMessage: t("hasSubdomainsCard.emptyMessage"),
      emptyHelpLink: domainEntityHelpLink,
      tableOptions
    }
  );
}

export { HasSubdomainsCard };
