// @swagger-api/apidom-ns-workflows-1/src/refractor/registration.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
exports.__esModule = true;
var _WorkflowsSpecification = _interopRequireDefault(require("../elements/WorkflowsSpecification1.cjs"));
exports.WorkflowsSpecification1Element = _WorkflowsSpecification.default;
var _WorkflowsSpec = _interopRequireDefault(require("../elements/WorkflowsSpec.cjs"));
exports.WorkflowsSpecElement = _WorkflowsSpec.default;
var _Info = _interopRequireDefault(require("../elements/Info.cjs"));
exports.InfoElement = _Info.default;
var _SourceDescription = _interopRequireDefault(require("../elements/SourceDescription.cjs"));
exports.SourceDescriptionElement = _SourceDescription.default;
var _Workflow = _interopRequireDefault(require("../elements/Workflow.cjs"));
exports.WorkflowElement = _Workflow.default;
var _Step = _interopRequireDefault(require("../elements/Step.cjs"));
exports.StepElement = _Step.default;
var _Parameter = _interopRequireDefault(require("../elements/Parameter.cjs"));
exports.ParameterElement = _Parameter.default;
var _SuccessAction = _interopRequireDefault(require("../elements/SuccessAction.cjs"));
exports.SuccessActionElement = _SuccessAction.default;
var _FailureAction = _interopRequireDefault(require("../elements/FailureAction.cjs"));
exports.FailureActionElement = _FailureAction.default;
var _Components = _interopRequireDefault(require("../elements/Components.cjs"));
exports.ComponentsElement = _Components.default;
var _Criterion = _interopRequireDefault(require("../elements/Criterion.cjs"));
exports.CriterionElement = _Criterion.default;
var _Reference = _interopRequireDefault(require("../elements/Reference.cjs"));
exports.ReferenceElement = _Reference.default;
var _JSONSchema = _interopRequireDefault(require("../elements/JSONSchema.cjs"));
exports.JSONSchemaElement = _JSONSchema.default;
var _index = require("./index.cjs");
_Info.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Info', '$visitor']);
_WorkflowsSpec.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'WorkflowsSpecification', 'fixedFields', 'workflowsSpec']);
_WorkflowsSpecification.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'WorkflowsSpecification', '$visitor']);
_SourceDescription.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'SourceDescription', '$visitor']);
_Workflow.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Workflow', '$visitor']);
_Step.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Step', '$visitor']);
_Parameter.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Parameter', '$visitor']);
_SuccessAction.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'SuccessAction', '$visitor']);
_FailureAction.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'FailureAction', '$visitor']);
_Components.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Components', '$visitor']);
_Criterion.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Criterion', '$visitor']);
_Reference.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Reference', '$visitor']);
_JSONSchema.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'JSONSchema', '$visitor']);
