// @backstage/plugin-auth-backend-module-bitbucket-server-provider/dist/module.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var pluginAuthNode = require('@backstage/plugin-auth-node');
var authenticator = require('./authenticator.cjs.js');
var resolvers = require('./resolvers.cjs.js');

const authModuleBitbucketServerProvider = backendPluginApi.createBackendModule({
  pluginId: "auth",
  moduleId: "bitbucket-server-provider",
  register(reg) {
    reg.registerInit({
      deps: {
        providers: pluginAuthNode.authProvidersExtensionPoint
      },
      async init({ providers }) {
        providers.registerProvider({
          providerId: "bitbucketServer",
          factory: pluginAuthNode.createOAuthProviderFactory({
            authenticator: authenticator.bitbucketServerAuthenticator,
            signInResolverFactories: {
              ...resolvers.bitbucketServerSignInResolvers,
              ...pluginAuthNode.commonSignInResolvers
            }
          })
        });
      }
    });
  }
});

exports.authModuleBitbucketServerProvider = authModuleBitbucketServerProvider;
