// @backstage/plugin-search/dist/components/SearchType/SearchType.esm.js
import Checkbox from '@material-ui/core/Checkbox';
import Chip from '@material-ui/core/Chip';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';
import ListItemText from '@material-ui/core/ListItemText';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';
import { makeStyles } from '@material-ui/core/styles';
import React from 'react';
import useEffectOnce from 'react-use/esm/useEffectOnce';
import { SearchTypeAccordion } from './SearchType.Accordion.esm.js';
import { SearchTypeTabs } from './SearchType.Tabs.esm.js';
import { useSearch } from '@backstage/plugin-search-react';

const useStyles = makeStyles((theme) => ({
  label: {
    textTransform: "capitalize"
  },
  chips: {
    display: "flex",
    flexWrap: "wrap",
    marginTop: theme.spacing(1)
  },
  chip: {
    margin: 2
  }
}));
const SearchType = (props) => {
  const { className, defaultValue, name, values = [] } = props;
  const classes = useStyles();
  const { types, setTypes } = useSearch();
  useEffectOnce(() => {
    if (!types.length) {
      if (defaultValue && Array.isArray(defaultValue)) {
        setTypes(defaultValue);
      } else if (defaultValue) {
        setTypes([defaultValue]);
      }
    }
  });
  const handleChange = (e) => {
    const value = e.target.value;
    setTypes(value);
  };
  return /* @__PURE__ */ React.createElement(
    FormControl,
    {
      className,
      variant: "filled",
      fullWidth: true,
      "data-testid": "search-typefilter-next"
    },
    /* @__PURE__ */ React.createElement(InputLabel, { className: classes.label, margin: "dense" }, name),
    /* @__PURE__ */ React.createElement(
      Select,
      {
        multiple: true,
        variant: "outlined",
        value: types,
        onChange: handleChange,
        placeholder: "All Results",
        renderValue: (selected) => /* @__PURE__ */ React.createElement("div", { className: classes.chips }, selected.map((value) => /* @__PURE__ */ React.createElement(
          Chip,
          {
            key: value,
            label: value,
            className: classes.chip,
            size: "small"
          }
        )))
      },
      values.map((value) => /* @__PURE__ */ React.createElement(MenuItem, { key: value, value }, /* @__PURE__ */ React.createElement(Checkbox, { checked: types.indexOf(value) > -1 }), /* @__PURE__ */ React.createElement(ListItemText, { primary: value })))
    )
  );
};
SearchType.Accordion = (props) => {
  return /* @__PURE__ */ React.createElement(SearchTypeAccordion, { ...props });
};
SearchType.Tabs = (props) => {
  return /* @__PURE__ */ React.createElement(SearchTypeTabs, { ...props });
};

export { SearchType };
