// @swagger-api/apidom-ns-api-design-systems/src/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard").default;
exports.__esModule = true;
exports.validateOpenAPI3_1 = exports.refractPluginOpenApi3_1StandardIdentifierSelectors = exports.refractPluginOpenApi3_1StandardIdentifierAccessors = exports.mediaTypes = exports.keyMap = exports.isStringElement = exports.isStandardIdentifierElement = exports.isStandardElement = exports.isScenarioElement = exports.isRequirementLevelElement = exports.isRequirementElement = exports.isRefElement = exports.isPrincipleElement = exports.isObjectElement = exports.isNumberElement = exports.isNullElement = exports.isMemberElement = exports.isMainElement = exports.isLinkElement = exports.isInfoElement = exports.isElement = exports.isBooleanElement = exports.isArrayElement = exports.getNodeType = exports.default = exports.Visitor = exports.StandardIdentifierElement = exports.StandardElement = exports.SpecificationVisitor = exports.ScenarioElement = exports.RequirementLevelElement = exports.RequirementElement = exports.PrincipleElement = exports.MainElement = exports.InfoElement = exports.FixedFieldsVisitor = exports.FallbackVisitor = exports.ApiDesignSystemsMediaTypes = void 0;
var _apidomCore = require("@swagger-api/apidom-core");
exports.isRefElement = _apidomCore.isRefElement;
exports.isLinkElement = _apidomCore.isLinkElement;
exports.isMemberElement = _apidomCore.isMemberElement;
exports.isObjectElement = _apidomCore.isObjectElement;
exports.isArrayElement = _apidomCore.isArrayElement;
exports.isBooleanElement = _apidomCore.isBooleanElement;
exports.isNullElement = _apidomCore.isNullElement;
exports.isElement = _apidomCore.isElement;
exports.isNumberElement = _apidomCore.isNumberElement;
exports.isStringElement = _apidomCore.isStringElement;
var _mediaTypes = _interopRequireWildcard(require("./media-types.cjs"));
exports.mediaTypes = _mediaTypes.default;
exports.ApiDesignSystemsMediaTypes = _mediaTypes.ApiDesignSystemsMediaTypes;
var _namespace = _interopRequireDefault(require("./namespace.cjs"));
exports.default = _namespace.default;
var _standardIdentifierSelectors = _interopRequireDefault(require("./refractor/plugins/openapi-3-1/standard-identifier-selectors.cjs"));
exports.refractPluginOpenApi3_1StandardIdentifierSelectors = _standardIdentifierSelectors.default;
var _standardIdentifierAccessors = _interopRequireDefault(require("./refractor/plugins/openapi-3-1/standard-identifier-accessors.cjs"));
exports.refractPluginOpenApi3_1StandardIdentifierAccessors = _standardIdentifierAccessors.default;
var _validator = _interopRequireDefault(require("./validator/openapi-3-1/validator.cjs"));
exports.validateOpenAPI3_1 = _validator.default;
var _predicates = require("./predicates.cjs");
exports.isInfoElement = _predicates.isInfoElement;
exports.isMainElement = _predicates.isMainElement;
exports.isPrincipleElement = _predicates.isPrincipleElement;
exports.isRequirementElement = _predicates.isRequirementElement;
exports.isRequirementLevelElement = _predicates.isRequirementLevelElement;
exports.isScenarioElement = _predicates.isScenarioElement;
exports.isStandardElement = _predicates.isStandardElement;
exports.isStandardIdentifierElement = _predicates.isStandardIdentifierElement;
var _FixedFieldsVisitor = _interopRequireDefault(require("./refractor/visitors/generics/FixedFieldsVisitor.cjs"));
exports.FixedFieldsVisitor = _FixedFieldsVisitor.default;
var _SpecificationVisitor = _interopRequireDefault(require("./refractor/visitors/SpecificationVisitor.cjs"));
exports.SpecificationVisitor = _SpecificationVisitor.default;
var _FallbackVisitor = _interopRequireDefault(require("./refractor/visitors/FallbackVisitor.cjs"));
exports.FallbackVisitor = _FallbackVisitor.default;
var _Visitor = _interopRequireDefault(require("./refractor/visitors/Visitor.cjs"));
exports.Visitor = _Visitor.default;
var _visitor = require("./traversal/visitor.cjs");
exports.keyMap = _visitor.keyMap;
exports.getNodeType = _visitor.getNodeType;
var _registration = require("./refractor/registration.cjs");
exports.MainElement = _registration.MainElement;
exports.InfoElement = _registration.InfoElement;
exports.PrincipleElement = _registration.PrincipleElement;
exports.RequirementElement = _registration.RequirementElement;
exports.RequirementLevelElement = _registration.RequirementLevelElement;
exports.ScenarioElement = _registration.ScenarioElement;
exports.StandardElement = _registration.StandardElement;
exports.StandardIdentifierElement = _registration.StandardIdentifierElement;
