// @stoplight/spectral-core/dist/ruleset/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rule = exports.Formats = exports.Ruleset = exports.createRulesetFunction = exports.getDiagnosticSeverity = exports.RulesetValidationError = exports.assertValidRuleset = void 0;
var index_1 = require("./validation/index");
Object.defineProperty(exports, "assertValidRuleset", { enumerable: true, get: function () { return index_1.assertValidRuleset; } });
Object.defineProperty(exports, "RulesetValidationError", { enumerable: true, get: function () { return index_1.RulesetValidationError; } });
var severity_1 = require("./utils/severity");
Object.defineProperty(exports, "getDiagnosticSeverity", { enumerable: true, get: function () { return severity_1.getDiagnosticSeverity; } });
var function_1 = require("./function");
Object.defineProperty(exports, "createRulesetFunction", { enumerable: true, get: function () { return function_1.createRulesetFunction; } });
var ruleset_1 = require("./ruleset");
Object.defineProperty(exports, "Ruleset", { enumerable: true, get: function () { return ruleset_1.Ruleset; } });
var formats_1 = require("./formats");
Object.defineProperty(exports, "Formats", { enumerable: true, get: function () { return formats_1.Formats; } });
var rule_1 = require("./rule");
Object.defineProperty(exports, "Rule", { enumerable: true, get: function () { return rule_1.Rule; } });
