// @backstage/frontend-plugin-api/dist/index.esm.js
export { AnalyticsContext } from './analytics/AnalyticsContext.esm.js';
export { useAnalytics } from './analytics/useAnalytics.esm.js';
export { appTreeApiRef } from './apis/definitions/AppTreeApi.esm.js';
export { SessionState, alertApiRef, appThemeApiRef, atlassianAuthApiRef, bitbucketAuthApiRef, bitbucketServerAuthApiRef, configApiRef, createApiFactory, createApiRef, discoveryApiRef, errorApiRef, featureFlagsApiRef, fetchApiRef, githubAuthApiRef, gitlabAuthApiRef, googleAuthApiRef, identityApiRef, microsoftAuthApiRef, oauthRequestApiRef, oktaAuthApiRef, oneloginAuthApiRef, storageApiRef, useApi, useApiHolder, vmwareCloudAuthApiRef, withApis } from '@backstage/core-plugin-api';
export { componentsApiRef, useComponentRef } from './apis/definitions/ComponentsApi.esm.js';
export { iconsApiRef } from './apis/definitions/IconsApi.esm.js';
export { routeResolutionApiRef } from './apis/definitions/RouteResolutionApi.esm.js';
export { analyticsApiRef } from './apis/definitions/AnalyticsApi.esm.js';
export { ApiBlueprint } from './blueprints/ApiBlueprint.esm.js';
export { AppRootElementBlueprint } from './blueprints/AppRootElementBlueprint.esm.js';
export { AppRootWrapperBlueprint } from './blueprints/AppRootWrapperBlueprint.esm.js';
export { IconBundleBlueprint } from './blueprints/IconBundleBlueprint.esm.js';
export { NavItemBlueprint } from './blueprints/NavItemBlueprint.esm.js';
export { NavLogoBlueprint } from './blueprints/NavLogoBlueprint.esm.js';
export { PageBlueprint } from './blueprints/PageBlueprint.esm.js';
export { RouterBlueprint } from './blueprints/RouterBlueprint.esm.js';
export { SignInPageBlueprint } from './blueprints/SignInPageBlueprint.esm.js';
export { ThemeBlueprint } from './blueprints/ThemeBlueprint.esm.js';
export { TranslationBlueprint } from './blueprints/TranslationBlueprint.esm.js';
export { ExtensionBoundary } from './components/ExtensionBoundary.esm.js';
export { coreComponentRefs } from './components/coreComponentRefs.esm.js';
export { createComponentRef } from './components/createComponentRef.esm.js';
export { createComponentExtension } from './extensions/createComponentExtension.esm.js';
export { createRouteRef } from './routing/RouteRef.esm.js';
export { createSubRouteRef } from './routing/SubRouteRef.esm.js';
export { createExternalRouteRef } from './routing/ExternalRouteRef.esm.js';
export { useRouteRef } from './routing/useRouteRef.esm.js';
export { useRouteRefParams } from './routing/useRouteRefParams.esm.js';
export { createTranslationMessages, createTranslationRef, createTranslationResource, useTranslationRef } from '@backstage/core-plugin-api/alpha';
export { coreExtensionData } from './wiring/coreExtensionData.esm.js';
export { createExtension } from './wiring/createExtension.esm.js';
export { createExtensionInput } from './wiring/createExtensionInput.esm.js';
export { createExtensionDataRef } from './wiring/createExtensionDataRef.esm.js';
export { createFrontendPlugin } from './wiring/createFrontendPlugin.esm.js';
export { createFrontendModule } from './wiring/createFrontendModule.esm.js';
export { createExtensionBlueprint } from './wiring/createExtensionBlueprint.esm.js';
