// @backstage/plugin-scaffolder-react/dist/next/components/Form/DescriptionFieldTemplate.esm.js
import React from 'react';
import { MarkdownContent } from '@backstage/core-components';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles((theme) => ({
  markdownDescription: {
    fontSize: theme.typography.caption.fontSize,
    margin: 0,
    color: theme.palette.text.secondary,
    "& :first-child": {
      margin: 0,
      marginTop: "3px"
      // to keep the standard browser padding
    }
  }
}));
const DescriptionFieldTemplate = (props) => {
  const { id, description } = props;
  const classes = useStyles();
  if (description) {
    if (typeof description === "string") {
      return /* @__PURE__ */ React.createElement(
        MarkdownContent,
        {
          content: description,
          linkTarget: "_blank",
          className: classes.markdownDescription
        }
      );
    }
    return /* @__PURE__ */ React.createElement(Typography, { id, variant: "subtitle2", style: { marginTop: "5px" } }, description);
  }
  return null;
};

export { DescriptionFieldTemplate };
