// @swagger-api/apidom-ns-json-schema-draft-7/src/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard").default;
exports.__esModule = true;
exports.specificationObj = exports.refractorPluginReplaceEmptyElement = exports.refract = exports.mediaTypes = exports.keyMap = exports.isStringElement = exports.isRefElement = exports.isObjectElement = exports.isNumberElement = exports.isNullElement = exports.isMemberElement = exports.isLinkElement = exports.isLinkDescriptionElement = exports.isJSONSchemaElement = exports.isJSONReferenceLikeElement = exports.isJSONReferenceElement = exports.isElement = exports.isBooleanElement = exports.isArrayElement = exports.getNodeType = exports.default = exports.createRefractor = exports.Visitor = exports.SpecificationVisitor = exports.PatternedFieldsVisitor = exports.ParentSchemaAwareVisitor = exports.MapVisitor = exports.LinkDescriptionElement = exports.JSONSchemaElement = exports.JSONSchemaDraft7MediaTypes = exports.JSONReferenceElement = exports.FixedFieldsVisitor = exports.FallbackVisitor = exports.AlternatingVisitor = void 0;
var _apidomCore = require("@swagger-api/apidom-core");
exports.isRefElement = _apidomCore.isRefElement;
exports.isLinkElement = _apidomCore.isLinkElement;
exports.isMemberElement = _apidomCore.isMemberElement;
exports.isObjectElement = _apidomCore.isObjectElement;
exports.isArrayElement = _apidomCore.isArrayElement;
exports.isBooleanElement = _apidomCore.isBooleanElement;
exports.isNullElement = _apidomCore.isNullElement;
exports.isElement = _apidomCore.isElement;
exports.isNumberElement = _apidomCore.isNumberElement;
exports.isStringElement = _apidomCore.isStringElement;
var _mediaTypes = _interopRequireWildcard(require("./media-types.cjs"));
exports.mediaTypes = _mediaTypes.default;
exports.JSONSchemaDraft7MediaTypes = _mediaTypes.JSONSchemaDraft7MediaTypes;
var _namespace = _interopRequireDefault(require("./namespace.cjs"));
exports.default = _namespace.default;
var _replaceEmptyElement = _interopRequireDefault(require("./refractor/plugins/replace-empty-element.cjs"));
exports.refractorPluginReplaceEmptyElement = _replaceEmptyElement.default;
var _index = _interopRequireWildcard(require("./refractor/index.cjs"));
exports.refract = _index.default;
exports.createRefractor = _index.createRefractor;
var _specification = _interopRequireDefault(require("./refractor/specification.cjs"));
exports.specificationObj = _specification.default;
var _predicates = require("./predicates.cjs");
exports.isJSONReferenceElement = _predicates.isJSONReferenceElement;
exports.isJSONSchemaElement = _predicates.isJSONSchemaElement;
exports.isLinkDescriptionElement = _predicates.isLinkDescriptionElement;
var _apidomNsJsonSchemaDraft = require("@swagger-api/apidom-ns-json-schema-draft-6");
exports.isJSONReferenceLikeElement = _apidomNsJsonSchemaDraft.isJSONReferenceLikeElement;
exports.SpecificationVisitor = _apidomNsJsonSchemaDraft.SpecificationVisitor;
exports.FallbackVisitor = _apidomNsJsonSchemaDraft.FallbackVisitor;
exports.FixedFieldsVisitor = _apidomNsJsonSchemaDraft.FixedFieldsVisitor;
exports.PatternedFieldsVisitor = _apidomNsJsonSchemaDraft.PatternedFieldsVisitor;
exports.MapVisitor = _apidomNsJsonSchemaDraft.MapVisitor;
exports.AlternatingVisitor = _apidomNsJsonSchemaDraft.AlternatingVisitor;
exports.ParentSchemaAwareVisitor = _apidomNsJsonSchemaDraft.ParentSchemaAwareVisitor;
exports.Visitor = _apidomNsJsonSchemaDraft.Visitor;
exports.JSONReferenceElement = _apidomNsJsonSchemaDraft.JSONReferenceElement;
var _visitor = require("./traversal/visitor.cjs");
exports.keyMap = _visitor.keyMap;
exports.getNodeType = _visitor.getNodeType;
var _registration = require("./refractor/registration.cjs");
exports.JSONSchemaElement = _registration.JSONSchemaElement;
exports.LinkDescriptionElement = _registration.LinkDescriptionElement;
