// @backstage/frontend-test-utils/dist/app/renderInTestApp.esm.js
import React from 'react';
import { Link, MemoryRouter } from 'react-router-dom';
import { createSpecializedApp } from '@backstage/frontend-app-api';
import { render } from '@testing-library/react';
import { ConfigReader } from '@backstage/config';
import { coreExtensionData, NavItemBlueprint, useRouteRef, createExtension, RouterBlueprint, createFrontendPlugin } from '@backstage/frontend-plugin-api';
import appPlugin from '@backstage/plugin-app';

const NavItem = (props) => {
  const { routeRef, title, icon: Icon } = props;
  const link = useRouteRef(routeRef);
  if (!link) {
    return null;
  }
  return /* @__PURE__ */ React.createElement("li", null, /* @__PURE__ */ React.createElement(Link, { to: link() }, /* @__PURE__ */ React.createElement(Icon, null), " ", title));
};
const appPluginOverride = appPlugin.withOverrides({
  extensions: [
    appPlugin.getExtension("sign-in-page:app").override({
      disabled: true
    }),
    appPlugin.getExtension("app/nav").override({
      output: [coreExtensionData.reactElement],
      factory(_originalFactory, { inputs }) {
        return [
          coreExtensionData.reactElement(
            /* @__PURE__ */ React.createElement("nav", null, /* @__PURE__ */ React.createElement("ul", null, inputs.items.map((item, index) => {
              const { icon, title, routeRef } = item.get(
                NavItemBlueprint.dataRefs.target
              );
              return /* @__PURE__ */ React.createElement(
                NavItem,
                {
                  key: index,
                  icon,
                  title,
                  routeRef
                }
              );
            })))
          )
        ];
      }
    })
  ]
});
function renderInTestApp(element, options) {
  const extensions = [
    createExtension({
      attachTo: { id: "app/routes", input: "routes" },
      output: [coreExtensionData.reactElement, coreExtensionData.routePath],
      factory: () => {
        return [
          coreExtensionData.reactElement(element),
          coreExtensionData.routePath("/")
        ];
      }
    }),
    RouterBlueprint.make({
      params: {
        Component: ({ children }) => /* @__PURE__ */ React.createElement(MemoryRouter, null, children)
      }
    })
  ];
  if (options?.mountedRoutes) {
    for (const [path, routeRef] of Object.entries(options.mountedRoutes)) {
      extensions.push(
        createExtension({
          kind: "test-route",
          name: path,
          attachTo: { id: "app/root", input: "elements" },
          output: [
            coreExtensionData.reactElement,
            coreExtensionData.routePath,
            coreExtensionData.routeRef
          ],
          factory: () => [
            coreExtensionData.reactElement(/* @__PURE__ */ React.createElement(React.Fragment, null)),
            coreExtensionData.routePath(path),
            coreExtensionData.routeRef(routeRef)
          ]
        })
      );
    }
  }
  if (options?.extensions) {
    extensions.push(...options.extensions);
  }
  const features = [
    createFrontendPlugin({
      id: "test",
      extensions
    }),
    appPluginOverride
  ];
  if (options?.features) {
    features.push(...options.features);
  }
  const app = createSpecializedApp({
    features,
    config: ConfigReader.fromConfigs([
      { context: "render-config", data: options?.config ?? {} }
    ])
  });
  return render(app.createRoot());
}

export { renderInTestApp };
