// @backstage/frontend-plugin-api/dist/blueprints/SignInPageBlueprint.esm.js
import React, { lazy } from 'react';
import '../wiring/coreExtensionData.esm.js';
import 'zod';
import 'zod-to-json-schema';
import { createExtensionDataRef } from '../wiring/createExtensionDataRef.esm.js';
import { createExtensionBlueprint } from '../wiring/createExtensionBlueprint.esm.js';
import { ExtensionBoundary } from '../components/ExtensionBoundary.esm.js';

const componentDataRef = createExtensionDataRef().with({ id: "core.sign-in-page.component" });
const SignInPageBlueprint = createExtensionBlueprint({
  kind: "sign-in-page",
  attachTo: { id: "app/root", input: "signInPage" },
  output: [componentDataRef],
  dataRefs: {
    component: componentDataRef
  },
  *factory({
    loader
  }, { node }) {
    const ExtensionComponent = lazy(
      () => loader().then((component) => ({ default: component }))
    );
    yield componentDataRef((props) => /* @__PURE__ */ React.createElement(ExtensionBoundary, { node, routable: true }, /* @__PURE__ */ React.createElement(ExtensionComponent, { ...props })));
  }
});

export { SignInPageBlueprint };
