// @swagger-api/apidom-ast/src/index.cjs
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
exports.__esModule = true;
exports.visit = exports.mergeAllVisitors = exports.isYamlTag = exports.isYamlStream = exports.isYamlSequence = exports.isYamlScalar = exports.isYamlMapping = exports.isYamlKeyValuePair = exports.isYamlDocument = exports.isYamlDirective = exports.isYamlComment = exports.isYamlAnchor = exports.isYamlAlias = exports.isPosition = exports.isPoint = exports.isParseResult = exports.isNode = exports.isLiteral = exports.isJsonTrue = exports.isJsonStringContent = exports.isJsonString = exports.isJsonProperty = exports.isJsonObject = exports.isJsonNumber = exports.isJsonNull = exports.isJsonKey = exports.isJsonFalse = exports.isJsonEscapeSequence = exports.isJsonDocument = exports.isJsonArray = exports.getVisitFn = exports.getNodeType = exports.cloneNode = exports.YamlTagError = exports.YamlTag = exports.YamlStyleGroup = exports.YamlStyle = exports.YamlStream = exports.YamlSequence = exports.YamlSchemaError = exports.YamlScalar = exports.YamlReferenceManager = exports.YamlReferenceError = exports.YamlNodeKind = exports.YamlNode = exports.YamlMapping = exports.YamlKeyValuePair = exports.YamlJsonSchema = exports.YamlFailsafeSchema = exports.YamlError = exports.YamlDocument = exports.YamlDirective = exports.YamlComment = exports.YamlCollection = exports.YamlAnchor = exports.YamlAlias = exports.Position = exports.Point = exports.ParseResult = exports.Literal = exports.JsonValue = exports.JsonTrue = exports.JsonStringContent = exports.JsonString = exports.JsonProperty = exports.JsonObject = exports.JsonNumber = exports.JsonNull = exports.JsonNode = exports.JsonKey = exports.JsonFalse = exports.JsonEscapeSequence = exports.JsonDocument = exports.JsonArray = exports.Error = exports.BREAK = void 0;
var _JsonNode = _interopRequireDefault(require("./json/nodes/JsonNode.cjs"));
exports.JsonNode = _JsonNode.default;
var _JsonDocument = _interopRequireDefault(require("./json/nodes/JsonDocument.cjs"));
exports.JsonDocument = _JsonDocument.default;
var _JsonObject = _interopRequireDefault(require("./json/nodes/JsonObject.cjs"));
exports.JsonObject = _JsonObject.default;
var _JsonProperty = _interopRequireDefault(require("./json/nodes/JsonProperty.cjs"));
exports.JsonProperty = _JsonProperty.default;
var _JsonArray = _interopRequireDefault(require("./json/nodes/JsonArray.cjs"));
exports.JsonArray = _JsonArray.default;
var _JsonValue = _interopRequireDefault(require("./json/nodes/JsonValue.cjs"));
exports.JsonValue = _JsonValue.default;
var _JsonKey = _interopRequireDefault(require("./json/nodes/JsonKey.cjs"));
exports.JsonKey = _JsonKey.default;
var _JsonString = _interopRequireDefault(require("./json/nodes/JsonString.cjs"));
exports.JsonString = _JsonString.default;
var _JsonStringContent = _interopRequireDefault(require("./json/nodes/JsonStringContent.cjs"));
exports.JsonStringContent = _JsonStringContent.default;
var _JsonEscapeSequence = _interopRequireDefault(require("./json/nodes/JsonEscapeSequence.cjs"));
exports.JsonEscapeSequence = _JsonEscapeSequence.default;
var _JsonNumber = _interopRequireDefault(require("./json/nodes/JsonNumber.cjs"));
exports.JsonNumber = _JsonNumber.default;
var _JsonTrue = _interopRequireDefault(require("./json/nodes/JsonTrue.cjs"));
exports.JsonTrue = _JsonTrue.default;
var _JsonFalse = _interopRequireDefault(require("./json/nodes/JsonFalse.cjs"));
exports.JsonFalse = _JsonFalse.default;
var _JsonNull = _interopRequireDefault(require("./json/nodes/JsonNull.cjs"));
exports.JsonNull = _JsonNull.default;
var _predicates = require("./json/nodes/predicates.cjs");
exports.isJsonDocument = _predicates.isDocument;
exports.isJsonFalse = _predicates.isFalse;
exports.isJsonProperty = _predicates.isProperty;
exports.isJsonStringContent = _predicates.isStringContent;
exports.isJsonEscapeSequence = _predicates.isEscapeSequence;
exports.isJsonArray = _predicates.isArray;
exports.isJsonKey = _predicates.isKey;
exports.isJsonNull = _predicates.isNull;
exports.isJsonNumber = _predicates.isNumber;
exports.isJsonObject = _predicates.isObject;
exports.isJsonString = _predicates.isString;
exports.isJsonTrue = _predicates.isTrue;
var _YamlAlias = _interopRequireDefault(require("./yaml/nodes/YamlAlias.cjs"));
exports.YamlAlias = _YamlAlias.default;
var _YamlCollection = _interopRequireDefault(require("./yaml/nodes/YamlCollection.cjs"));
exports.YamlCollection = _YamlCollection.default;
var _YamlComment = _interopRequireDefault(require("./yaml/nodes/YamlComment.cjs"));
exports.YamlComment = _YamlComment.default;
var _YamlDirective = _interopRequireDefault(require("./yaml/nodes/YamlDirective.cjs"));
exports.YamlDirective = _YamlDirective.default;
var _YamlDocument = _interopRequireDefault(require("./yaml/nodes/YamlDocument.cjs"));
exports.YamlDocument = _YamlDocument.default;
var _YamlKeyValuePair = _interopRequireDefault(require("./yaml/nodes/YamlKeyValuePair.cjs"));
exports.YamlKeyValuePair = _YamlKeyValuePair.default;
var _YamlMapping = _interopRequireDefault(require("./yaml/nodes/YamlMapping.cjs"));
exports.YamlMapping = _YamlMapping.default;
var _YamlNode = _interopRequireDefault(require("./yaml/nodes/YamlNode.cjs"));
exports.YamlNode = _YamlNode.default;
var _YamlScalar = _interopRequireDefault(require("./yaml/nodes/YamlScalar.cjs"));
exports.YamlScalar = _YamlScalar.default;
var _YamlSequence = _interopRequireDefault(require("./yaml/nodes/YamlSequence.cjs"));
exports.YamlSequence = _YamlSequence.default;
var _YamlStream = _interopRequireDefault(require("./yaml/nodes/YamlStream.cjs"));
exports.YamlStream = _YamlStream.default;
var _YamlTag = _interopRequireWildcard(require("./yaml/nodes/YamlTag.cjs"));
exports.YamlTag = _YamlTag.default;
exports.YamlNodeKind = _YamlTag.YamlNodeKind;
var _YamlAnchor = _interopRequireDefault(require("./yaml/nodes/YamlAnchor.cjs"));
exports.YamlAnchor = _YamlAnchor.default;
var _YamlStyle = require("./yaml/nodes/YamlStyle.cjs");
exports.YamlStyle = _YamlStyle.YamlStyle;
exports.YamlStyleGroup = _YamlStyle.YamlStyleGroup;
var _index = _interopRequireDefault(require("./yaml/schemas/failsafe/index.cjs"));
exports.YamlFailsafeSchema = _index.default;
var _index2 = _interopRequireDefault(require("./yaml/schemas/json/index.cjs"));
exports.YamlJsonSchema = _index2.default;
var _ReferenceManager = _interopRequireDefault(require("./yaml/anchors-aliases/ReferenceManager.cjs"));
exports.YamlReferenceManager = _ReferenceManager.default;
var _predicates2 = require("./yaml/nodes/predicates.cjs");
exports.isYamlAlias = _predicates2.isAlias;
exports.isYamlAnchor = _predicates2.isAnchor;
exports.isYamlComment = _predicates2.isComment;
exports.isYamlKeyValuePair = _predicates2.isKeyValuePair;
exports.isYamlDirective = _predicates2.isDirective;
exports.isYamlDocument = _predicates2.isDocument;
exports.isYamlMapping = _predicates2.isMapping;
exports.isYamlScalar = _predicates2.isScalar;
exports.isYamlSequence = _predicates2.isSequence;
exports.isYamlStream = _predicates2.isStream;
exports.isYamlTag = _predicates2.isTag;
var _YamlError = _interopRequireDefault(require("./yaml/errors/YamlError.cjs"));
exports.YamlError = _YamlError.default;
var _YamlReferenceError = _interopRequireDefault(require("./yaml/errors/YamlReferenceError.cjs"));
exports.YamlReferenceError = _YamlReferenceError.default;
var _YamlSchemaError = _interopRequireDefault(require("./yaml/errors/YamlSchemaError.cjs"));
exports.YamlSchemaError = _YamlSchemaError.default;
var _YamlTagError = _interopRequireDefault(require("./yaml/errors/YamlTagError.cjs"));
exports.YamlTagError = _YamlTagError.default;
var _Literal = _interopRequireDefault(require("./Literal.cjs"));
exports.Literal = _Literal.default;
var _Position = _interopRequireWildcard(require("./Position.cjs"));
exports.Point = _Position.Point;
exports.Position = _Position.default;
var _Error = _interopRequireDefault(require("./Error.cjs"));
exports.Error = _Error.default;
var _ParseResult = _interopRequireDefault(require("./ParseResult.cjs"));
exports.ParseResult = _ParseResult.default;
var _predicates3 = require("./predicates.cjs");
exports.isParseResult = _predicates3.isParseResult;
exports.isLiteral = _predicates3.isLiteral;
exports.isPoint = _predicates3.isPoint;
exports.isPosition = _predicates3.isPosition;
var _visitor = require("./traversal/visitor.cjs");
exports.BREAK = _visitor.BREAK;
exports.mergeAllVisitors = _visitor.mergeAll;
exports.getVisitFn = _visitor.getVisitFn;
exports.visit = _visitor.visit;
exports.getNodeType = _visitor.getNodeType;
exports.isNode = _visitor.isNode;
exports.cloneNode = _visitor.cloneNode;
