// @asyncapi/react-component/lib/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MESSAGE_HEADERS_TEXT = exports.HEADERS_TEXT = exports.PARAMETERS_TEXT = exports.CHANNELS_TEXT = exports.SCHEMAS_TEXT = exports.MESSAGES_TEXT = exports.OPERATIONS_TEXT = exports.SERVERS_TEXT = exports.GENERATED_TEXT = exports.REQUIRED_TEXT = exports.REPLIER_LABEL_DEFAULT_TEXT = exports.REPLIER_TEXT = exports.REQUEST_LABEL_DEFAULT_TEXT = exports.REQUEST_TEXT = exports.RECEIVE_TEXT_LABEL_DEFAULT_TEXT = exports.RECEIVE_TEXT = exports.SUBSCRIBE_LABEL_DEFAULT_TEXT = exports.SUBSCRIBE_TEXT = exports.SEND_LABEL_DEFAULT_TEXT = exports.SEND_TEXT = exports.PUBLISH_LABEL_DEFAULT_TEXT = exports.PUBLISH_TEXT = exports.DEPRECATED_TEXT = exports.SPECIFICATION_TEXT = exports.TYPE_TEXT = exports.LOCATION_TEXT = exports.EXTERAL_DOCUMENTATION_TEXT = exports.EMAIL_SUPPORT_TEXT = exports.URL_SUPPORT_TEXT = exports.TERMS_OF_SERVICE_TEXT = exports.LICENSE_TEXT = exports.EMAIL_TEXT = exports.URL_TEXT = exports.NAM_TEXTE = exports.CONTACT_TEXT = exports.RAW_MESSAGE_PUBLISH_TEXT = exports.RAW_MESSAGE_SUBSCRIBE_TEXT = exports.ONE_OF_FOLLOWING_MESSAGES_SUBSCRIBE_SINGLE_TEXT = exports.ONE_OF_FOLLOWING_MESSAGES_SUBSCRIBE_TEXT = exports.ONE_OF_FOLLOWING_MESSAGES_PUBLISH_SINGLE_TEXT = exports.ONE_OF_FOLLOWING_MESSAGES_PUBLISH_TEXT = exports.GENERATED_BADGE_TEXT = exports.ANY_OF_PAYLOADS_TEXT = exports.ONE_OF_PAYLOADS_TEXT = exports.SERVERS = exports.VALIDATION_ERRORS_TYPE = exports.EXPAND_ALL_TEXT = exports.COLLAPSE_ALL_TEXT = exports.CONTENT_TYPES_SITE = exports.CSS_PREFIX = void 0;
exports.URL_VARIABLES_TEXT = exports.SECURITY_TEXT = exports.COLLAPSE_ERROR_BUTTON_TEXT = exports.EXPAND_ERROR_BUTTON_TEXT = exports.ERROR_TEXT = exports.ANY_TEXT = exports.NONE_TEXT = exports.BINDINGS_SCHEMA_OBJECT_TEXT = exports.MESSAGE_BINDINGS_TEXT = exports.OPERATION_BINDINGS_TEXT = exports.CHANNEL_BINDINGS_TEXT = exports.SERVER_BINDINGS_TEXT = exports.SCHEMA_EXAMPLE_TEXT = exports.PAYLOAD_EXAMPLE_TEXT = exports.MESSAGE_PAYLOAD_TEXT = exports.PAYLOAD_TEXT = exports.TAGS_TEXT = exports.HEADERS_EXAMPLE_TEXT = void 0;
exports.CSS_PREFIX = 'asyncapi';
exports.CONTENT_TYPES_SITE = 'https://www.iana.org/assignments/media-types';
exports.COLLAPSE_ALL_TEXT = 'Collapse All';
exports.EXPAND_ALL_TEXT = 'Expand All';
exports.VALIDATION_ERRORS_TYPE = 'https://github.com/asyncapi/parser-js/validation-errors';
exports.SERVERS = 'Servers';
exports.ONE_OF_PAYLOADS_TEXT = 'One of those payloads:';
exports.ANY_OF_PAYLOADS_TEXT = 'Any of those payloads:';
exports.GENERATED_BADGE_TEXT = 'generated';
exports.ONE_OF_FOLLOWING_MESSAGES_PUBLISH_TEXT = 'You can send one of the following messages:';
exports.ONE_OF_FOLLOWING_MESSAGES_PUBLISH_SINGLE_TEXT = 'You can send the following message:';
exports.ONE_OF_FOLLOWING_MESSAGES_SUBSCRIBE_TEXT = 'You can subscribe to one of the following messages:';
exports.ONE_OF_FOLLOWING_MESSAGES_SUBSCRIBE_SINGLE_TEXT = 'You can subscribe to the following message:';
exports.RAW_MESSAGE_SUBSCRIBE_TEXT = 'You can subscribe to the following message:';
exports.RAW_MESSAGE_PUBLISH_TEXT = 'You can send the following message:';
exports.CONTACT_TEXT = 'Contact';
exports.NAM_TEXTE = 'Name';
exports.URL_TEXT = 'Url';
exports.EMAIL_TEXT = 'Email';
exports.LICENSE_TEXT = 'License';
exports.TERMS_OF_SERVICE_TEXT = 'Terms of service';
exports.URL_SUPPORT_TEXT = 'Support';
exports.EMAIL_SUPPORT_TEXT = 'Email support';
exports.EXTERAL_DOCUMENTATION_TEXT = 'External Docs';
exports.LOCATION_TEXT = 'Location';
exports.TYPE_TEXT = 'Type';
exports.SPECIFICATION_TEXT = 'Specification';
exports.DEPRECATED_TEXT = 'Deprecated';
exports.PUBLISH_TEXT = 'Publish';
exports.PUBLISH_LABEL_DEFAULT_TEXT = 'PUB';
exports.SEND_TEXT = 'Send';
exports.SEND_LABEL_DEFAULT_TEXT = 'SEND';
exports.SUBSCRIBE_TEXT = 'Subscribe';
exports.SUBSCRIBE_LABEL_DEFAULT_TEXT = 'SUB';
exports.RECEIVE_TEXT = 'Receive';
exports.RECEIVE_TEXT_LABEL_DEFAULT_TEXT = 'RECEIVE';
exports.REQUEST_TEXT = 'Request';
exports.REQUEST_LABEL_DEFAULT_TEXT = 'REQUEST';
exports.REPLIER_TEXT = 'Reply';
exports.REPLIER_LABEL_DEFAULT_TEXT = 'REPLY';
exports.REQUIRED_TEXT = 'Required';
exports.GENERATED_TEXT = 'Generated';
exports.SERVERS_TEXT = 'Servers';
exports.OPERATIONS_TEXT = 'Operations';
exports.MESSAGES_TEXT = 'Messages';
exports.SCHEMAS_TEXT = 'Schemas';
exports.CHANNELS_TEXT = 'Channels';
exports.PARAMETERS_TEXT = 'Parameters';
exports.HEADERS_TEXT = 'Headers';
exports.MESSAGE_HEADERS_TEXT = 'Message Headers';
exports.HEADERS_EXAMPLE_TEXT = 'Example of headers';
exports.TAGS_TEXT = 'Tags';
exports.PAYLOAD_TEXT = 'Payload';
exports.MESSAGE_PAYLOAD_TEXT = 'Message Payload';
exports.PAYLOAD_EXAMPLE_TEXT = 'Example of payload';
exports.SCHEMA_EXAMPLE_TEXT = 'Example';
exports.SERVER_BINDINGS_TEXT = 'Server Bindings';
exports.CHANNEL_BINDINGS_TEXT = 'Channel Bindings';
exports.OPERATION_BINDINGS_TEXT = 'Operation Bindings';
exports.MESSAGE_BINDINGS_TEXT = 'Message Bindings';
exports.BINDINGS_SCHEMA_OBJECT_TEXT = 'Schema Object';
exports.NONE_TEXT = 'None';
exports.ANY_TEXT = 'Any';
exports.ERROR_TEXT = 'Error';
exports.EXPAND_ERROR_BUTTON_TEXT = 'Expand';
exports.COLLAPSE_ERROR_BUTTON_TEXT = 'Collapse';
exports.SECURITY_TEXT = 'Security';
exports.URL_VARIABLES_TEXT = 'URL Variables';
