// @backstage/integration/dist/index.cjs.js
'use strict';

var AwsS3Integration = require('./awsS3/AwsS3Integration.cjs.js');
var config = require('./awsS3/config.cjs.js');
var AwsCodeCommitIntegration = require('./awsCodeCommit/AwsCodeCommitIntegration.cjs.js');
var config$1 = require('./awsCodeCommit/config.cjs.js');
var AzureBlobStorageIntegration = require('./azureBlobStorage/AzureBlobStorageIntegration.cjs.js');
var config$2 = require('./azureBlobStorage/config.cjs.js');
var DefaultAzureCredentialsProvider = require('./azureBlobStorage/DefaultAzureCredentialsProvider.cjs.js');
var AzureIntegration = require('./azure/AzureIntegration.cjs.js');
var config$3 = require('./azure/config.cjs.js');
var core = require('./azure/core.cjs.js');
var DefaultAzureDevOpsCredentialsProvider = require('./azure/DefaultAzureDevOpsCredentialsProvider.cjs.js');
var deprecated = require('./azure/deprecated.cjs.js');
var BitbucketIntegration = require('./bitbucket/BitbucketIntegration.cjs.js');
var config$4 = require('./bitbucket/config.cjs.js');
var core$1 = require('./bitbucket/core.cjs.js');
var BitbucketCloudIntegration = require('./bitbucketCloud/BitbucketCloudIntegration.cjs.js');
var config$5 = require('./bitbucketCloud/config.cjs.js');
var core$2 = require('./bitbucketCloud/core.cjs.js');
var BitbucketServerIntegration = require('./bitbucketServer/BitbucketServerIntegration.cjs.js');
var config$6 = require('./bitbucketServer/config.cjs.js');
var core$3 = require('./bitbucketServer/core.cjs.js');
var GerritIntegration = require('./gerrit/GerritIntegration.cjs.js');
var config$7 = require('./gerrit/config.cjs.js');
var core$4 = require('./gerrit/core.cjs.js');
var GiteaIntegration = require('./gitea/GiteaIntegration.cjs.js');
var core$5 = require('./gitea/core.cjs.js');
var config$8 = require('./gitea/config.cjs.js');
var config$9 = require('./github/config.cjs.js');
var core$6 = require('./github/core.cjs.js');
var DefaultGithubCredentialsProvider = require('./github/DefaultGithubCredentialsProvider.cjs.js');
var SingleInstanceGithubCredentialsProvider = require('./github/SingleInstanceGithubCredentialsProvider.cjs.js');
var GithubIntegration = require('./github/GithubIntegration.cjs.js');
var config$a = require('./gitlab/config.cjs.js');
var core$7 = require('./gitlab/core.cjs.js');
var GitLabIntegration = require('./gitlab/GitLabIntegration.cjs.js');
var DefaultGitlabCredentialsProvider = require('./gitlab/DefaultGitlabCredentialsProvider.cjs.js');
var config$b = require('./googleGcs/config.cjs.js');
var HarnessIntegration = require('./harness/HarnessIntegration.cjs.js');
var core$8 = require('./harness/core.cjs.js');
var config$c = require('./harness/config.cjs.js');
var helpers = require('./helpers.cjs.js');
var ScmIntegrations = require('./ScmIntegrations.cjs.js');



exports.AwsS3Integration = AwsS3Integration.AwsS3Integration;
exports.readAwsS3IntegrationConfig = config.readAwsS3IntegrationConfig;
exports.readAwsS3IntegrationConfigs = config.readAwsS3IntegrationConfigs;
exports.AwsCodeCommitIntegration = AwsCodeCommitIntegration.AwsCodeCommitIntegration;
exports.readAwsCodeCommitIntegrationConfig = config$1.readAwsCodeCommitIntegrationConfig;
exports.readAwsCodeCommitIntegrationConfigs = config$1.readAwsCodeCommitIntegrationConfigs;
exports.AzureBlobStorageIntergation = AzureBlobStorageIntegration.AzureBlobStorageIntergation;
exports.readAzureBlobStorageIntegrationConfig = config$2.readAzureBlobStorageIntegrationConfig;
exports.readAzureBlobStorageIntegrationConfigs = config$2.readAzureBlobStorageIntegrationConfigs;
exports.DefaultAzureCredentialsManager = DefaultAzureCredentialsProvider.DefaultAzureCredentialsManager;
exports.AzureIntegration = AzureIntegration.AzureIntegration;
exports.readAzureIntegrationConfig = config$3.readAzureIntegrationConfig;
exports.readAzureIntegrationConfigs = config$3.readAzureIntegrationConfigs;
exports.getAzureCommitsUrl = core.getAzureCommitsUrl;
exports.getAzureDownloadUrl = core.getAzureDownloadUrl;
exports.getAzureFileFetchUrl = core.getAzureFileFetchUrl;
exports.DefaultAzureDevOpsCredentialsProvider = DefaultAzureDevOpsCredentialsProvider.DefaultAzureDevOpsCredentialsProvider;
exports.getAzureRequestOptions = deprecated.getAzureRequestOptions;
exports.BitbucketIntegration = BitbucketIntegration.BitbucketIntegration;
exports.readBitbucketIntegrationConfig = config$4.readBitbucketIntegrationConfig;
exports.readBitbucketIntegrationConfigs = config$4.readBitbucketIntegrationConfigs;
exports.getBitbucketDefaultBranch = core$1.getBitbucketDefaultBranch;
exports.getBitbucketDownloadUrl = core$1.getBitbucketDownloadUrl;
exports.getBitbucketFileFetchUrl = core$1.getBitbucketFileFetchUrl;
exports.getBitbucketRequestOptions = core$1.getBitbucketRequestOptions;
exports.BitbucketCloudIntegration = BitbucketCloudIntegration.BitbucketCloudIntegration;
exports.readBitbucketCloudIntegrationConfig = config$5.readBitbucketCloudIntegrationConfig;
exports.readBitbucketCloudIntegrationConfigs = config$5.readBitbucketCloudIntegrationConfigs;
exports.getBitbucketCloudDefaultBranch = core$2.getBitbucketCloudDefaultBranch;
exports.getBitbucketCloudDownloadUrl = core$2.getBitbucketCloudDownloadUrl;
exports.getBitbucketCloudFileFetchUrl = core$2.getBitbucketCloudFileFetchUrl;
exports.getBitbucketCloudRequestOptions = core$2.getBitbucketCloudRequestOptions;
exports.BitbucketServerIntegration = BitbucketServerIntegration.BitbucketServerIntegration;
exports.readBitbucketServerIntegrationConfig = config$6.readBitbucketServerIntegrationConfig;
exports.readBitbucketServerIntegrationConfigs = config$6.readBitbucketServerIntegrationConfigs;
exports.getBitbucketServerDefaultBranch = core$3.getBitbucketServerDefaultBranch;
exports.getBitbucketServerDownloadUrl = core$3.getBitbucketServerDownloadUrl;
exports.getBitbucketServerFileFetchUrl = core$3.getBitbucketServerFileFetchUrl;
exports.getBitbucketServerRequestOptions = core$3.getBitbucketServerRequestOptions;
exports.GerritIntegration = GerritIntegration.GerritIntegration;
exports.readGerritIntegrationConfig = config$7.readGerritIntegrationConfig;
exports.readGerritIntegrationConfigs = config$7.readGerritIntegrationConfigs;
exports.buildGerritGitilesArchiveUrl = core$4.buildGerritGitilesArchiveUrl;
exports.getGerritBranchApiUrl = core$4.getGerritBranchApiUrl;
exports.getGerritCloneRepoUrl = core$4.getGerritCloneRepoUrl;
exports.getGerritFileContentsApiUrl = core$4.getGerritFileContentsApiUrl;
exports.getGerritProjectsApiUrl = core$4.getGerritProjectsApiUrl;
exports.getGerritRequestOptions = core$4.getGerritRequestOptions;
exports.parseGerritGitilesUrl = core$4.parseGerritGitilesUrl;
exports.parseGerritJsonResponse = core$4.parseGerritJsonResponse;
exports.parseGitilesUrlRef = core$4.parseGitilesUrlRef;
exports.GiteaIntegration = GiteaIntegration.GiteaIntegration;
exports.getGiteaArchiveUrl = core$5.getGiteaArchiveUrl;
exports.getGiteaEditContentsUrl = core$5.getGiteaEditContentsUrl;
exports.getGiteaFileContentsUrl = core$5.getGiteaFileContentsUrl;
exports.getGiteaLatestCommitUrl = core$5.getGiteaLatestCommitUrl;
exports.getGiteaRequestOptions = core$5.getGiteaRequestOptions;
exports.parseGiteaUrl = core$5.parseGiteaUrl;
exports.readGiteaConfig = config$8.readGiteaConfig;
exports.readGithubIntegrationConfig = config$9.readGithubIntegrationConfig;
exports.readGithubIntegrationConfigs = config$9.readGithubIntegrationConfigs;
exports.getGitHubRequestOptions = core$6.getGitHubRequestOptions;
exports.getGithubFileFetchUrl = core$6.getGithubFileFetchUrl;
exports.DefaultGithubCredentialsProvider = DefaultGithubCredentialsProvider.DefaultGithubCredentialsProvider;
exports.GithubAppCredentialsMux = SingleInstanceGithubCredentialsProvider.GithubAppCredentialsMux;
exports.SingleInstanceGithubCredentialsProvider = SingleInstanceGithubCredentialsProvider.SingleInstanceGithubCredentialsProvider;
exports.GithubIntegration = GithubIntegration.GithubIntegration;
exports.replaceGithubUrlType = GithubIntegration.replaceGithubUrlType;
exports.getGitLabIntegrationRelativePath = config$a.getGitLabIntegrationRelativePath;
exports.readGitLabIntegrationConfig = config$a.readGitLabIntegrationConfig;
exports.readGitLabIntegrationConfigs = config$a.readGitLabIntegrationConfigs;
exports.getGitLabFileFetchUrl = core$7.getGitLabFileFetchUrl;
exports.getGitLabRequestOptions = core$7.getGitLabRequestOptions;
exports.GitLabIntegration = GitLabIntegration.GitLabIntegration;
exports.replaceGitLabUrlType = GitLabIntegration.replaceGitLabUrlType;
exports.DefaultGitlabCredentialsProvider = DefaultGitlabCredentialsProvider.DefaultGitlabCredentialsProvider;
exports.readGoogleGcsIntegrationConfig = config$b.readGoogleGcsIntegrationConfig;
exports.HarnessIntegration = HarnessIntegration.HarnessIntegration;
exports.getHarnessArchiveUrl = core$8.getHarnessArchiveUrl;
exports.getHarnessFileContentsUrl = core$8.getHarnessFileContentsUrl;
exports.getHarnessLatestCommitUrl = core$8.getHarnessLatestCommitUrl;
exports.getHarnessRequestOptions = core$8.getHarnessRequestOptions;
exports.parseHarnessUrl = core$8.parseHarnessUrl;
exports.readHarnessConfig = config$c.readHarnessConfig;
exports.defaultScmResolveUrl = helpers.defaultScmResolveUrl;
exports.ScmIntegrations = ScmIntegrations.ScmIntegrations;
