// @backstage/plugin-scaffolder-backend-module-bitbucket/dist/actions/bitbucket.examples.cjs.js
'use strict';

var yaml = require('yaml');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var yaml__default = /*#__PURE__*/_interopDefaultCompat(yaml);

const examples = [
  {
    description: "Initializes a git repository with the content in the workspace, and publishes it to Bitbucket with the default configuration.",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "publish",
          action: "publish:bitbucket",
          name: "Publish to Bitbucket",
          input: {
            repoUrl: "bitbucket.org?repo=repo&workspace=workspace&project=project"
          }
        }
      ]
    })
  },
  {
    description: "Initializes a Bitbucket repository with a description.",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "publish",
          action: "publish:bitbucket",
          name: "Publish to Bitbucket",
          input: {
            repoUrl: "bitbucket.org?repo=repo&workspace=workspace&project=project",
            description: "Initialize a git repository"
          }
        }
      ]
    })
  },
  {
    description: "Initializes a Bitbucket repository with public repo visibility, if not set defaults to private",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "publish",
          action: "publish:bitbucket",
          name: "Publish to Bitbucket",
          input: {
            repoUrl: "bitbucket.org?repo=repo&workspace=workspace&project=project",
            repoVisibility: "public"
          }
        }
      ]
    })
  },
  {
    description: "Initializes a Bitbucket repository with a default branch, if not set defaults to master",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "publish",
          action: "publish:bitbucket",
          name: "Publish to Bitbucket",
          input: {
            repoUrl: "bitbucket.org?repo=repo&workspace=workspace&project=project",
            defaultBranch: "main"
          }
        }
      ]
    })
  },
  {
    description: "Path within the workspace that will be used as the repository root. If omitted, the entire workspace will be published as the repository",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "publish",
          action: "publish:bitbucket",
          name: "Publish to Bitbucket",
          input: {
            repoUrl: "bitbucket.org?repo=repo&workspace=workspace&project=project",
            sourcePath: "./repoRoot"
          }
        }
      ]
    })
  },
  {
    description: "Initializes a Bitbucket repository with LFS enabled",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "publish",
          action: "publish:bitbucket",
          name: "Publish to Bitbucket",
          input: {
            repoUrl: "hosted.bitbucket.com?project=project&repo=repo",
            enableLFS: true
          }
        }
      ]
    })
  },
  {
    description: "Initializes a Bitbucket repository with a custom authentication token",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "publish",
          action: "publish:bitbucket",
          name: "Publish to Bitbucket",
          input: {
            repoUrl: "bitbucket.org?repo=repo&workspace=workspace&project=project",
            token: "your-auth-token"
          }
        }
      ]
    })
  },
  {
    description: "Initializes a Bitbucket repository with an initial commit message, if not set defaults to `initial commit`",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "publish",
          action: "publish:bitbucket",
          name: "Publish to Bitbucket",
          input: {
            repoUrl: "bitbucket.org?repo=repo&workspace=workspace&project=project",
            gitCommitMessage: "Initial commit with custom message"
          }
        }
      ]
    })
  },
  {
    description: "Initializes a Bitbucket repository with a custom author",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "publish",
          action: "publish:bitbucket",
          name: "Publish to Bitbucket",
          input: {
            repoUrl: "bitbucket.org?repo=repo&workspace=workspace&project=project",
            gitAuthorName: "Your Name",
            gitAuthorEmail: "your.email@example.com"
          }
        }
      ]
    })
  },
  {
    description: "Initializes a Bitbucket repository with all proporties being set",
    example: yaml__default.default.stringify({
      steps: [
        {
          id: "publish",
          action: "publish:bitbucket",
          name: "Publish to Bitbucket",
          input: {
            repoUrl: "bitbucket.org?repo=repo&workspace=workspace&project=project",
            description: "Initialize a git repository",
            repoVisibility: "public",
            defaultBranch: "main",
            token: "your-auth-token",
            gitCommitMessage: "Initial commit with custom message",
            gitAuthorName: "Your Name",
            gitAuthorEmail: "your.email@example.com"
          }
        }
      ]
    })
  }
];

exports.examples = examples;
