// @backstage/plugin-scaffolder-backend/dist/scaffolder/actions/builtin/fetch/plain.examples.cjs.js
'use strict';

var yaml = require('yaml');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var yaml__default = /*#__PURE__*/_interopDefaultCompat(yaml);

const examples = [
  {
    description: "Downloads content and places it in the workspace.",
    example: yaml__default.default.stringify({
      steps: [
        {
          action: "fetch:plain",
          id: "fetch-plain",
          name: "Fetch plain",
          input: {
            url: "https://github.com/backstage/community/tree/main/backstage-community-sessions/assets"
          }
        }
      ]
    })
  },
  {
    description: "Optionally, if you would prefer the data to be downloaded to a subdirectory in the workspace you may specify the \u2018targetPath\u2019 input option.",
    example: yaml__default.default.stringify({
      steps: [
        {
          action: "fetch:plain",
          id: "fetch-plain",
          name: "Fetch plain",
          input: {
            url: "https://github.com/backstage/community/tree/main/backstage-community-sessions/assets",
            targetPath: "fetched-data"
          }
        }
      ]
    })
  }
];

exports.examples = examples;
