// @backstage/plugin-scaffolder-backend/dist/scaffolder/actions/builtin/fetch/plainFile.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var plainFile_examples = require('./plainFile.examples.cjs.js');
var pluginScaffolderNode = require('@backstage/plugin-scaffolder-node');

function createFetchPlainFileAction(options) {
  const { reader, integrations } = options;
  return pluginScaffolderNode.createTemplateAction({
    id: "fetch:plain:file",
    description: "Downloads single file and places it in the workspace.",
    examples: plainFile_examples.examples,
    schema: {
      input: {
        type: "object",
        required: ["url", "targetPath"],
        properties: {
          url: {
            title: "Fetch URL",
            description: "Relative path or absolute URL pointing to the single file to fetch.",
            type: "string"
          },
          targetPath: {
            title: "Target Path",
            description: "Target path within the working directory to download the file as.",
            type: "string"
          },
          token: {
            title: "Token",
            description: "An optional token to use for authentication when reading the resources.",
            type: "string"
          }
        }
      }
    },
    supportsDryRun: true,
    async handler(ctx) {
      ctx.logger.info("Fetching plain content from remote URL");
      const outputPath = backendPluginApi.resolveSafeChildPath(
        ctx.workspacePath,
        ctx.input.targetPath
      );
      await pluginScaffolderNode.fetchFile({
        reader,
        integrations,
        baseUrl: ctx.templateInfo?.baseUrl,
        fetchUrl: ctx.input.url,
        outputPath,
        token: ctx.input.token
      });
    }
  });
}

exports.createFetchPlainFileAction = createFetchPlainFileAction;
