// @backstage/plugin-scaffolder-node/dist/index.cjs.js
'use strict';

var createTemplateAction = require('./actions/createTemplateAction.cjs.js');
var executeShellCommand = require('./actions/executeShellCommand.cjs.js');
var fetch = require('./actions/fetch.cjs.js');
var gitHelpers = require('./actions/gitHelpers.cjs.js');
var util = require('./actions/util.cjs.js');
var serializeDirectoryContents = require('./files/serializeDirectoryContents.cjs.js');
var deserializeDirectoryContents = require('./files/deserializeDirectoryContents.cjs.js');



exports.createTemplateAction = createTemplateAction.createTemplateAction;
exports.executeShellCommand = executeShellCommand.executeShellCommand;
exports.fetchContents = fetch.fetchContents;
exports.fetchFile = fetch.fetchFile;
exports.addFiles = gitHelpers.addFiles;
exports.cloneRepo = gitHelpers.cloneRepo;
exports.commitAndPushBranch = gitHelpers.commitAndPushBranch;
exports.commitAndPushRepo = gitHelpers.commitAndPushRepo;
exports.createBranch = gitHelpers.createBranch;
exports.initRepoAndPush = gitHelpers.initRepoAndPush;
exports.getRepoSourceDirectory = util.getRepoSourceDirectory;
exports.parseRepoUrl = util.parseRepoUrl;
exports.serializeDirectoryContents = serializeDirectoryContents.serializeDirectoryContents;
exports.deserializeDirectoryContents = deserializeDirectoryContents.deserializeDirectoryContents;
