// @backstage/plugin-auth-backend/dist/service/router.cjs.js
'use strict';

var express = require('express');
var Router = require('express-promise-router');
var cookieParser = require('cookie-parser');
var providers = require('../providers/providers.cjs.js');
var router = require('../providers/router.cjs.js');
require('@backstage/plugin-auth-node');
var backendCommon = require('@backstage/backend-common');
var errors = require('@backstage/errors');
var router$1 = require('../identity/router.cjs.js');
var TokenFactory = require('../identity/TokenFactory.cjs.js');
require('luxon');
require('@google-cloud/firestore');
var KeyStores = require('../identity/KeyStores.cjs.js');
var UserInfoDatabaseHandler = require('../identity/UserInfoDatabaseHandler.cjs.js');
var session = require('express-session');
var connectSessionKnex = require('connect-session-knex');
var passport = require('passport');
var AuthDatabase = require('../database/AuthDatabase.cjs.js');
var readBackstageTokenExpiration = require('./readBackstageTokenExpiration.cjs.js');
var StaticTokenIssuer = require('../identity/StaticTokenIssuer.cjs.js');
var StaticKeyStore = require('../identity/StaticKeyStore.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var express__default = /*#__PURE__*/_interopDefaultCompat(express);
var Router__default = /*#__PURE__*/_interopDefaultCompat(Router);
var cookieParser__default = /*#__PURE__*/_interopDefaultCompat(cookieParser);
var session__default = /*#__PURE__*/_interopDefaultCompat(session);
var connectSessionKnex__default = /*#__PURE__*/_interopDefaultCompat(connectSessionKnex);
var passport__default = /*#__PURE__*/_interopDefaultCompat(passport);

async function createRouter(options) {
  const {
    logger,
    config,
    discovery,
    database,
    tokenFactoryAlgorithm,
    providerFactories = {}
  } = options;
  const { auth, httpAuth } = backendCommon.createLegacyAuthAdapters(options);
  const router$2 = Router__default.default();
  const appUrl = config.getString("app.baseUrl");
  const authUrl = await discovery.getExternalBaseUrl("auth");
  const backstageTokenExpiration = readBackstageTokenExpiration.readBackstageTokenExpiration(config);
  const authDb = AuthDatabase.AuthDatabase.create(database);
  const keyStore = await KeyStores.KeyStores.fromConfig(config, {
    logger,
    database: authDb
  });
  const userInfoDatabaseHandler = new UserInfoDatabaseHandler.UserInfoDatabaseHandler(
    await authDb.get()
  );
  let tokenIssuer;
  if (keyStore instanceof StaticKeyStore.StaticKeyStore) {
    tokenIssuer = new StaticTokenIssuer.StaticTokenIssuer(
      {
        logger: logger.child({ component: "token-factory" }),
        issuer: authUrl,
        sessionExpirationSeconds: backstageTokenExpiration
      },
      keyStore
    );
  } else {
    tokenIssuer = new TokenFactory.TokenFactory({
      issuer: authUrl,
      keyStore,
      keyDurationSeconds: backstageTokenExpiration,
      logger: logger.child({ component: "token-factory" }),
      algorithm: tokenFactoryAlgorithm ?? config.getOptionalString("auth.identityTokenAlgorithm"),
      userInfoDatabaseHandler
    });
  }
  const secret = config.getOptionalString("auth.session.secret");
  if (secret) {
    router$2.use(cookieParser__default.default(secret));
    const enforceCookieSSL = authUrl.startsWith("https");
    const KnexSessionStore = connectSessionKnex__default.default(session__default.default);
    router$2.use(
      session__default.default({
        secret,
        saveUninitialized: false,
        resave: false,
        cookie: { secure: enforceCookieSSL ? "auto" : false },
        store: new KnexSessionStore({
          createtable: false,
          knex: await authDb.get()
        })
      })
    );
    router$2.use(passport__default.default.initialize());
    router$2.use(passport__default.default.session());
  } else {
    router$2.use(cookieParser__default.default());
  }
  router$2.use(express__default.default.urlencoded({ extended: false }));
  router$2.use(express__default.default.json());
  const providers$1 = options.disableDefaultProviderFactories ? providerFactories : {
    ...providers.defaultAuthProviderFactories,
    ...providerFactories
  };
  router.bindProviderRouters(router$2, {
    providers: providers$1,
    appUrl,
    baseUrl: authUrl,
    tokenIssuer,
    ...options,
    auth,
    httpAuth
  });
  router$1.bindOidcRouter(router$2, {
    auth,
    tokenIssuer,
    baseUrl: authUrl,
    userInfoDatabaseHandler
  });
  router$2.use("/:provider/", (req) => {
    const { provider } = req.params;
    throw new errors.NotFoundError(`Unknown auth provider '${provider}'`);
  });
  return router$2;
}

exports.createRouter = createRouter;
