// @backstage/plugin-auth-backend/dist/providers/providers.cjs.js
'use strict';

var provider$e = require('./atlassian/provider.cjs.js');
var provider$5 = require('./auth0/provider.cjs.js');
var provider$b = require('./aws-alb/provider.cjs.js');
var provider$c = require('./bitbucket/provider.cjs.js');
var provider$f = require('./cloudflare-access/provider.cjs.js');
var provider$g = require('./gcp-iap/provider.cjs.js');
var provider$1 = require('./github/provider.cjs.js');
var provider$2 = require('./gitlab/provider.cjs.js');
var provider = require('./google/provider.cjs.js');
var provider$6 = require('./microsoft/provider.cjs.js');
var provider$8 = require('./oauth2/provider.cjs.js');
var provider$h = require('./oauth2-proxy/provider.cjs.js');
var provider$9 = require('./oidc/provider.cjs.js');
var provider$4 = require('./okta/provider.cjs.js');
var provider$a = require('./onelogin/provider.cjs.js');
var provider$3 = require('./saml/provider.cjs.js');
var provider$d = require('./bitbucketServer/provider.cjs.js');
var provider$7 = require('./azure-easyauth/provider.cjs.js');

const providers = Object.freeze({
  atlassian: provider$e.atlassian,
  auth0: provider$5.auth0,
  awsAlb: provider$b.awsAlb,
  bitbucket: provider$c.bitbucket,
  bitbucketServer: provider$d.bitbucketServer,
  cfAccess: provider$f.cfAccess,
  gcpIap: provider$g.gcpIap,
  github: provider$1.github,
  gitlab: provider$2.gitlab,
  google: provider.google,
  microsoft: provider$6.microsoft,
  oauth2: provider$8.oauth2,
  oauth2Proxy: provider$h.oauth2Proxy,
  oidc: provider$9.oidc,
  okta: provider$4.okta,
  onelogin: provider$a.onelogin,
  saml: provider$3.saml,
  easyAuth: provider$7.easyAuth
});
const defaultAuthProviderFactories = {
  google: provider.google.create(),
  github: provider$1.github.create(),
  gitlab: provider$2.gitlab.create(),
  saml: provider$3.saml.create(),
  okta: provider$4.okta.create(),
  auth0: provider$5.auth0.create(),
  microsoft: provider$6.microsoft.create(),
  easyAuth: provider$7.easyAuth.create(),
  oauth2: provider$8.oauth2.create(),
  oidc: provider$9.oidc.create(),
  onelogin: provider$a.onelogin.create(),
  awsalb: provider$b.awsAlb.create(),
  bitbucket: provider$c.bitbucket.create(),
  bitbucketServer: provider$d.bitbucketServer.create(),
  atlassian: provider$e.atlassian.create()
};

exports.defaultAuthProviderFactories = defaultAuthProviderFactories;
exports.providers = providers;
