// @backstage/plugin-bitbucket-cloud-common/dist/BitbucketCloudClient.cjs.js
'use strict';

var fetch = require('cross-fetch');
var pagination = require('./pagination.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fetch__default = /*#__PURE__*/_interopDefaultCompat(fetch);

class BitbucketCloudClient {
  constructor(config) {
    this.config = config;
  }
  static fromConfig(config) {
    return new BitbucketCloudClient(config);
  }
  searchCode(workspace, query, options) {
    const workspaceEnc = encodeURIComponent(workspace);
    return new pagination.WithPagination(
      (paginationOptions) => this.createUrl(`/workspaces/${workspaceEnc}/search/code`, {
        ...paginationOptions,
        ...options,
        search_query: query
      }),
      (url) => this.getTypeMapped(url)
    );
  }
  listRepositoriesByWorkspace(workspace, options) {
    const workspaceEnc = encodeURIComponent(workspace);
    return new pagination.WithPagination(
      (paginationOptions) => this.createUrl(`/repositories/${workspaceEnc}`, {
        ...paginationOptions,
        ...options
      }),
      (url) => this.getTypeMapped(url)
    );
  }
  listProjectsByWorkspace(workspace, options) {
    const workspaceEnc = encodeURIComponent(workspace);
    return new pagination.WithPagination(
      (paginationOptions) => this.createUrl(`/workspaces/${workspaceEnc}/projects`, {
        ...paginationOptions,
        ...options
      }),
      (url) => this.getTypeMapped(url)
    );
  }
  listWorkspaces(options) {
    return new pagination.WithPagination(
      (paginationOptions) => this.createUrl("/workspaces", { ...paginationOptions, ...options }),
      (url) => this.getTypeMapped(url)
    );
  }
  listBranchesByRepository(repository, workspace, options) {
    const workspaceEnc = encodeURIComponent(workspace);
    return new pagination.WithPagination(
      (paginationOptions) => this.createUrl(
        `/repositories/${workspaceEnc}/${repository}/refs/branches`,
        {
          ...paginationOptions,
          ...options
        }
      ),
      (url) => this.getTypeMapped(url)
    );
  }
  createUrl(endpoint, options) {
    const request = new URL(this.config.apiBaseUrl + endpoint);
    for (const key in options) {
      if (options[key]) {
        request.searchParams.append(key, options[key].toString());
      }
    }
    return request;
  }
  async getTypeMapped(url) {
    return this.get(url).then(
      (response) => response.json()
    );
  }
  async get(url) {
    return this.request(new fetch.Request(url.toString(), { method: "GET" }));
  }
  async request(req) {
    return fetch__default.default(req, { headers: this.getAuthHeaders() }).then(
      (response) => {
        if (!response.ok) {
          throw new Error(
            `Unexpected response for ${req.method} ${req.url}. Expected 200 but got ${response.status} - ${response.statusText}`
          );
        }
        return response;
      }
    );
  }
  getAuthHeaders() {
    const headers = {};
    if (this.config.username) {
      const buffer = Buffer.from(
        `${this.config.username}:${this.config.appPassword}`,
        "utf8"
      );
      headers.Authorization = `Basic ${buffer.toString("base64")}`;
    } else if (this.config.token) {
      headers.Authorization = `Bearer ${this.config.token}`;
    }
    return headers;
  }
}

exports.BitbucketCloudClient = BitbucketCloudClient;
