// @backstage/plugin-catalog/dist/components/RelatedEntitiesCard/presets.esm.js
import { EntityTable } from '@backstage/plugin-catalog-react';

const componentEntityColumns = [
  EntityTable.columns.createEntityRefColumn({ defaultKind: "component" }),
  EntityTable.columns.createOwnerColumn(),
  EntityTable.columns.createSpecTypeColumn(),
  EntityTable.columns.createSpecLifecycleColumn(),
  EntityTable.columns.createMetadataDescriptionColumn()
];
const componentEntityHelpLink = "https://backstage.io/docs/features/software-catalog/descriptor-format#kind-component";
const asComponentEntities = (entities) => entities;
const resourceEntityColumns = [
  EntityTable.columns.createEntityRefColumn({ defaultKind: "resource" }),
  EntityTable.columns.createOwnerColumn(),
  EntityTable.columns.createSpecTypeColumn(),
  EntityTable.columns.createSpecLifecycleColumn(),
  EntityTable.columns.createMetadataDescriptionColumn()
];
const resourceEntityHelpLink = "https://backstage.io/docs/features/software-catalog/descriptor-format#kind-resource";
const asResourceEntities = (entities) => entities;
const systemEntityColumns = [
  EntityTable.columns.createEntityRefColumn({ defaultKind: "system" }),
  EntityTable.columns.createOwnerColumn(),
  EntityTable.columns.createMetadataDescriptionColumn()
];
const systemEntityHelpLink = "https://backstage.io/docs/features/software-catalog/descriptor-format#kind-system";
const asSystemEntities = (entities) => entities;
const domainEntityColumns = [
  EntityTable.columns.createEntityRefColumn({ defaultKind: "domain" }),
  EntityTable.columns.createOwnerColumn(),
  EntityTable.columns.createMetadataDescriptionColumn()
];
const domainEntityHelpLink = "https://backstage.io/docs/features/software-catalog/descriptor-format#kind-domain";
const asDomainEntities = (entities) => entities;

export { asComponentEntities, asDomainEntities, asResourceEntities, asSystemEntities, componentEntityColumns, componentEntityHelpLink, domainEntityColumns, domainEntityHelpLink, resourceEntityColumns, resourceEntityHelpLink, systemEntityColumns, systemEntityHelpLink };
