// @backstage/frontend-defaults/dist/createPublicSignInApp.esm.js
import { createFrontendModule, coreExtensionData, useApi, identityApiRef } from '@backstage/frontend-plugin-api';
import React from 'react';
import { useAsync, useMountEffect } from '@react-hookz/web';
import { createApp } from './createApp.esm.js';
import appPlugin from '@backstage/plugin-app';

function InternalCookieAuthRedirect() {
  const identityApi = useApi(identityApiRef);
  const [state, actions] = useAsync(async () => {
    const { token } = await identityApi.getCredentials();
    if (!token) {
      throw new Error("Expected Backstage token in sign-in response");
    }
    return token;
  });
  useMountEffect(actions.execute);
  if (state.status === "error" && state.error) {
    return /* @__PURE__ */ React.createElement(React.Fragment, null, "An error occurred: ", state.error.message);
  }
  if (state.status === "success" && state.result) {
    return /* @__PURE__ */ React.createElement(
      "form",
      {
        ref: (form) => form?.submit(),
        action: window.location.href,
        method: "POST",
        style: { visibility: "hidden" }
      },
      /* @__PURE__ */ React.createElement("input", { type: "hidden", name: "type", value: "sign-in" }),
      /* @__PURE__ */ React.createElement("input", { type: "hidden", name: "token", value: state.result }),
      /* @__PURE__ */ React.createElement("input", { type: "submit", value: "Continue" })
    );
  }
  return null;
}
function createPublicSignInApp(options) {
  return createApp({
    ...options,
    features: [
      ...options?.features ?? [],
      // This is a rather than app plugin override in order for it to take precedence over any supplied app plugin override
      createFrontendModule({
        pluginId: "app",
        extensions: [
          appPlugin.getExtension("app/layout").override({
            factory: () => [
              coreExtensionData.reactElement(/* @__PURE__ */ React.createElement(InternalCookieAuthRedirect, null))
            ]
          })
        ]
      })
    ]
  });
}

export { InternalCookieAuthRedirect, createPublicSignInApp };
