// @backstage/plugin-scaffolder-backend-module-github/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var githubActionsDispatch = require('./actions/githubActionsDispatch.cjs.js');
var githubIssuesLabel = require('./actions/githubIssuesLabel.cjs.js');
var githubRepoCreate = require('./actions/githubRepoCreate.cjs.js');
var githubRepoPush = require('./actions/githubRepoPush.cjs.js');
var githubWebhook = require('./actions/githubWebhook.cjs.js');
var githubDeployKey = require('./actions/githubDeployKey.cjs.js');
var githubEnvironment = require('./actions/githubEnvironment.cjs.js');
var githubPullRequest = require('./actions/githubPullRequest.cjs.js');
var github = require('./actions/github.cjs.js');
var githubAutolinks = require('./actions/githubAutolinks.cjs.js');
var githubPagesEnable = require('./actions/githubPagesEnable.cjs.js');
var githubBranchProtection = require('./actions/githubBranchProtection.cjs.js');
var helpers = require('./actions/helpers.cjs.js');
var module$1 = require('./module.cjs.js');



exports.createGithubActionsDispatchAction = githubActionsDispatch.createGithubActionsDispatchAction;
exports.createGithubIssuesLabelAction = githubIssuesLabel.createGithubIssuesLabelAction;
exports.createGithubRepoCreateAction = githubRepoCreate.createGithubRepoCreateAction;
exports.createGithubRepoPushAction = githubRepoPush.createGithubRepoPushAction;
exports.createGithubWebhookAction = githubWebhook.createGithubWebhookAction;
exports.createGithubDeployKeyAction = githubDeployKey.createGithubDeployKeyAction;
exports.createGithubEnvironmentAction = githubEnvironment.createGithubEnvironmentAction;
exports.createPublishGithubPullRequestAction = githubPullRequest.createPublishGithubPullRequestAction;
exports.createPublishGithubAction = github.createPublishGithubAction;
exports.createGithubAutolinksAction = githubAutolinks.createGithubAutolinksAction;
exports.createGithubPagesEnableAction = githubPagesEnable.createGithubPagesEnableAction;
exports.createGithubBranchProtectionAction = githubBranchProtection.createGithubBranchProtectionAction;
exports.getOctokitOptions = helpers.getOctokitOptions;
exports.default = module$1.githubModule;
