// @asyncapi/parser/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiagnosticSeverity = exports.isOldAsyncAPIDocument = exports.isAsyncAPIDocument = exports.toAsyncAPIDocument = exports.createAsyncAPIDocument = exports.fromFile = exports.fromURL = exports.unstringify = exports.stringify = exports.Parser = void 0;
const types_1 = require("@stoplight/types");
Object.defineProperty(exports, "DiagnosticSeverity", { enumerable: true, get: function () { return types_1.DiagnosticSeverity; } });
const parser_1 = require("./parser");
Object.defineProperty(exports, "Parser", { enumerable: true, get: function () { return parser_1.Parser; } });
__exportStar(require("./models"), exports);
var stringify_1 = require("./stringify");
Object.defineProperty(exports, "stringify", { enumerable: true, get: function () { return stringify_1.stringify; } });
Object.defineProperty(exports, "unstringify", { enumerable: true, get: function () { return stringify_1.unstringify; } });
var from_1 = require("./from");
Object.defineProperty(exports, "fromURL", { enumerable: true, get: function () { return from_1.fromURL; } });
Object.defineProperty(exports, "fromFile", { enumerable: true, get: function () { return from_1.fromFile; } });
var document_1 = require("./document");
Object.defineProperty(exports, "createAsyncAPIDocument", { enumerable: true, get: function () { return document_1.createAsyncAPIDocument; } });
Object.defineProperty(exports, "toAsyncAPIDocument", { enumerable: true, get: function () { return document_1.toAsyncAPIDocument; } });
Object.defineProperty(exports, "isAsyncAPIDocument", { enumerable: true, get: function () { return document_1.isAsyncAPIDocument; } });
Object.defineProperty(exports, "isOldAsyncAPIDocument", { enumerable: true, get: function () { return document_1.isOldAsyncAPIDocument; } });
__exportStar(require("./old-api"), exports);
exports.default = parser_1.Parser;
