// @backstage/plugin-scaffolder-react/dist/next/components/PasswordWidget/PasswordWidget.esm.js
import TextField from '@material-ui/core/TextField';
import React from 'react';
import FormHelperText from '@material-ui/core/FormHelperText';
import { MarkdownContent } from '@backstage/core-components';

const PasswordWidget = (props) => {
  const {
    value,
    onChange,
    schema: { title }
  } = props;
  return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(
    TextField,
    {
      id: title,
      label: title,
      "aria-describedby": title,
      onChange: (e) => {
        onChange(e.target.value);
      },
      value,
      autoComplete: "off"
    }
  ), /* @__PURE__ */ React.createElement(FormHelperText, { error: true }, /* @__PURE__ */ React.createElement(
    MarkdownContent,
    {
      content: "This widget is insecure. Please use [`ui:field: Secret`](https://backstage.io/docs/features/software-templates/writing-templates/#using-secrets) instead of\n          `ui:widget: password`"
    }
  )));
};

export { PasswordWidget };
