// @backstage/plugin-scaffolder-backend/dist/scaffolder/actions/deprecated.cjs.js
'use strict';

var github = require('@backstage/plugin-scaffolder-backend-module-github');
var gitlab = require('@backstage/plugin-scaffolder-backend-module-gitlab');
var azure = require('@backstage/plugin-scaffolder-backend-module-azure');
var bitbucket = require('@backstage/plugin-scaffolder-backend-module-bitbucket');
var bitbucketCloud = require('@backstage/plugin-scaffolder-backend-module-bitbucket-cloud');
var bitbucketServer = require('@backstage/plugin-scaffolder-backend-module-bitbucket-server');
var gerrit = require('@backstage/plugin-scaffolder-backend-module-gerrit');

function _interopNamespaceCompat(e) {
  if (e && typeof e === 'object' && 'default' in e) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var github__namespace = /*#__PURE__*/_interopNamespaceCompat(github);
var gitlab__namespace = /*#__PURE__*/_interopNamespaceCompat(gitlab);
var azure__namespace = /*#__PURE__*/_interopNamespaceCompat(azure);
var bitbucket__namespace = /*#__PURE__*/_interopNamespaceCompat(bitbucket);
var bitbucketCloud__namespace = /*#__PURE__*/_interopNamespaceCompat(bitbucketCloud);
var bitbucketServer__namespace = /*#__PURE__*/_interopNamespaceCompat(bitbucketServer);
var gerrit__namespace = /*#__PURE__*/_interopNamespaceCompat(gerrit);

const createGithubActionsDispatchAction = github__namespace.createGithubActionsDispatchAction;
const createGithubDeployKeyAction = github__namespace.createGithubDeployKeyAction;
const createGithubEnvironmentAction = github__namespace.createGithubEnvironmentAction;
const createGithubIssuesLabelAction = github__namespace.createGithubIssuesLabelAction;
const createGithubRepoCreateAction = github__namespace.createGithubRepoCreateAction;
const createGithubRepoPushAction = github__namespace.createGithubRepoPushAction;
const createGithubWebhookAction = github__namespace.createGithubWebhookAction;
const createPublishGithubAction = github__namespace.createPublishGithubAction;
const createPublishGithubPullRequestAction = github__namespace.createPublishGithubPullRequestAction;
const createPublishBitbucketAction = bitbucket__namespace.createPublishBitbucketAction;
const createPublishBitbucketCloudAction = bitbucketCloud__namespace.createPublishBitbucketCloudAction;
const createPublishBitbucketServerAction = bitbucketServer__namespace.createPublishBitbucketServerAction;
const createPublishBitbucketServerPullRequestAction = bitbucketServer__namespace.createPublishBitbucketServerPullRequestAction;
const createPublishAzureAction = azure__namespace.createPublishAzureAction;
const createPublishGerritAction = gerrit__namespace.createPublishGerritAction;
const createPublishGerritReviewAction = gerrit__namespace.createPublishGerritReviewAction;
const createPublishGitlabAction = gitlab__namespace.createPublishGitlabAction;
const createPublishGitlabMergeRequestAction = gitlab__namespace.createPublishGitlabMergeRequestAction;

exports.createGithubActionsDispatchAction = createGithubActionsDispatchAction;
exports.createGithubDeployKeyAction = createGithubDeployKeyAction;
exports.createGithubEnvironmentAction = createGithubEnvironmentAction;
exports.createGithubIssuesLabelAction = createGithubIssuesLabelAction;
exports.createGithubRepoCreateAction = createGithubRepoCreateAction;
exports.createGithubRepoPushAction = createGithubRepoPushAction;
exports.createGithubWebhookAction = createGithubWebhookAction;
exports.createPublishAzureAction = createPublishAzureAction;
exports.createPublishBitbucketAction = createPublishBitbucketAction;
exports.createPublishBitbucketCloudAction = createPublishBitbucketCloudAction;
exports.createPublishBitbucketServerAction = createPublishBitbucketServerAction;
exports.createPublishBitbucketServerPullRequestAction = createPublishBitbucketServerPullRequestAction;
exports.createPublishGerritAction = createPublishGerritAction;
exports.createPublishGerritReviewAction = createPublishGerritReviewAction;
exports.createPublishGithubAction = createPublishGithubAction;
exports.createPublishGithubPullRequestAction = createPublishGithubPullRequestAction;
exports.createPublishGitlabAction = createPublishGitlabAction;
exports.createPublishGitlabMergeRequestAction = createPublishGitlabMergeRequestAction;
