// @swagger-api/apidom-ns-openapi-2/src/refractor/registration.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
exports.__esModule = true;
var _Swagger = _interopRequireDefault(require("../elements/Swagger.cjs"));
exports.SwaggerElement = _Swagger.default;
var _SwaggerVersion = _interopRequireDefault(require("../elements/SwaggerVersion.cjs"));
exports.SwaggerVersionElement = _SwaggerVersion.default;
var _Info = _interopRequireDefault(require("../elements/Info.cjs"));
exports.InfoElement = _Info.default;
var _Contact = _interopRequireDefault(require("../elements/Contact.cjs"));
exports.ContactElement = _Contact.default;
var _License = _interopRequireDefault(require("../elements/License.cjs"));
exports.LicenseElement = _License.default;
var _Paths = _interopRequireDefault(require("../elements/Paths.cjs"));
exports.PathsElement = _Paths.default;
var _PathItem = _interopRequireDefault(require("../elements/PathItem.cjs"));
exports.PathItemElement = _PathItem.default;
var _Operation = _interopRequireDefault(require("../elements/Operation.cjs"));
exports.OperationElement = _Operation.default;
var _ExternalDocumentation = _interopRequireDefault(require("../elements/ExternalDocumentation.cjs"));
exports.ExternalDocumentationElement = _ExternalDocumentation.default;
var _Parameter = _interopRequireDefault(require("../elements/Parameter.cjs"));
exports.ParameterElement = _Parameter.default;
var _Items = _interopRequireDefault(require("../elements/Items.cjs"));
exports.ItemsElement = _Items.default;
var _Responses = _interopRequireDefault(require("../elements/Responses.cjs"));
exports.ResponsesElement = _Responses.default;
var _Response = _interopRequireDefault(require("../elements/Response.cjs"));
exports.ResponseElement = _Response.default;
var _Headers = _interopRequireDefault(require("../elements/Headers.cjs"));
exports.HeadersElement = _Headers.default;
var _Example = _interopRequireDefault(require("../elements/Example.cjs"));
exports.ExampleElement = _Example.default;
var _Header = _interopRequireDefault(require("../elements/Header.cjs"));
exports.HeaderElement = _Header.default;
var _Tag = _interopRequireDefault(require("../elements/Tag.cjs"));
exports.TagElement = _Tag.default;
var _Reference = _interopRequireDefault(require("../elements/Reference.cjs"));
exports.ReferenceElement = _Reference.default;
var _Schema = _interopRequireDefault(require("../elements/Schema.cjs"));
exports.SchemaElement = _Schema.default;
var _Xml = _interopRequireDefault(require("../elements/Xml.cjs"));
exports.XmlElement = _Xml.default;
var _Definitions = _interopRequireDefault(require("../elements/Definitions.cjs"));
exports.DefinitionsElement = _Definitions.default;
var _ParametersDefinitions = _interopRequireDefault(require("../elements/ParametersDefinitions.cjs"));
exports.ParametersDefinitionsElement = _ParametersDefinitions.default;
var _ResponsesDefinitions = _interopRequireDefault(require("../elements/ResponsesDefinitions.cjs"));
exports.ResponsesDefinitionsElement = _ResponsesDefinitions.default;
var _SecurityDefinitions = _interopRequireDefault(require("../elements/SecurityDefinitions.cjs"));
exports.SecurityDefinitionsElement = _SecurityDefinitions.default;
var _SecurityScheme = _interopRequireDefault(require("../elements/SecurityScheme.cjs"));
exports.SecuritySchemeElement = _SecurityScheme.default;
var _Scopes = _interopRequireDefault(require("../elements/Scopes.cjs"));
exports.ScopesElement = _Scopes.default;
var _SecurityRequirement = _interopRequireDefault(require("../elements/SecurityRequirement.cjs"));
exports.SecurityRequirementElement = _SecurityRequirement.default;
var _index = require("./index.cjs");
// register refractors specific to element types
_Swagger.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Swagger', '$visitor']);
_SwaggerVersion.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'SwaggerVersion', '$visitor']);
_Info.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Info', '$visitor']);
_Contact.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Contact', '$visitor']);
_License.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'License', '$visitor']);
_Paths.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Paths', '$visitor']);
_PathItem.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'PathItem', '$visitor']);
_Operation.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Operation', '$visitor']);
_ExternalDocumentation.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'ExternalDocumentation', '$visitor']);
_Parameter.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Parameter', '$visitor']);
_Items.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Items', '$visitor']);
_Responses.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Responses', '$visitor']);
_Response.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Response', '$visitor']);
_Headers.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Headers', '$visitor']);
_Example.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Example', '$visitor']);
_Header.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Header', '$visitor']);
_Tag.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Tag', '$visitor']);
_Reference.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Reference', '$visitor']);
_Schema.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Schema', '$visitor']);
_Xml.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'XML', '$visitor']);
_Definitions.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Definitions', '$visitor']);
_ParametersDefinitions.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'ParametersDefinitions', '$visitor']);
_ResponsesDefinitions.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'ResponsesDefinitions', '$visitor']);
_SecurityDefinitions.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'SecurityDefinitions', '$visitor']);
_SecurityScheme.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'SecurityScheme', '$visitor']);
_Scopes.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Scopes', '$visitor']);
_SecurityRequirement.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'SecurityRequirement', '$visitor']);
