// @backstage/plugin-catalog/dist/alpha/entityContents.esm.js
import React from 'react';
import { createExtensionInput, coreExtensionData } from '@backstage/frontend-plugin-api';
import { EntityContentBlueprint } from '@backstage/plugin-catalog-react/alpha';

const catalogOverviewEntityContent = EntityContentBlueprint.makeWithOverrides({
  name: "overview",
  inputs: {
    cards: createExtensionInput([
      coreExtensionData.reactElement,
      EntityContentBlueprint.dataRefs.filterFunction.optional(),
      EntityContentBlueprint.dataRefs.filterExpression.optional()
    ])
  },
  factory: (originalFactory, { inputs }) => {
    return originalFactory({
      defaultPath: "/",
      defaultTitle: "Overview",
      loader: async () => import('./EntityOverviewPage.esm.js').then((m) => /* @__PURE__ */ React.createElement(
        m.EntityOverviewPage,
        {
          cards: inputs.cards.map((c) => ({
            element: c.get(coreExtensionData.reactElement),
            filterFunction: c.get(
              EntityContentBlueprint.dataRefs.filterFunction
            ),
            filterExpression: c.get(
              EntityContentBlueprint.dataRefs.filterExpression
            )
          }))
        }
      ))
    });
  }
});
var entityContents = [catalogOverviewEntityContent];

export { catalogOverviewEntityContent, entityContents as default };
