// @backstage/plugin-catalog-react/dist/alpha/blueprints/EntityContentBlueprint.esm.js
import { createExtensionBlueprint, coreExtensionData, ExtensionBoundary } from '@backstage/frontend-plugin-api';
import { entityContentTitleDataRef, entityFilterFunctionDataRef, entityFilterExpressionDataRef } from './extensionData.esm.js';

const EntityContentBlueprint = createExtensionBlueprint({
  kind: "entity-content",
  attachTo: { id: "page:catalog/entity", input: "contents" },
  output: [
    coreExtensionData.reactElement,
    coreExtensionData.routePath,
    entityContentTitleDataRef,
    coreExtensionData.routeRef.optional(),
    entityFilterFunctionDataRef.optional(),
    entityFilterExpressionDataRef.optional()
  ],
  dataRefs: {
    title: entityContentTitleDataRef,
    filterFunction: entityFilterFunctionDataRef,
    filterExpression: entityFilterExpressionDataRef
  },
  config: {
    schema: {
      path: (z) => z.string().optional(),
      title: (z) => z.string().optional(),
      filter: (z) => z.string().optional()
    }
  },
  *factory({
    loader,
    defaultPath,
    defaultTitle,
    filter,
    routeRef
  }, { node, config }) {
    const path = config.path ?? defaultPath;
    const title = config.title ?? defaultTitle;
    yield coreExtensionData.reactElement(ExtensionBoundary.lazy(node, loader));
    yield coreExtensionData.routePath(path);
    yield entityContentTitleDataRef(title);
    if (routeRef) {
      yield coreExtensionData.routeRef(routeRef);
    }
    if (config.filter) {
      yield entityFilterExpressionDataRef(config.filter);
    } else if (typeof filter === "string") {
      yield entityFilterExpressionDataRef(filter);
    } else if (typeof filter === "function") {
      yield entityFilterFunctionDataRef(filter);
    }
  }
});

export { EntityContentBlueprint };
