// @backstage/plugin-catalog-react/dist/components/EntityNamespacePicker/EntityNamespacePicker.esm.js
import { makeStyles } from '@material-ui/core/styles';
import React from 'react';
import { EntityNamespaceFilter } from '../../filters.esm.js';
import { EntityAutocompletePicker } from '../EntityAutocompletePicker/EntityAutocompletePicker.esm.js';
import { catalogReactTranslationRef } from '../../translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

const useStyles = makeStyles(
  {
    input: {}
  },
  {
    name: "CatalogReactEntityNamespacePicker"
  }
);
const EntityNamespacePicker = (props) => {
  const { initiallySelectedNamespaces } = props;
  const classes = useStyles();
  const { t } = useTranslationRef(catalogReactTranslationRef);
  return /* @__PURE__ */ React.createElement(
    EntityAutocompletePicker,
    {
      label: t("entityNamespacePicker.title"),
      name: "namespace",
      path: "metadata.namespace",
      Filter: EntityNamespaceFilter,
      InputProps: { className: classes.input },
      initialSelectedOptions: initiallySelectedNamespaces
    }
  );
};

export { EntityNamespacePicker };
