// @backstage/plugin-search-react/dist/alpha/blueprints/SearchResultListItemBlueprint.esm.js
import React, { lazy } from 'react';
import { createExtensionBlueprint, ExtensionBoundary } from '@backstage/frontend-plugin-api';
import { searchResultListItemDataRef } from './types.esm.js';
import { SearchResultListItemExtension } from '../../extensions.esm.js';

const SearchResultListItemBlueprint = createExtensionBlueprint({
  kind: "search-result-list-item",
  attachTo: {
    id: "page:search",
    input: "items"
  },
  config: {
    schema: {
      noTrack: (z) => z.boolean().default(false)
    }
  },
  output: [searchResultListItemDataRef],
  dataRefs: {
    item: searchResultListItemDataRef
  },
  *factory(params, { config, node }) {
    const ExtensionComponent = lazy(
      () => params.component({ config }).then((component) => ({ default: component }))
    );
    yield searchResultListItemDataRef({
      predicate: params.predicate,
      component: (props) => /* @__PURE__ */ React.createElement(ExtensionBoundary, { node }, /* @__PURE__ */ React.createElement(
        SearchResultListItemExtension,
        {
          rank: props.rank,
          result: props.result,
          noTrack: config.noTrack
        },
        /* @__PURE__ */ React.createElement(ExtensionComponent, { ...props })
      ))
    });
  }
});

export { SearchResultListItemBlueprint };
