// @backstage/core-components/dist/components/LogViewer/LogViewerControls.esm.js
import React__default from 'react';
import IconButton from '@material-ui/core/IconButton';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import ChevronRight from '@material-ui/icons/ChevronRight';
import FilterList from '@material-ui/icons/FilterList';

function LogViewerControls(props) {
  const { resultCount, resultIndexStep, toggleShouldFilter } = props;
  const resultIndex = props.resultIndex ?? 0;
  const handleKeyPress = (event) => {
    if (event.key === "Enter") {
      if (event.metaKey || event.ctrlKey || event.altKey) {
        toggleShouldFilter();
      } else {
        resultIndexStep(event.shiftKey);
      }
    }
  };
  return /* @__PURE__ */ React__default.createElement(React__default.Fragment, null, resultCount !== void 0 && /* @__PURE__ */ React__default.createElement(React__default.Fragment, null, /* @__PURE__ */ React__default.createElement(IconButton, { size: "small", onClick: () => resultIndexStep(true) }, /* @__PURE__ */ React__default.createElement(ChevronLeftIcon, null)), /* @__PURE__ */ React__default.createElement(Typography, null, Math.min(resultIndex + 1, resultCount), "/", resultCount), /* @__PURE__ */ React__default.createElement(IconButton, { size: "small", onClick: () => resultIndexStep() }, /* @__PURE__ */ React__default.createElement(ChevronRight, null))), /* @__PURE__ */ React__default.createElement(
    TextField,
    {
      size: "small",
      variant: "standard",
      placeholder: "Search",
      value: props.searchInput,
      onKeyPress: handleKeyPress,
      onChange: (e) => props.setSearchInput(e.target.value)
    }
  ), /* @__PURE__ */ React__default.createElement(IconButton, { size: "small", onClick: toggleShouldFilter }, props.shouldFilter ? /* @__PURE__ */ React__default.createElement(FilterList, { color: "primary" }) : /* @__PURE__ */ React__default.createElement(FilterList, { color: "disabled" })));
}

export { LogViewerControls };
