// @backstage/plugin-scaffolder/dist/components/OngoingTask/ContextMenu.esm.js
import IconButton from '@material-ui/core/IconButton';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import MenuItem from '@material-ui/core/MenuItem';
import MenuList from '@material-ui/core/MenuList';
import Popover from '@material-ui/core/Popover';
import { makeStyles, useTheme } from '@material-ui/core/styles';
import { useAsync } from '@react-hookz/web';
import Cancel from '@material-ui/icons/Cancel';
import Repeat from '@material-ui/icons/Repeat';
import Replay from '@material-ui/icons/Replay';
import Toc from '@material-ui/icons/Toc';
import ControlPointIcon from '@material-ui/icons/ControlPoint';
import MoreVert from '@material-ui/icons/MoreVert';
import React, { useState } from 'react';
import { useApi, useAnalytics } from '@backstage/core-plugin-api';
import { scaffolderApiRef } from '@backstage/plugin-scaffolder-react';
import { usePermission } from '@backstage/plugin-permission-react';
import { taskCancelPermission, taskReadPermission, taskCreatePermission } from '@backstage/plugin-scaffolder-common/alpha';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../translation.esm.js';

const useStyles = makeStyles(() => ({
  button: {
    color: ({ fontColor }) => fontColor
  }
}));
const ContextMenu = (props) => {
  const {
    cancelEnabled,
    canRetry,
    isRetryableTask,
    logsVisible,
    buttonBarVisible,
    onRetry,
    onStartOver,
    onToggleLogs,
    onToggleButtonBar,
    taskId
  } = props;
  const { getPageTheme } = useTheme();
  const pageTheme = getPageTheme({ themeId: "website" });
  const classes = useStyles({ fontColor: pageTheme.fontColor });
  const scaffolderApi = useApi(scaffolderApiRef);
  const analytics = useAnalytics();
  const [anchorEl, setAnchorEl] = useState();
  const { t } = useTranslationRef(scaffolderTranslationRef);
  const [{ status: cancelStatus }, { execute: cancel }] = useAsync(async () => {
    if (taskId) {
      analytics.captureEvent("cancelled", "Template has been cancelled");
      await scaffolderApi.cancelTask(taskId);
    }
  });
  const { allowed: canCancelTask } = usePermission({
    permission: taskCancelPermission
  });
  const { allowed: canReadTask } = usePermission({
    permission: taskReadPermission
  });
  const { allowed: canCreateTask } = usePermission({
    permission: taskCreatePermission
  });
  const canStartOver = canReadTask && canCreateTask;
  return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(
    IconButton,
    {
      "aria-label": "more",
      "aria-controls": "long-menu",
      "aria-haspopup": "true",
      onClick: (event) => {
        setAnchorEl(event.currentTarget);
      },
      "data-testid": "menu-button",
      className: classes.button
    },
    /* @__PURE__ */ React.createElement(MoreVert, null)
  ), /* @__PURE__ */ React.createElement(
    Popover,
    {
      open: Boolean(anchorEl),
      onClose: () => setAnchorEl(void 0),
      anchorEl,
      anchorOrigin: { vertical: "bottom", horizontal: "right" },
      transformOrigin: { vertical: "top", horizontal: "right" }
    },
    /* @__PURE__ */ React.createElement(MenuList, null, /* @__PURE__ */ React.createElement(MenuItem, { onClick: () => onToggleLogs?.(!logsVisible) }, /* @__PURE__ */ React.createElement(ListItemIcon, null, /* @__PURE__ */ React.createElement(Toc, { fontSize: "small" })), /* @__PURE__ */ React.createElement(
      ListItemText,
      {
        primary: logsVisible ? t("ongoingTask.contextMenu.hideLogs") : t("ongoingTask.contextMenu.showLogs")
      }
    )), /* @__PURE__ */ React.createElement(MenuItem, { onClick: () => onToggleButtonBar?.(!buttonBarVisible) }, /* @__PURE__ */ React.createElement(ListItemIcon, null, /* @__PURE__ */ React.createElement(ControlPointIcon, { fontSize: "small" })), /* @__PURE__ */ React.createElement(
      ListItemText,
      {
        primary: buttonBarVisible ? t("ongoingTask.contextMenu.hideButtonBar") : t("ongoingTask.contextMenu.showButtonBar")
      }
    )), /* @__PURE__ */ React.createElement(
      MenuItem,
      {
        onClick: onStartOver,
        disabled: cancelEnabled || !canStartOver,
        "data-testid": "start-over-task"
      },
      /* @__PURE__ */ React.createElement(ListItemIcon, null, /* @__PURE__ */ React.createElement(Repeat, { fontSize: "small" })),
      /* @__PURE__ */ React.createElement(ListItemText, { primary: t("ongoingTask.contextMenu.startOver") })
    ), isRetryableTask && /* @__PURE__ */ React.createElement(
      MenuItem,
      {
        onClick: onRetry,
        disabled: cancelEnabled || !canRetry,
        "data-testid": "retry-task"
      },
      /* @__PURE__ */ React.createElement(ListItemIcon, null, /* @__PURE__ */ React.createElement(Replay, { fontSize: "small" })),
      /* @__PURE__ */ React.createElement(ListItemText, { primary: t("ongoingTask.contextMenu.retry") })
    ), /* @__PURE__ */ React.createElement(
      MenuItem,
      {
        onClick: cancel,
        disabled: !cancelEnabled || cancelStatus !== "not-executed" || !canCancelTask,
        "data-testid": "cancel-task"
      },
      /* @__PURE__ */ React.createElement(ListItemIcon, null, /* @__PURE__ */ React.createElement(Cancel, { fontSize: "small" })),
      /* @__PURE__ */ React.createElement(ListItemText, { primary: t("ongoingTask.contextMenu.cancel") })
    ))
  ));
};

export { ContextMenu };
