// @backstage/plugin-app/dist/extensions/AppRoutes.esm.js
import React from 'react';
import { createExtension, createExtensionInput, coreExtensionData, useComponentRef, coreComponentRefs } from '@backstage/frontend-plugin-api';
import { useRoutes } from 'react-router-dom';

const AppRoutes = createExtension({
  name: "routes",
  attachTo: { id: "app/layout", input: "content" },
  inputs: {
    routes: createExtensionInput([
      coreExtensionData.routePath,
      coreExtensionData.routeRef.optional(),
      coreExtensionData.reactElement
    ])
  },
  output: [coreExtensionData.reactElement],
  factory({ inputs }) {
    const Routes = () => {
      const NotFoundErrorPage = useComponentRef(
        coreComponentRefs.notFoundErrorPage
      );
      const element = useRoutes([
        ...inputs.routes.map((route) => ({
          path: `${route.get(coreExtensionData.routePath)}/*`,
          element: route.get(coreExtensionData.reactElement)
        })),
        {
          path: "*",
          element: /* @__PURE__ */ React.createElement(NotFoundErrorPage, null)
        }
      ]);
      return element;
    };
    return [coreExtensionData.reactElement(/* @__PURE__ */ React.createElement(Routes, null))];
  }
});

export { AppRoutes };
