// @backstage/plugin-app/dist/extensions/components.esm.js
import React from 'react';
import Button from '@material-ui/core/Button';
import { createComponentExtension, coreComponentRefs } from '@backstage/frontend-plugin-api';
import { ErrorPanel } from '@backstage/core-components';
import '../packages/app-defaults/src/defaults/apis.esm.js';
import { components } from '../packages/app-defaults/src/defaults/components.esm.js';
import '@material-ui/icons/Apartment';
import '@material-ui/icons/BrokenImage';
import '@material-ui/icons/Category';
import '@material-ui/icons/CreateNewFolder';
import '@material-ui/icons/Subject';
import '@material-ui/icons/Search';
import '@material-ui/icons/Chat';
import '@material-ui/icons/Dashboard';
import '@material-ui/icons/Description';
import '@material-ui/icons/Email';
import '@material-ui/icons/Extension';
import '@material-ui/icons/GitHub';
import '@material-ui/icons/Help';
import '@material-ui/icons/LocationOn';
import '@material-ui/icons/Memory';
import '@material-ui/icons/MenuBook';
import '@material-ui/icons/People';
import '@material-ui/icons/Person';
import '@material-ui/icons/Warning';
import '@material-ui/icons/Storage';
import '@material-ui/icons/FeaturedPlayList';
import '@material-ui/icons/Star';
import '@material-ui/icons/StarBorder';
import '@material-ui/icons/OpenInNew';
import '@backstage/theme';
import '@material-ui/icons/Brightness2';
import '@material-ui/icons/WbSunny';

const DefaultProgressComponent = createComponentExtension({
  ref: coreComponentRefs.progress,
  loader: { sync: () => components.Progress }
});
const DefaultNotFoundErrorPageComponent = createComponentExtension({
  ref: coreComponentRefs.notFoundErrorPage,
  loader: { sync: () => components.NotFoundErrorPage }
});
const DefaultErrorBoundaryComponent = createComponentExtension({
  ref: coreComponentRefs.errorBoundaryFallback,
  loader: {
    sync: () => (props) => {
      const { plugin, error, resetError } = props;
      const title = `Error in ${plugin?.id}`;
      return /* @__PURE__ */ React.createElement(ErrorPanel, { title, error, defaultExpanded: true }, /* @__PURE__ */ React.createElement(Button, { variant: "outlined", onClick: resetError }, "Retry"));
    }
  }
});

export { DefaultErrorBoundaryComponent, DefaultNotFoundErrorPageComponent, DefaultProgressComponent };
