// @backstage/plugin-app/dist/extensions/elements.esm.js
import { OAuthRequestDialog, AlertDisplay } from '@backstage/core-components';
import { AppRootElementBlueprint } from '@backstage/frontend-plugin-api';
import React from 'react';

const oauthRequestDialogAppRootElement = AppRootElementBlueprint.make({
  name: "oauth-request-dialog",
  params: {
    element: /* @__PURE__ */ React.createElement(OAuthRequestDialog, null)
  }
});
const alertDisplayAppRootElement = AppRootElementBlueprint.makeWithOverrides({
  name: "alert-display",
  config: {
    schema: {
      transientTimeoutMs: (z) => z.number().default(5e3),
      anchorOrigin: (z) => z.object({
        vertical: z.enum(["top", "bottom"]).default("top"),
        horizontal: z.enum(["left", "center", "right"]).default("center")
      }).default({})
    }
  },
  factory: (originalFactory, { config }) => {
    return originalFactory({
      element: () => /* @__PURE__ */ React.createElement(AlertDisplay, { ...config })
    });
  }
});

export { alertDisplayAppRootElement, oauthRequestDialogAppRootElement };
