// @backstage/core-components/dist/layout/SignInPage/LegacyUserIdentity.esm.js
function parseJwtPayload(token) {
  const [_header, payload, _signature] = token.split(".");
  return JSON.parse(window.atob(payload));
}
class LegacyUserIdentity {
  constructor(result) {
    this.result = result;
  }
  getUserId() {
    return this.result.userId;
  }
  static fromResult(result) {
    return new LegacyUserIdentity(result);
  }
  async getIdToken() {
    return this.result.getIdToken?.();
  }
  getProfile() {
    return this.result.profile;
  }
  async getProfileInfo() {
    return this.result.profile;
  }
  async getBackstageIdentity() {
    const token = await this.getIdToken();
    if (!token) {
      const userEntityRef = `user:default/${this.getUserId()}`;
      return {
        type: "user",
        userEntityRef,
        ownershipEntityRefs: [userEntityRef]
      };
    }
    const { sub, ent } = parseJwtPayload(token);
    return {
      type: "user",
      userEntityRef: sub,
      ownershipEntityRefs: ent ?? []
    };
  }
  async getCredentials() {
    const token = await this.result.getIdToken?.();
    return { token };
  }
  async signOut() {
    return this.result.signOut?.();
  }
}

export { LegacyUserIdentity };
