// @aws-sdk/client-codecommit/dist-es/models/models_0.js
import { CodeCommitServiceException as __BaseException } from "./CodeCommitServiceException";
export class ActorDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "ActorDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "ActorDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ActorDoesNotExistException.prototype);
    }
}
export const ApprovalState = {
    APPROVE: "APPROVE",
    REVOKE: "REVOKE",
};
export class ApprovalRuleContentRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ApprovalRuleContentRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ApprovalRuleContentRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ApprovalRuleContentRequiredException.prototype);
    }
}
export class ApprovalRuleDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "ApprovalRuleDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "ApprovalRuleDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ApprovalRuleDoesNotExistException.prototype);
    }
}
export class ApprovalRuleNameAlreadyExistsException extends __BaseException {
    constructor(opts) {
        super({
            name: "ApprovalRuleNameAlreadyExistsException",
            $fault: "client",
            ...opts,
        });
        this.name = "ApprovalRuleNameAlreadyExistsException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ApprovalRuleNameAlreadyExistsException.prototype);
    }
}
export class ApprovalRuleNameRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ApprovalRuleNameRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ApprovalRuleNameRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ApprovalRuleNameRequiredException.prototype);
    }
}
export const OverrideStatus = {
    OVERRIDE: "OVERRIDE",
    REVOKE: "REVOKE",
};
export class ApprovalRuleTemplateContentRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ApprovalRuleTemplateContentRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ApprovalRuleTemplateContentRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ApprovalRuleTemplateContentRequiredException.prototype);
    }
}
export class ApprovalRuleTemplateDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "ApprovalRuleTemplateDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "ApprovalRuleTemplateDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ApprovalRuleTemplateDoesNotExistException.prototype);
    }
}
export class ApprovalRuleTemplateInUseException extends __BaseException {
    constructor(opts) {
        super({
            name: "ApprovalRuleTemplateInUseException",
            $fault: "client",
            ...opts,
        });
        this.name = "ApprovalRuleTemplateInUseException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ApprovalRuleTemplateInUseException.prototype);
    }
}
export class ApprovalRuleTemplateNameAlreadyExistsException extends __BaseException {
    constructor(opts) {
        super({
            name: "ApprovalRuleTemplateNameAlreadyExistsException",
            $fault: "client",
            ...opts,
        });
        this.name = "ApprovalRuleTemplateNameAlreadyExistsException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ApprovalRuleTemplateNameAlreadyExistsException.prototype);
    }
}
export class ApprovalRuleTemplateNameRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ApprovalRuleTemplateNameRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ApprovalRuleTemplateNameRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ApprovalRuleTemplateNameRequiredException.prototype);
    }
}
export class ApprovalStateRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ApprovalStateRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ApprovalStateRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ApprovalStateRequiredException.prototype);
    }
}
export class EncryptionIntegrityChecksFailedException extends __BaseException {
    constructor(opts) {
        super({
            name: "EncryptionIntegrityChecksFailedException",
            $fault: "server",
            ...opts,
        });
        this.name = "EncryptionIntegrityChecksFailedException";
        this.$fault = "server";
        Object.setPrototypeOf(this, EncryptionIntegrityChecksFailedException.prototype);
    }
}
export class EncryptionKeyAccessDeniedException extends __BaseException {
    constructor(opts) {
        super({
            name: "EncryptionKeyAccessDeniedException",
            $fault: "client",
            ...opts,
        });
        this.name = "EncryptionKeyAccessDeniedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, EncryptionKeyAccessDeniedException.prototype);
    }
}
export class EncryptionKeyDisabledException extends __BaseException {
    constructor(opts) {
        super({
            name: "EncryptionKeyDisabledException",
            $fault: "client",
            ...opts,
        });
        this.name = "EncryptionKeyDisabledException";
        this.$fault = "client";
        Object.setPrototypeOf(this, EncryptionKeyDisabledException.prototype);
    }
}
export class EncryptionKeyNotFoundException extends __BaseException {
    constructor(opts) {
        super({
            name: "EncryptionKeyNotFoundException",
            $fault: "client",
            ...opts,
        });
        this.name = "EncryptionKeyNotFoundException";
        this.$fault = "client";
        Object.setPrototypeOf(this, EncryptionKeyNotFoundException.prototype);
    }
}
export class EncryptionKeyUnavailableException extends __BaseException {
    constructor(opts) {
        super({
            name: "EncryptionKeyUnavailableException",
            $fault: "client",
            ...opts,
        });
        this.name = "EncryptionKeyUnavailableException";
        this.$fault = "client";
        Object.setPrototypeOf(this, EncryptionKeyUnavailableException.prototype);
    }
}
export class InvalidApprovalRuleTemplateNameException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidApprovalRuleTemplateNameException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidApprovalRuleTemplateNameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidApprovalRuleTemplateNameException.prototype);
    }
}
export class InvalidRepositoryNameException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidRepositoryNameException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRepositoryNameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRepositoryNameException.prototype);
    }
}
export class MaximumRuleTemplatesAssociatedWithRepositoryException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaximumRuleTemplatesAssociatedWithRepositoryException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaximumRuleTemplatesAssociatedWithRepositoryException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaximumRuleTemplatesAssociatedWithRepositoryException.prototype);
    }
}
export class RepositoryDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "RepositoryDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "RepositoryDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RepositoryDoesNotExistException.prototype);
    }
}
export class RepositoryNameRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "RepositoryNameRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "RepositoryNameRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RepositoryNameRequiredException.prototype);
    }
}
export class AuthorDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "AuthorDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "AuthorDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, AuthorDoesNotExistException.prototype);
    }
}
export class MaximumRepositoryNamesExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaximumRepositoryNamesExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaximumRepositoryNamesExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaximumRepositoryNamesExceededException.prototype);
    }
}
export class RepositoryNamesRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "RepositoryNamesRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "RepositoryNamesRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RepositoryNamesRequiredException.prototype);
    }
}
export const ConflictDetailLevelTypeEnum = {
    FILE_LEVEL: "FILE_LEVEL",
    LINE_LEVEL: "LINE_LEVEL",
};
export const ConflictResolutionStrategyTypeEnum = {
    ACCEPT_DESTINATION: "ACCEPT_DESTINATION",
    ACCEPT_SOURCE: "ACCEPT_SOURCE",
    AUTOMERGE: "AUTOMERGE",
    NONE: "NONE",
};
export const MergeOptionTypeEnum = {
    FAST_FORWARD_MERGE: "FAST_FORWARD_MERGE",
    SQUASH_MERGE: "SQUASH_MERGE",
    THREE_WAY_MERGE: "THREE_WAY_MERGE",
};
export const FileModeTypeEnum = {
    EXECUTABLE: "EXECUTABLE",
    NORMAL: "NORMAL",
    SYMLINK: "SYMLINK",
};
export const ChangeTypeEnum = {
    ADDED: "A",
    DELETED: "D",
    MODIFIED: "M",
};
export const ObjectTypeEnum = {
    DIRECTORY: "DIRECTORY",
    FILE: "FILE",
    GIT_LINK: "GIT_LINK",
    SYMBOLIC_LINK: "SYMBOLIC_LINK",
};
export class CommitDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "CommitDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "CommitDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CommitDoesNotExistException.prototype);
    }
}
export class CommitRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "CommitRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "CommitRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CommitRequiredException.prototype);
    }
}
export class InvalidCommitException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidCommitException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidCommitException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidCommitException.prototype);
    }
}
export class InvalidConflictDetailLevelException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidConflictDetailLevelException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidConflictDetailLevelException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidConflictDetailLevelException.prototype);
    }
}
export class InvalidConflictResolutionStrategyException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidConflictResolutionStrategyException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidConflictResolutionStrategyException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidConflictResolutionStrategyException.prototype);
    }
}
export class InvalidContinuationTokenException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidContinuationTokenException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidContinuationTokenException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidContinuationTokenException.prototype);
    }
}
export class InvalidMaxConflictFilesException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidMaxConflictFilesException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidMaxConflictFilesException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidMaxConflictFilesException.prototype);
    }
}
export class InvalidMaxMergeHunksException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidMaxMergeHunksException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidMaxMergeHunksException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidMaxMergeHunksException.prototype);
    }
}
export class InvalidMergeOptionException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidMergeOptionException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidMergeOptionException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidMergeOptionException.prototype);
    }
}
export class MaximumFileContentToLoadExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaximumFileContentToLoadExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaximumFileContentToLoadExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaximumFileContentToLoadExceededException.prototype);
    }
}
export class MaximumItemsToCompareExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaximumItemsToCompareExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaximumItemsToCompareExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaximumItemsToCompareExceededException.prototype);
    }
}
export class MergeOptionRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "MergeOptionRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "MergeOptionRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MergeOptionRequiredException.prototype);
    }
}
export class TipsDivergenceExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "TipsDivergenceExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "TipsDivergenceExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, TipsDivergenceExceededException.prototype);
    }
}
export class CommitIdsLimitExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "CommitIdsLimitExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "CommitIdsLimitExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CommitIdsLimitExceededException.prototype);
    }
}
export class CommitIdsListRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "CommitIdsListRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "CommitIdsListRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CommitIdsListRequiredException.prototype);
    }
}
export const BatchGetRepositoriesErrorCodeEnum = {
    ENCRYPTION_INTEGRITY_CHECKS_FAILED_EXCEPTION: "EncryptionIntegrityChecksFailedException",
    ENCRYPTION_KEY_ACCESS_DENIED_EXCEPTION: "EncryptionKeyAccessDeniedException",
    ENCRYPTION_KEY_DISABLED_EXCEPTION: "EncryptionKeyDisabledException",
    ENCRYPTION_KEY_NOT_FOUND_EXCEPTION: "EncryptionKeyNotFoundException",
    ENCRYPTION_KEY_UNAVAILABLE_EXCEPTION: "EncryptionKeyUnavailableException",
    REPOSITORY_DOES_NOT_EXIST_EXCEPTION: "RepositoryDoesNotExistException",
};
export class BeforeCommitIdAndAfterCommitIdAreSameException extends __BaseException {
    constructor(opts) {
        super({
            name: "BeforeCommitIdAndAfterCommitIdAreSameException",
            $fault: "client",
            ...opts,
        });
        this.name = "BeforeCommitIdAndAfterCommitIdAreSameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, BeforeCommitIdAndAfterCommitIdAreSameException.prototype);
    }
}
export class BlobIdDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "BlobIdDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "BlobIdDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, BlobIdDoesNotExistException.prototype);
    }
}
export class BlobIdRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "BlobIdRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "BlobIdRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, BlobIdRequiredException.prototype);
    }
}
export class BranchDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "BranchDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "BranchDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, BranchDoesNotExistException.prototype);
    }
}
export class BranchNameExistsException extends __BaseException {
    constructor(opts) {
        super({
            name: "BranchNameExistsException",
            $fault: "client",
            ...opts,
        });
        this.name = "BranchNameExistsException";
        this.$fault = "client";
        Object.setPrototypeOf(this, BranchNameExistsException.prototype);
    }
}
export class BranchNameIsTagNameException extends __BaseException {
    constructor(opts) {
        super({
            name: "BranchNameIsTagNameException",
            $fault: "client",
            ...opts,
        });
        this.name = "BranchNameIsTagNameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, BranchNameIsTagNameException.prototype);
    }
}
export class BranchNameRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "BranchNameRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "BranchNameRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, BranchNameRequiredException.prototype);
    }
}
export class CannotDeleteApprovalRuleFromTemplateException extends __BaseException {
    constructor(opts) {
        super({
            name: "CannotDeleteApprovalRuleFromTemplateException",
            $fault: "client",
            ...opts,
        });
        this.name = "CannotDeleteApprovalRuleFromTemplateException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CannotDeleteApprovalRuleFromTemplateException.prototype);
    }
}
export class CannotModifyApprovalRuleFromTemplateException extends __BaseException {
    constructor(opts) {
        super({
            name: "CannotModifyApprovalRuleFromTemplateException",
            $fault: "client",
            ...opts,
        });
        this.name = "CannotModifyApprovalRuleFromTemplateException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CannotModifyApprovalRuleFromTemplateException.prototype);
    }
}
export class ClientRequestTokenRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ClientRequestTokenRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ClientRequestTokenRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ClientRequestTokenRequiredException.prototype);
    }
}
export class InvalidApprovalRuleTemplateContentException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidApprovalRuleTemplateContentException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidApprovalRuleTemplateContentException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidApprovalRuleTemplateContentException.prototype);
    }
}
export class InvalidApprovalRuleTemplateDescriptionException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidApprovalRuleTemplateDescriptionException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidApprovalRuleTemplateDescriptionException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidApprovalRuleTemplateDescriptionException.prototype);
    }
}
export class NumberOfRuleTemplatesExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "NumberOfRuleTemplatesExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "NumberOfRuleTemplatesExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NumberOfRuleTemplatesExceededException.prototype);
    }
}
export class CommitIdRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "CommitIdRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "CommitIdRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CommitIdRequiredException.prototype);
    }
}
export class InvalidBranchNameException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidBranchNameException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidBranchNameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidBranchNameException.prototype);
    }
}
export class InvalidCommitIdException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidCommitIdException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidCommitIdException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidCommitIdException.prototype);
    }
}
export class CommitMessageLengthExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "CommitMessageLengthExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "CommitMessageLengthExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CommitMessageLengthExceededException.prototype);
    }
}
export class DirectoryNameConflictsWithFileNameException extends __BaseException {
    constructor(opts) {
        super({
            name: "DirectoryNameConflictsWithFileNameException",
            $fault: "client",
            ...opts,
        });
        this.name = "DirectoryNameConflictsWithFileNameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, DirectoryNameConflictsWithFileNameException.prototype);
    }
}
export class FileContentAndSourceFileSpecifiedException extends __BaseException {
    constructor(opts) {
        super({
            name: "FileContentAndSourceFileSpecifiedException",
            $fault: "client",
            ...opts,
        });
        this.name = "FileContentAndSourceFileSpecifiedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, FileContentAndSourceFileSpecifiedException.prototype);
    }
}
export class FileContentSizeLimitExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "FileContentSizeLimitExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "FileContentSizeLimitExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, FileContentSizeLimitExceededException.prototype);
    }
}
export class FileDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "FileDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "FileDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, FileDoesNotExistException.prototype);
    }
}
export class FileEntryRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "FileEntryRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "FileEntryRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, FileEntryRequiredException.prototype);
    }
}
export class FileModeRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "FileModeRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "FileModeRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, FileModeRequiredException.prototype);
    }
}
export class FileNameConflictsWithDirectoryNameException extends __BaseException {
    constructor(opts) {
        super({
            name: "FileNameConflictsWithDirectoryNameException",
            $fault: "client",
            ...opts,
        });
        this.name = "FileNameConflictsWithDirectoryNameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, FileNameConflictsWithDirectoryNameException.prototype);
    }
}
export class FilePathConflictsWithSubmodulePathException extends __BaseException {
    constructor(opts) {
        super({
            name: "FilePathConflictsWithSubmodulePathException",
            $fault: "client",
            ...opts,
        });
        this.name = "FilePathConflictsWithSubmodulePathException";
        this.$fault = "client";
        Object.setPrototypeOf(this, FilePathConflictsWithSubmodulePathException.prototype);
    }
}
export class FolderContentSizeLimitExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "FolderContentSizeLimitExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "FolderContentSizeLimitExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, FolderContentSizeLimitExceededException.prototype);
    }
}
export class InvalidDeletionParameterException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidDeletionParameterException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidDeletionParameterException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidDeletionParameterException.prototype);
    }
}
export class InvalidEmailException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidEmailException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidEmailException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidEmailException.prototype);
    }
}
export class InvalidFileModeException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidFileModeException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidFileModeException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidFileModeException.prototype);
    }
}
export class InvalidParentCommitIdException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidParentCommitIdException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidParentCommitIdException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidParentCommitIdException.prototype);
    }
}
export class InvalidPathException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidPathException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidPathException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidPathException.prototype);
    }
}
export class MaximumFileEntriesExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaximumFileEntriesExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaximumFileEntriesExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaximumFileEntriesExceededException.prototype);
    }
}
export class NameLengthExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "NameLengthExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "NameLengthExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NameLengthExceededException.prototype);
    }
}
export class NoChangeException extends __BaseException {
    constructor(opts) {
        super({
            name: "NoChangeException",
            $fault: "client",
            ...opts,
        });
        this.name = "NoChangeException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NoChangeException.prototype);
    }
}
export class ParentCommitDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "ParentCommitDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "ParentCommitDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ParentCommitDoesNotExistException.prototype);
    }
}
export class ParentCommitIdOutdatedException extends __BaseException {
    constructor(opts) {
        super({
            name: "ParentCommitIdOutdatedException",
            $fault: "client",
            ...opts,
        });
        this.name = "ParentCommitIdOutdatedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ParentCommitIdOutdatedException.prototype);
    }
}
export class ParentCommitIdRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ParentCommitIdRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ParentCommitIdRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ParentCommitIdRequiredException.prototype);
    }
}
export class PathRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "PathRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "PathRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, PathRequiredException.prototype);
    }
}
export class PutFileEntryConflictException extends __BaseException {
    constructor(opts) {
        super({
            name: "PutFileEntryConflictException",
            $fault: "client",
            ...opts,
        });
        this.name = "PutFileEntryConflictException";
        this.$fault = "client";
        Object.setPrototypeOf(this, PutFileEntryConflictException.prototype);
    }
}
export class RestrictedSourceFileException extends __BaseException {
    constructor(opts) {
        super({
            name: "RestrictedSourceFileException",
            $fault: "client",
            ...opts,
        });
        this.name = "RestrictedSourceFileException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RestrictedSourceFileException.prototype);
    }
}
export class SamePathRequestException extends __BaseException {
    constructor(opts) {
        super({
            name: "SamePathRequestException",
            $fault: "client",
            ...opts,
        });
        this.name = "SamePathRequestException";
        this.$fault = "client";
        Object.setPrototypeOf(this, SamePathRequestException.prototype);
    }
}
export class SourceFileOrContentRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "SourceFileOrContentRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "SourceFileOrContentRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, SourceFileOrContentRequiredException.prototype);
    }
}
export const PullRequestStatusEnum = {
    CLOSED: "CLOSED",
    OPEN: "OPEN",
};
export class IdempotencyParameterMismatchException extends __BaseException {
    constructor(opts) {
        super({
            name: "IdempotencyParameterMismatchException",
            $fault: "client",
            ...opts,
        });
        this.name = "IdempotencyParameterMismatchException";
        this.$fault = "client";
        Object.setPrototypeOf(this, IdempotencyParameterMismatchException.prototype);
    }
}
export class InvalidClientRequestTokenException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidClientRequestTokenException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidClientRequestTokenException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidClientRequestTokenException.prototype);
    }
}
export class InvalidDescriptionException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidDescriptionException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidDescriptionException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidDescriptionException.prototype);
    }
}
export class InvalidReferenceNameException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidReferenceNameException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidReferenceNameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidReferenceNameException.prototype);
    }
}
export class InvalidTargetException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidTargetException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidTargetException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidTargetException.prototype);
    }
}
export class InvalidTargetsException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidTargetsException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidTargetsException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidTargetsException.prototype);
    }
}
export class InvalidTitleException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidTitleException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidTitleException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidTitleException.prototype);
    }
}
export class MaximumOpenPullRequestsExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaximumOpenPullRequestsExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaximumOpenPullRequestsExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaximumOpenPullRequestsExceededException.prototype);
    }
}
export class MultipleRepositoriesInPullRequestException extends __BaseException {
    constructor(opts) {
        super({
            name: "MultipleRepositoriesInPullRequestException",
            $fault: "client",
            ...opts,
        });
        this.name = "MultipleRepositoriesInPullRequestException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MultipleRepositoriesInPullRequestException.prototype);
    }
}
export class ReferenceDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "ReferenceDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "ReferenceDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ReferenceDoesNotExistException.prototype);
    }
}
export class ReferenceNameRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ReferenceNameRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ReferenceNameRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ReferenceNameRequiredException.prototype);
    }
}
export class ReferenceTypeNotSupportedException extends __BaseException {
    constructor(opts) {
        super({
            name: "ReferenceTypeNotSupportedException",
            $fault: "client",
            ...opts,
        });
        this.name = "ReferenceTypeNotSupportedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ReferenceTypeNotSupportedException.prototype);
    }
}
export class SourceAndDestinationAreSameException extends __BaseException {
    constructor(opts) {
        super({
            name: "SourceAndDestinationAreSameException",
            $fault: "client",
            ...opts,
        });
        this.name = "SourceAndDestinationAreSameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, SourceAndDestinationAreSameException.prototype);
    }
}
export class TargetRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "TargetRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "TargetRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, TargetRequiredException.prototype);
    }
}
export class TargetsRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "TargetsRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "TargetsRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, TargetsRequiredException.prototype);
    }
}
export class TitleRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "TitleRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "TitleRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, TitleRequiredException.prototype);
    }
}
export class InvalidApprovalRuleContentException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidApprovalRuleContentException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidApprovalRuleContentException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidApprovalRuleContentException.prototype);
    }
}
export class InvalidApprovalRuleNameException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidApprovalRuleNameException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidApprovalRuleNameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidApprovalRuleNameException.prototype);
    }
}
export class InvalidPullRequestIdException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidPullRequestIdException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidPullRequestIdException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidPullRequestIdException.prototype);
    }
}
export class NumberOfRulesExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "NumberOfRulesExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "NumberOfRulesExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NumberOfRulesExceededException.prototype);
    }
}
export class PullRequestAlreadyClosedException extends __BaseException {
    constructor(opts) {
        super({
            name: "PullRequestAlreadyClosedException",
            $fault: "client",
            ...opts,
        });
        this.name = "PullRequestAlreadyClosedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, PullRequestAlreadyClosedException.prototype);
    }
}
export class PullRequestDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "PullRequestDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "PullRequestDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, PullRequestDoesNotExistException.prototype);
    }
}
export class PullRequestIdRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "PullRequestIdRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "PullRequestIdRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, PullRequestIdRequiredException.prototype);
    }
}
export class EncryptionKeyInvalidIdException extends __BaseException {
    constructor(opts) {
        super({
            name: "EncryptionKeyInvalidIdException",
            $fault: "client",
            ...opts,
        });
        this.name = "EncryptionKeyInvalidIdException";
        this.$fault = "client";
        Object.setPrototypeOf(this, EncryptionKeyInvalidIdException.prototype);
    }
}
export class EncryptionKeyInvalidUsageException extends __BaseException {
    constructor(opts) {
        super({
            name: "EncryptionKeyInvalidUsageException",
            $fault: "client",
            ...opts,
        });
        this.name = "EncryptionKeyInvalidUsageException";
        this.$fault = "client";
        Object.setPrototypeOf(this, EncryptionKeyInvalidUsageException.prototype);
    }
}
export class InvalidRepositoryDescriptionException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidRepositoryDescriptionException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRepositoryDescriptionException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRepositoryDescriptionException.prototype);
    }
}
export class InvalidSystemTagUsageException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidSystemTagUsageException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidSystemTagUsageException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidSystemTagUsageException.prototype);
    }
}
export class InvalidTagsMapException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidTagsMapException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidTagsMapException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidTagsMapException.prototype);
    }
}
export class OperationNotAllowedException extends __BaseException {
    constructor(opts) {
        super({
            name: "OperationNotAllowedException",
            $fault: "client",
            ...opts,
        });
        this.name = "OperationNotAllowedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, OperationNotAllowedException.prototype);
    }
}
export class RepositoryLimitExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "RepositoryLimitExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "RepositoryLimitExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RepositoryLimitExceededException.prototype);
    }
}
export class RepositoryNameExistsException extends __BaseException {
    constructor(opts) {
        super({
            name: "RepositoryNameExistsException",
            $fault: "client",
            ...opts,
        });
        this.name = "RepositoryNameExistsException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RepositoryNameExistsException.prototype);
    }
}
export class TagPolicyException extends __BaseException {
    constructor(opts) {
        super({
            name: "TagPolicyException",
            $fault: "client",
            ...opts,
        });
        this.name = "TagPolicyException";
        this.$fault = "client";
        Object.setPrototypeOf(this, TagPolicyException.prototype);
    }
}
export class TooManyTagsException extends __BaseException {
    constructor(opts) {
        super({
            name: "TooManyTagsException",
            $fault: "client",
            ...opts,
        });
        this.name = "TooManyTagsException";
        this.$fault = "client";
        Object.setPrototypeOf(this, TooManyTagsException.prototype);
    }
}
export class ConcurrentReferenceUpdateException extends __BaseException {
    constructor(opts) {
        super({
            name: "ConcurrentReferenceUpdateException",
            $fault: "client",
            ...opts,
        });
        this.name = "ConcurrentReferenceUpdateException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ConcurrentReferenceUpdateException.prototype);
    }
}
export const ReplacementTypeEnum = {
    KEEP_BASE: "KEEP_BASE",
    KEEP_DESTINATION: "KEEP_DESTINATION",
    KEEP_SOURCE: "KEEP_SOURCE",
    USE_NEW_CONTENT: "USE_NEW_CONTENT",
};
export class InvalidConflictResolutionException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidConflictResolutionException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidConflictResolutionException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidConflictResolutionException.prototype);
    }
}
export class InvalidReplacementContentException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidReplacementContentException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidReplacementContentException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidReplacementContentException.prototype);
    }
}
export class InvalidReplacementTypeException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidReplacementTypeException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidReplacementTypeException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidReplacementTypeException.prototype);
    }
}
export class ManualMergeRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ManualMergeRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ManualMergeRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ManualMergeRequiredException.prototype);
    }
}
export class MaximumConflictResolutionEntriesExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaximumConflictResolutionEntriesExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaximumConflictResolutionEntriesExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaximumConflictResolutionEntriesExceededException.prototype);
    }
}
export class MultipleConflictResolutionEntriesException extends __BaseException {
    constructor(opts) {
        super({
            name: "MultipleConflictResolutionEntriesException",
            $fault: "client",
            ...opts,
        });
        this.name = "MultipleConflictResolutionEntriesException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MultipleConflictResolutionEntriesException.prototype);
    }
}
export class ReplacementContentRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ReplacementContentRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ReplacementContentRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ReplacementContentRequiredException.prototype);
    }
}
export class ReplacementTypeRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ReplacementTypeRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ReplacementTypeRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ReplacementTypeRequiredException.prototype);
    }
}
export class DefaultBranchCannotBeDeletedException extends __BaseException {
    constructor(opts) {
        super({
            name: "DefaultBranchCannotBeDeletedException",
            $fault: "client",
            ...opts,
        });
        this.name = "DefaultBranchCannotBeDeletedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, DefaultBranchCannotBeDeletedException.prototype);
    }
}
export class CommentDeletedException extends __BaseException {
    constructor(opts) {
        super({
            name: "CommentDeletedException",
            $fault: "client",
            ...opts,
        });
        this.name = "CommentDeletedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CommentDeletedException.prototype);
    }
}
export class CommentDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "CommentDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "CommentDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CommentDoesNotExistException.prototype);
    }
}
export class CommentIdRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "CommentIdRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "CommentIdRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CommentIdRequiredException.prototype);
    }
}
export class InvalidCommentIdException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidCommentIdException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidCommentIdException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidCommentIdException.prototype);
    }
}
export const PullRequestEventType = {
    PULL_REQUEST_APPROVAL_RULE_CREATED: "PULL_REQUEST_APPROVAL_RULE_CREATED",
    PULL_REQUEST_APPROVAL_RULE_DELETED: "PULL_REQUEST_APPROVAL_RULE_DELETED",
    PULL_REQUEST_APPROVAL_RULE_OVERRIDDEN: "PULL_REQUEST_APPROVAL_RULE_OVERRIDDEN",
    PULL_REQUEST_APPROVAL_RULE_UPDATED: "PULL_REQUEST_APPROVAL_RULE_UPDATED",
    PULL_REQUEST_APPROVAL_STATE_CHANGED: "PULL_REQUEST_APPROVAL_STATE_CHANGED",
    PULL_REQUEST_CREATED: "PULL_REQUEST_CREATED",
    PULL_REQUEST_MERGE_STATE_CHANGED: "PULL_REQUEST_MERGE_STATE_CHANGED",
    PULL_REQUEST_SOURCE_REFERENCE_UPDATED: "PULL_REQUEST_SOURCE_REFERENCE_UPDATED",
    PULL_REQUEST_STATUS_CHANGED: "PULL_REQUEST_STATUS_CHANGED",
};
export class InvalidActorArnException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidActorArnException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidActorArnException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidActorArnException.prototype);
    }
}
export class InvalidMaxResultsException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidMaxResultsException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidMaxResultsException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidMaxResultsException.prototype);
    }
}
export class InvalidPullRequestEventTypeException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidPullRequestEventTypeException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidPullRequestEventTypeException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidPullRequestEventTypeException.prototype);
    }
}
export class InvalidRevisionIdException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidRevisionIdException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRevisionIdException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRevisionIdException.prototype);
    }
}
export class RevisionIdRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "RevisionIdRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "RevisionIdRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RevisionIdRequiredException.prototype);
    }
}
export class RevisionNotCurrentException extends __BaseException {
    constructor(opts) {
        super({
            name: "RevisionNotCurrentException",
            $fault: "client",
            ...opts,
        });
        this.name = "RevisionNotCurrentException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RevisionNotCurrentException.prototype);
    }
}
export class FileTooLargeException extends __BaseException {
    constructor(opts) {
        super({
            name: "FileTooLargeException",
            $fault: "client",
            ...opts,
        });
        this.name = "FileTooLargeException";
        this.$fault = "client";
        Object.setPrototypeOf(this, FileTooLargeException.prototype);
    }
}
export class InvalidBlobIdException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidBlobIdException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidBlobIdException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidBlobIdException.prototype);
    }
}
export class InvalidReactionUserArnException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidReactionUserArnException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidReactionUserArnException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidReactionUserArnException.prototype);
    }
}
export const RelativeFileVersionEnum = {
    AFTER: "AFTER",
    BEFORE: "BEFORE",
};
export class RepositoryNotAssociatedWithPullRequestException extends __BaseException {
    constructor(opts) {
        super({
            name: "RepositoryNotAssociatedWithPullRequestException",
            $fault: "client",
            ...opts,
        });
        this.name = "RepositoryNotAssociatedWithPullRequestException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RepositoryNotAssociatedWithPullRequestException.prototype);
    }
}
export class CommitIdDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "CommitIdDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "CommitIdDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CommitIdDoesNotExistException.prototype);
    }
}
export class PathDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "PathDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "PathDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, PathDoesNotExistException.prototype);
    }
}
export class FolderDoesNotExistException extends __BaseException {
    constructor(opts) {
        super({
            name: "FolderDoesNotExistException",
            $fault: "client",
            ...opts,
        });
        this.name = "FolderDoesNotExistException";
        this.$fault = "client";
        Object.setPrototypeOf(this, FolderDoesNotExistException.prototype);
    }
}
export class InvalidDestinationCommitSpecifierException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidDestinationCommitSpecifierException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidDestinationCommitSpecifierException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidDestinationCommitSpecifierException.prototype);
    }
}
export class InvalidSourceCommitSpecifierException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidSourceCommitSpecifierException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidSourceCommitSpecifierException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidSourceCommitSpecifierException.prototype);
    }
}
export const RepositoryTriggerEventEnum = {
    ALL: "all",
    CREATE_REFERENCE: "createReference",
    DELETE_REFERENCE: "deleteReference",
    UPDATE_REFERENCE: "updateReference",
};
