// @aws-sdk/client-codecommit/dist-es/models/models_1.js
import { CodeCommitServiceException as __BaseException } from "./CodeCommitServiceException";
export class InvalidAuthorArnException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidAuthorArnException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidAuthorArnException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidAuthorArnException.prototype);
    }
}
export class InvalidPullRequestStatusException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidPullRequestStatusException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidPullRequestStatusException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidPullRequestStatusException.prototype);
    }
}
export class InvalidOrderException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidOrderException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidOrderException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidOrderException.prototype);
    }
}
export class InvalidSortByException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidSortByException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidSortByException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidSortByException.prototype);
    }
}
export const OrderEnum = {
    ASCENDING: "ascending",
    DESCENDING: "descending",
};
export const SortByEnum = {
    MODIFIED_DATE: "lastModifiedDate",
    REPOSITORY_NAME: "repositoryName",
};
export class InvalidResourceArnException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidResourceArnException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidResourceArnException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidResourceArnException.prototype);
    }
}
export class ResourceArnRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ResourceArnRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ResourceArnRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceArnRequiredException.prototype);
    }
}
export class InvalidTargetBranchException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidTargetBranchException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidTargetBranchException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidTargetBranchException.prototype);
    }
}
export class PullRequestApprovalRulesNotSatisfiedException extends __BaseException {
    constructor(opts) {
        super({
            name: "PullRequestApprovalRulesNotSatisfiedException",
            $fault: "client",
            ...opts,
        });
        this.name = "PullRequestApprovalRulesNotSatisfiedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, PullRequestApprovalRulesNotSatisfiedException.prototype);
    }
}
export class TipOfSourceReferenceIsDifferentException extends __BaseException {
    constructor(opts) {
        super({
            name: "TipOfSourceReferenceIsDifferentException",
            $fault: "client",
            ...opts,
        });
        this.name = "TipOfSourceReferenceIsDifferentException";
        this.$fault = "client";
        Object.setPrototypeOf(this, TipOfSourceReferenceIsDifferentException.prototype);
    }
}
export class InvalidOverrideStatusException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidOverrideStatusException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidOverrideStatusException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidOverrideStatusException.prototype);
    }
}
export class OverrideAlreadySetException extends __BaseException {
    constructor(opts) {
        super({
            name: "OverrideAlreadySetException",
            $fault: "client",
            ...opts,
        });
        this.name = "OverrideAlreadySetException";
        this.$fault = "client";
        Object.setPrototypeOf(this, OverrideAlreadySetException.prototype);
    }
}
export class OverrideStatusRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "OverrideStatusRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "OverrideStatusRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, OverrideStatusRequiredException.prototype);
    }
}
export class CommentContentRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "CommentContentRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "CommentContentRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CommentContentRequiredException.prototype);
    }
}
export class CommentContentSizeLimitExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "CommentContentSizeLimitExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "CommentContentSizeLimitExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CommentContentSizeLimitExceededException.prototype);
    }
}
export class InvalidFileLocationException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidFileLocationException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidFileLocationException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidFileLocationException.prototype);
    }
}
export class InvalidFilePositionException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidFilePositionException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidFilePositionException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidFilePositionException.prototype);
    }
}
export class InvalidRelativeFileVersionEnumException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidRelativeFileVersionEnumException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRelativeFileVersionEnumException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRelativeFileVersionEnumException.prototype);
    }
}
export class InvalidReactionValueException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidReactionValueException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidReactionValueException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidReactionValueException.prototype);
    }
}
export class ReactionLimitExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "ReactionLimitExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "ReactionLimitExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ReactionLimitExceededException.prototype);
    }
}
export class ReactionValueRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "ReactionValueRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "ReactionValueRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ReactionValueRequiredException.prototype);
    }
}
export class FileContentRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "FileContentRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "FileContentRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, FileContentRequiredException.prototype);
    }
}
export class SameFileContentException extends __BaseException {
    constructor(opts) {
        super({
            name: "SameFileContentException",
            $fault: "client",
            ...opts,
        });
        this.name = "SameFileContentException";
        this.$fault = "client";
        Object.setPrototypeOf(this, SameFileContentException.prototype);
    }
}
export class InvalidRepositoryTriggerBranchNameException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidRepositoryTriggerBranchNameException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRepositoryTriggerBranchNameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRepositoryTriggerBranchNameException.prototype);
    }
}
export class InvalidRepositoryTriggerCustomDataException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidRepositoryTriggerCustomDataException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRepositoryTriggerCustomDataException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRepositoryTriggerCustomDataException.prototype);
    }
}
export class InvalidRepositoryTriggerDestinationArnException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidRepositoryTriggerDestinationArnException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRepositoryTriggerDestinationArnException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRepositoryTriggerDestinationArnException.prototype);
    }
}
export class InvalidRepositoryTriggerEventsException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidRepositoryTriggerEventsException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRepositoryTriggerEventsException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRepositoryTriggerEventsException.prototype);
    }
}
export class InvalidRepositoryTriggerNameException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidRepositoryTriggerNameException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRepositoryTriggerNameException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRepositoryTriggerNameException.prototype);
    }
}
export class InvalidRepositoryTriggerRegionException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidRepositoryTriggerRegionException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRepositoryTriggerRegionException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRepositoryTriggerRegionException.prototype);
    }
}
export class MaximumBranchesExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaximumBranchesExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaximumBranchesExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaximumBranchesExceededException.prototype);
    }
}
export class MaximumRepositoryTriggersExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaximumRepositoryTriggersExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaximumRepositoryTriggersExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaximumRepositoryTriggersExceededException.prototype);
    }
}
export class RepositoryTriggerBranchNameListRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "RepositoryTriggerBranchNameListRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "RepositoryTriggerBranchNameListRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RepositoryTriggerBranchNameListRequiredException.prototype);
    }
}
export class RepositoryTriggerDestinationArnRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "RepositoryTriggerDestinationArnRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "RepositoryTriggerDestinationArnRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RepositoryTriggerDestinationArnRequiredException.prototype);
    }
}
export class RepositoryTriggerEventsListRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "RepositoryTriggerEventsListRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "RepositoryTriggerEventsListRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RepositoryTriggerEventsListRequiredException.prototype);
    }
}
export class RepositoryTriggerNameRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "RepositoryTriggerNameRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "RepositoryTriggerNameRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RepositoryTriggerNameRequiredException.prototype);
    }
}
export class RepositoryTriggersListRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "RepositoryTriggersListRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "RepositoryTriggersListRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RepositoryTriggersListRequiredException.prototype);
    }
}
export class TagsMapRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "TagsMapRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "TagsMapRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, TagsMapRequiredException.prototype);
    }
}
export class InvalidTagKeysListException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidTagKeysListException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidTagKeysListException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidTagKeysListException.prototype);
    }
}
export class TagKeysListRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "TagKeysListRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "TagKeysListRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, TagKeysListRequiredException.prototype);
    }
}
export class InvalidRuleContentSha256Exception extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidRuleContentSha256Exception",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidRuleContentSha256Exception";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidRuleContentSha256Exception.prototype);
    }
}
export class CommentNotCreatedByCallerException extends __BaseException {
    constructor(opts) {
        super({
            name: "CommentNotCreatedByCallerException",
            $fault: "client",
            ...opts,
        });
        this.name = "CommentNotCreatedByCallerException";
        this.$fault = "client";
        Object.setPrototypeOf(this, CommentNotCreatedByCallerException.prototype);
    }
}
export class InvalidApprovalStateException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidApprovalStateException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidApprovalStateException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidApprovalStateException.prototype);
    }
}
export class MaximumNumberOfApprovalsExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "MaximumNumberOfApprovalsExceededException",
            $fault: "client",
            ...opts,
        });
        this.name = "MaximumNumberOfApprovalsExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MaximumNumberOfApprovalsExceededException.prototype);
    }
}
export class PullRequestCannotBeApprovedByAuthorException extends __BaseException {
    constructor(opts) {
        super({
            name: "PullRequestCannotBeApprovedByAuthorException",
            $fault: "client",
            ...opts,
        });
        this.name = "PullRequestCannotBeApprovedByAuthorException";
        this.$fault = "client";
        Object.setPrototypeOf(this, PullRequestCannotBeApprovedByAuthorException.prototype);
    }
}
export class InvalidPullRequestStatusUpdateException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidPullRequestStatusUpdateException",
            $fault: "client",
            ...opts,
        });
        this.name = "InvalidPullRequestStatusUpdateException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidPullRequestStatusUpdateException.prototype);
    }
}
export class PullRequestStatusRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "PullRequestStatusRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "PullRequestStatusRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, PullRequestStatusRequiredException.prototype);
    }
}
export class EncryptionKeyRequiredException extends __BaseException {
    constructor(opts) {
        super({
            name: "EncryptionKeyRequiredException",
            $fault: "client",
            ...opts,
        });
        this.name = "EncryptionKeyRequiredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, EncryptionKeyRequiredException.prototype);
    }
}
