// @backstage/plugin-auth-backend-module-cloudflare-access-provider/dist/module.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var pluginAuthNode = require('@backstage/plugin-auth-node');
var authenticator = require('./authenticator.cjs.js');
var resolvers = require('./resolvers.cjs.js');

const authModuleCloudflareAccessProvider = backendPluginApi.createBackendModule({
  pluginId: "auth",
  moduleId: "cloudflare-access-provider",
  register(reg) {
    reg.registerInit({
      deps: {
        authProviders: pluginAuthNode.authProvidersExtensionPoint,
        cache: backendPluginApi.coreServices.cache
      },
      async init({ authProviders, cache }) {
        authProviders.registerProvider({
          providerId: "cfaccess",
          factory: pluginAuthNode.createProxyAuthProviderFactory({
            authenticator: authenticator.createCloudflareAccessAuthenticator({ cache }),
            signInResolverFactories: {
              ...resolvers.cloudflareAccessSignInResolvers,
              ...pluginAuthNode.commonSignInResolvers
            }
          })
        });
      }
    });
  }
});

exports.authModuleCloudflareAccessProvider = authModuleCloudflareAccessProvider;
