// @backstage/plugin-catalog-react/dist/components/EntityProcessingStatusPicker/EntityProcessingStatusPicker.esm.js
import { EntityOrphanFilter, EntityErrorFilter } from '../../filters.esm.js';
import Box from '@material-ui/core/Box';
import Checkbox from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import { makeStyles, createStyles } from '@material-ui/core/styles';
import CheckBoxIcon from '@material-ui/icons/CheckBox';
import CheckBoxOutlineBlankIcon from '@material-ui/icons/CheckBoxOutlineBlank';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import React, { useState } from 'react';
import '../../hooks/useEntity.esm.js';
import { useEntityList } from '../../hooks/useEntityListProvider.esm.js';
import 'react-use/esm/useAsync';
import 'lodash/isEqual';
import 'lodash/sortBy';
import '@backstage/core-plugin-api';
import '../../api.esm.js';
import '@backstage/catalog-model';
import 'react-use/esm/useObservable';
import '../../apis/EntityPresentationApi/EntityPresentationApi.esm.js';
import 'lodash/get';
import '../../apis/StarredEntitiesApi/StarredEntitiesApi.esm.js';
import 'zen-observable';
import Autocomplete from '@material-ui/lab/Autocomplete';
import { catalogReactTranslationRef } from '../../translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

const useStyles = makeStyles(
  (theme) => createStyles({
    root: {},
    input: {
      backgroundColor: theme.palette.background.paper
    },
    label: {
      textTransform: "none",
      fontWeight: "bold"
    }
  }),
  { name: "CatalogReactEntityProcessingStatusPickerPicker" }
);
const icon = /* @__PURE__ */ React.createElement(CheckBoxOutlineBlankIcon, { fontSize: "small" });
const checkedIcon = /* @__PURE__ */ React.createElement(CheckBoxIcon, { fontSize: "small" });
const EntityProcessingStatusPicker = () => {
  const classes = useStyles();
  const { updateFilters } = useEntityList();
  const { t } = useTranslationRef(catalogReactTranslationRef);
  const [selectedAdvancedItems, setSelectedAdvancedItems] = useState(
    []
  );
  function orphanChange(value) {
    updateFilters({
      orphan: value ? new EntityOrphanFilter(value) : void 0
    });
  }
  function errorChange(value) {
    updateFilters({
      error: value ? new EntityErrorFilter(value) : void 0
    });
  }
  const availableAdvancedItems = ["Is Orphan", "Has Error"];
  return /* @__PURE__ */ React.createElement(Box, { className: classes.root, pb: 1, pt: 1 }, /* @__PURE__ */ React.createElement(Typography, { className: classes.label, variant: "button", component: "label" }, t("entityProcessingStatusPicker.title"), /* @__PURE__ */ React.createElement(
    Autocomplete,
    {
      PopperComponent: (popperProps) => /* @__PURE__ */ React.createElement("div", { ...popperProps }, popperProps.children),
      multiple: true,
      disableCloseOnSelect: true,
      options: availableAdvancedItems,
      value: selectedAdvancedItems,
      onChange: (_, value) => {
        setSelectedAdvancedItems(value);
        orphanChange(value.includes("Is Orphan"));
        errorChange(value.includes("Has Error"));
      },
      renderOption: (option, { selected }) => /* @__PURE__ */ React.createElement(
        FormControlLabel,
        {
          control: /* @__PURE__ */ React.createElement(
            Checkbox,
            {
              icon,
              checkedIcon,
              checked: selected
            }
          ),
          onClick: (event) => event.preventDefault(),
          label: option
        }
      ),
      size: "small",
      popupIcon: /* @__PURE__ */ React.createElement(ExpandMoreIcon, { "data-testid": "processing-status-picker-expand" }),
      renderInput: (params) => /* @__PURE__ */ React.createElement(
        TextField,
        {
          ...params,
          className: classes.input,
          variant: "outlined"
        }
      )
    }
  )));
};

export { EntityProcessingStatusPicker };
