// @backstage/plugin-catalog/dist/components/CatalogPage/DefaultCatalogPage.esm.js
import { PageWithHeader, Content, ContentHeader, CreateButton, SupportButton } from '@backstage/core-components';
import { useApi, configApiRef, useRouteRef } from '@backstage/core-plugin-api';
import { EntityListProvider, CatalogFilterLayout, DefaultFilters } from '@backstage/plugin-catalog-react';
import React from 'react';
import { createComponentRouteRef } from '../../routes.esm.js';
import { CatalogTable } from '../CatalogTable/CatalogTable.esm.js';
import { catalogTranslationRef } from '../../alpha/translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { catalogEntityCreatePermission } from '@backstage/plugin-catalog-common/alpha';
import { usePermission } from '@backstage/plugin-permission-react';

function BaseCatalogPage(props) {
  const { filters, content = /* @__PURE__ */ React.createElement(CatalogTable, null), pagination } = props;
  const orgName = useApi(configApiRef).getOptionalString("organization.name") ?? "Backstage";
  const createComponentLink = useRouteRef(createComponentRouteRef);
  const { t } = useTranslationRef(catalogTranslationRef);
  const { allowed } = usePermission({
    permission: catalogEntityCreatePermission
  });
  return /* @__PURE__ */ React.createElement(PageWithHeader, { title: t("indexPage.title", { orgName }), themeId: "home" }, /* @__PURE__ */ React.createElement(Content, null, /* @__PURE__ */ React.createElement(ContentHeader, { title: "" }, allowed && /* @__PURE__ */ React.createElement(
    CreateButton,
    {
      title: t("indexPage.createButtonTitle"),
      to: createComponentLink && createComponentLink()
    }
  ), /* @__PURE__ */ React.createElement(SupportButton, null, t("indexPage.supportButtonContent"))), /* @__PURE__ */ React.createElement(EntityListProvider, { pagination }, /* @__PURE__ */ React.createElement(CatalogFilterLayout, null, /* @__PURE__ */ React.createElement(CatalogFilterLayout.Filters, null, filters), /* @__PURE__ */ React.createElement(CatalogFilterLayout.Content, null, content)))));
}
function DefaultCatalogPage(props) {
  const {
    columns,
    actions,
    initiallySelectedFilter = "owned",
    initialKind = "component",
    tableOptions = {},
    emptyContent,
    pagination,
    ownerPickerMode,
    filters,
    initiallySelectedNamespaces
  } = props;
  return /* @__PURE__ */ React.createElement(
    BaseCatalogPage,
    {
      filters: filters ?? /* @__PURE__ */ React.createElement(
        DefaultFilters,
        {
          initialKind,
          initiallySelectedFilter,
          ownerPickerMode,
          initiallySelectedNamespaces
        }
      ),
      content: /* @__PURE__ */ React.createElement(
        CatalogTable,
        {
          columns,
          actions,
          tableOptions,
          emptyContent
        }
      ),
      pagination
    }
  );
}

export { BaseCatalogPage, DefaultCatalogPage };
