// @backstage/cli/dist/lib/new/factories/index.cjs.js
'use strict';

var frontendPlugin = require('./frontendPlugin.cjs.js');
var backendPlugin = require('./backendPlugin.cjs.js');
var backendModule = require('./backendModule.cjs.js');
var nodeLibraryPackage = require('./nodeLibraryPackage.cjs.js');
var webLibraryPackage = require('./webLibraryPackage.cjs.js');
var pluginCommon = require('./pluginCommon.cjs.js');
var pluginNode = require('./pluginNode.cjs.js');
var pluginWeb = require('./pluginWeb.cjs.js');
var scaffolderModule = require('./scaffolderModule.cjs.js');



exports.frontendPlugin = frontendPlugin.frontendPlugin;
exports.backendPlugin = backendPlugin.backendPlugin;
exports.backendModule = backendModule.backendModule;
exports.nodeLibraryPackage = nodeLibraryPackage.nodeLibraryPackage;
exports.webLibraryPackage = webLibraryPackage.webLibraryPackage;
exports.pluginCommon = pluginCommon.pluginCommon;
exports.pluginNode = pluginNode.pluginNode;
exports.pluginWeb = pluginWeb.pluginWeb;
exports.scaffolderModule = scaffolderModule.scaffolderModule;
