// @backstage/cli/dist/lib/new/factories/webLibraryPackage.cjs.js
'use strict';

var chalk = require('chalk');
var paths = require('../../paths.cjs.js');
var codeowners = require('../../codeowners/codeowners.cjs.js');
var types = require('../types.cjs.js');
var tasks = require('../../tasks.cjs.js');
var prompts = require('./common/prompts.cjs.js');
var tasks$1 = require('./common/tasks.cjs.js');
var util = require('./common/util.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);

const webLibraryPackage = types.createFactory({
  name: "web-library",
  description: "A new web-library package, exporting shared functionality for frontend plugins",
  optionsDiscovery: async () => ({
    codeOwnersPath: await codeowners.getCodeownersFilePath(paths.paths.targetRoot)
  }),
  optionsPrompts: [prompts.pluginIdPrompt(), prompts.ownerPrompt()],
  async create(options, ctx) {
    const { id } = options;
    const name = util.resolvePackageName({
      baseName: id,
      scope: ctx.scope,
      plugin: false
    });
    tasks.Task.log();
    tasks.Task.log(`Creating web-library package ${chalk__default.default.cyan(name)}`);
    const targetDir = ctx.isMonoRepo ? paths.paths.resolveTargetRoot("packages", id) : paths.paths.resolveTargetRoot(`${id}`);
    await tasks$1.executePluginPackageTemplate(ctx, {
      targetDir,
      templateName: "web-library-package",
      values: {
        id,
        name,
        pluginVersion: ctx.defaultVersion,
        privatePackage: ctx.private,
        npmRegistry: ctx.npmRegistry,
        license: ctx.license
      }
    });
    if (options.owner) {
      await codeowners.addCodeownersEntry(`/packages/${id}`, options.owner);
    }
    await tasks.Task.forCommand("yarn install", { cwd: targetDir, optional: true });
    await tasks.Task.forCommand("yarn lint --fix", {
      cwd: targetDir,
      optional: true
    });
  }
});

exports.webLibraryPackage = webLibraryPackage;
