// @backstage/core-plugin-api/dist/index.esm.js
export { AnalyticsContext } from './analytics/AnalyticsContext.esm.js';
export { useAnalytics } from './analytics/useAnalytics.esm.js';
export { useApi, useApiHolder, withApis } from './apis/system/useApi.esm.js';
export { createApiRef } from './apis/system/ApiRef.esm.js';
export { createApiFactory } from './apis/system/helpers.esm.js';
export { SessionState, atlassianAuthApiRef, bitbucketAuthApiRef, bitbucketServerAuthApiRef, githubAuthApiRef, gitlabAuthApiRef, googleAuthApiRef, microsoftAuthApiRef, oktaAuthApiRef, oneloginAuthApiRef, vmwareCloudAuthApiRef } from './apis/definitions/auth.esm.js';
export { alertApiRef } from './apis/definitions/AlertApi.esm.js';
export { analyticsApiRef } from './apis/definitions/AnalyticsApi.esm.js';
export { appThemeApiRef } from './apis/definitions/AppThemeApi.esm.js';
export { configApiRef } from './apis/definitions/ConfigApi.esm.js';
export { discoveryApiRef } from './apis/definitions/DiscoveryApi.esm.js';
export { errorApiRef } from './apis/definitions/ErrorApi.esm.js';
export { FeatureFlagState, featureFlagsApiRef } from './apis/definitions/FeatureFlagsApi.esm.js';
export { fetchApiRef } from './apis/definitions/FetchApi.esm.js';
export { identityApiRef } from './apis/definitions/IdentityApi.esm.js';
export { oauthRequestApiRef } from './apis/definitions/OAuthRequestApi.esm.js';
export { storageApiRef } from './apis/definitions/StorageApi.esm.js';
export { useApp } from './app/useApp.esm.js';
export { attachComponentData, getComponentData } from './extensions/componentData.esm.js';
export { createComponentExtension, createReactExtension, createRoutableExtension } from './extensions/extensions.esm.js';
export { useElementFilter } from './extensions/useElementFilter.esm.js';
export { createPlugin } from './plugin/Plugin.esm.js';
export { createRouteRef } from './routing/RouteRef.esm.js';
export { createSubRouteRef } from './routing/SubRouteRef.esm.js';
export { createExternalRouteRef } from './routing/ExternalRouteRef.esm.js';
export { useRouteRef } from './routing/useRouteRef.esm.js';
export { useRouteRefParams } from './routing/useRouteRefParams.esm.js';
