// @backstage/plugin-api-docs/dist/components/ApisCards/ConsumedApisCard.esm.js
import { RELATION_CONSUMES_API } from '@backstage/catalog-model';
import Typography from '@material-ui/core/Typography';
import { useEntity, useRelatedEntities, EntityTable } from '@backstage/plugin-catalog-react';
import React from 'react';
import { apiEntityColumns } from './presets.esm.js';
import { InfoCard, Progress, WarningPanel, CodeSnippet, Link } from '@backstage/core-components';

const ConsumedApisCard = (props) => {
  const {
    variant = "gridItem",
    title = "Consumed APIs",
    columns = apiEntityColumns,
    tableOptions = {}
  } = props;
  const { entity } = useEntity();
  const { entities, loading, error } = useRelatedEntities(entity, {
    type: RELATION_CONSUMES_API
  });
  if (loading) {
    return /* @__PURE__ */ React.createElement(InfoCard, { variant, title }, /* @__PURE__ */ React.createElement(Progress, null));
  }
  if (error || !entities) {
    return /* @__PURE__ */ React.createElement(InfoCard, { variant, title }, /* @__PURE__ */ React.createElement(
      WarningPanel,
      {
        severity: "error",
        title: "Could not load APIs",
        message: /* @__PURE__ */ React.createElement(CodeSnippet, { text: `${error}`, language: "text" })
      }
    ));
  }
  return /* @__PURE__ */ React.createElement(
    EntityTable,
    {
      title,
      variant,
      emptyContent: /* @__PURE__ */ React.createElement("div", { style: { textAlign: "center" } }, /* @__PURE__ */ React.createElement(Typography, { variant: "body1" }, "This ", entity.kind.toLocaleLowerCase("en-US"), " does not consume any APIs."), /* @__PURE__ */ React.createElement(Typography, { variant: "body2" }, /* @__PURE__ */ React.createElement(
        Link,
        {
          to: "https://backstage.io/docs/features/software-catalog/descriptor-format#specconsumesapis-optional",
          externalLinkIcon: true
        },
        "Learn how to change this"
      ))),
      columns,
      tableOptions,
      entities
    }
  );
};

export { ConsumedApisCard };
