// @backstage/plugin-api-docs/dist/components/ApisCards/HasApisCard.esm.js
import { RELATION_HAS_PART } from '@backstage/catalog-model';
import Typography from '@material-ui/core/Typography';
import { EntityTable, useEntity, useRelatedEntities } from '@backstage/plugin-catalog-react';
import React from 'react';
import { createSpecApiTypeColumn } from './presets.esm.js';
import { InfoCard, Progress, WarningPanel, CodeSnippet, Link } from '@backstage/core-components';

const presetColumns = [
  EntityTable.columns.createEntityRefColumn({ defaultKind: "API" }),
  EntityTable.columns.createOwnerColumn(),
  createSpecApiTypeColumn(),
  EntityTable.columns.createSpecLifecycleColumn(),
  EntityTable.columns.createMetadataDescriptionColumn()
];
const HasApisCard = (props) => {
  const {
    variant = "gridItem",
    title = "APIs",
    columns = presetColumns,
    tableOptions = {}
  } = props;
  const { entity } = useEntity();
  const { entities, loading, error } = useRelatedEntities(entity, {
    type: RELATION_HAS_PART,
    kind: "API"
  });
  if (loading) {
    return /* @__PURE__ */ React.createElement(InfoCard, { variant, title }, /* @__PURE__ */ React.createElement(Progress, null));
  }
  if (error || !entities) {
    return /* @__PURE__ */ React.createElement(InfoCard, { variant, title }, /* @__PURE__ */ React.createElement(
      WarningPanel,
      {
        severity: "error",
        title: "Could not load APIs",
        message: /* @__PURE__ */ React.createElement(CodeSnippet, { text: `${error}`, language: "text" })
      }
    ));
  }
  return /* @__PURE__ */ React.createElement(
    EntityTable,
    {
      title,
      variant,
      emptyContent: /* @__PURE__ */ React.createElement("div", { style: { textAlign: "center" } }, /* @__PURE__ */ React.createElement(Typography, { variant: "body1" }, "This ", entity.kind.toLocaleLowerCase("en-US"), " does not contain any APIs."), /* @__PURE__ */ React.createElement(Typography, { variant: "body2" }, /* @__PURE__ */ React.createElement(Link, { to: "https://backstage.io/docs/features/software-catalog/descriptor-format#kind-api" }, "Learn how to change this."))),
      columns,
      tableOptions,
      entities
    }
  );
};

export { HasApisCard };
