// @backstage/plugin-api-docs/dist/components/ApisCards/presets.esm.js
import { EntityTable } from '@backstage/plugin-catalog-react';
import ExtensionIcon from '@material-ui/icons/Extension';
import ToggleButton from '@material-ui/lab/ToggleButton';
import React, { useState } from 'react';
import '@material-ui/lab/Alert';
import '../../config.esm.js';
import '@backstage/core-components';
import '@backstage/core-plugin-api';
import '../AsyncApiDefinitionWidget/AsyncApiDefinitionWidget.esm.js';
import '../GraphQlDefinitionWidget/GraphQlDefinitionWidget.esm.js';
import '../OpenApiDefinitionWidget/OpenApiDefinitionWidget.esm.js';
import '@material-ui/core/styles';
import { ApiTypeTitle } from '../ApiDefinitionCard/ApiTypeTitle.esm.js';
import { ApiDefinitionDialog } from '../ApiDefinitionDialog/ApiDefinitionDialog.esm.js';

function createSpecApiTypeColumn() {
  return {
    title: "Type",
    field: "spec.type",
    render: (entity) => /* @__PURE__ */ React.createElement(ApiTypeTitle, { apiEntity: entity })
  };
}
const ApiDefinitionButton = ({ apiEntity }) => {
  const [dialogOpen, setDialogOpen] = useState(false);
  return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(
    ToggleButton,
    {
      "aria-label": "Toggle API Definition Dialog",
      onClick: () => setDialogOpen(!dialogOpen),
      value: dialogOpen
    },
    /* @__PURE__ */ React.createElement(ExtensionIcon, null)
  ), /* @__PURE__ */ React.createElement(
    ApiDefinitionDialog,
    {
      entity: apiEntity,
      open: dialogOpen,
      onClose: () => setDialogOpen(false)
    }
  ));
};
function createApiDefinitionColumn() {
  return {
    title: "API Definition",
    render: (entity) => /* @__PURE__ */ React.createElement(ApiDefinitionButton, { apiEntity: entity })
  };
}
const apiEntityColumns = [
  EntityTable.columns.createEntityRefColumn({ defaultKind: "API" }),
  EntityTable.columns.createSystemColumn(),
  EntityTable.columns.createOwnerColumn(),
  createSpecApiTypeColumn(),
  EntityTable.columns.createSpecLifecycleColumn(),
  EntityTable.columns.createMetadataDescriptionColumn(),
  createApiDefinitionColumn()
];

export { apiEntityColumns, createSpecApiTypeColumn };
