// @backstage/plugin-kubernetes-react/dist/components/DaemonSetsAccordions/DaemonSetsDrawer.esm.js
import React__default from 'react';
import { KubernetesStructuredMetadataTableDrawer } from '../KubernetesDrawer/KubernetesStructuredMetadataTableDrawer.esm.js';
import '../KubernetesDrawer/KubernetesDrawer.esm.js';
import '@backstage/core-components';
import '@material-ui/core/FormControlLabel';
import '@material-ui/core/Switch';
import 'js-yaml';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import Chip from '@material-ui/core/Chip';

const DaemonSetDrawer = ({
  daemonset,
  expanded
}) => {
  const namespace = daemonset.metadata?.namespace;
  return /* @__PURE__ */ React__default.createElement(
    KubernetesStructuredMetadataTableDrawer,
    {
      object: daemonset,
      expanded,
      kind: "DaemonSet",
      renderObject: (daemonsetObj) => {
        return {
          updateStrategyType: daemonsetObj.spec?.updateStrategy?.type ?? "???",
          minReadySeconds: daemonsetObj.spec?.minReadySeconds ?? "???",
          revisionHistoryLimit: daemonsetObj.spec?.revisionHistoryLimit ?? "???",
          currentNumberScheduled: daemonsetObj.status?.currentNumberScheduled ?? "???",
          desiredNumberScheduled: daemonsetObj.status?.desiredNumberScheduled ?? "???",
          numberAvailable: daemonsetObj.status?.numberAvailable ?? "???",
          numberMisscheduled: daemonsetObj.status?.numberMisscheduled ?? "???",
          numberReady: daemonsetObj.status?.numberReady ?? "???"
        };
      }
    },
    /* @__PURE__ */ React__default.createElement(
      Grid,
      {
        container: true,
        direction: "column",
        justifyContent: "flex-start",
        alignItems: "flex-start",
        spacing: 0
      },
      /* @__PURE__ */ React__default.createElement(Grid, { item: true }, /* @__PURE__ */ React__default.createElement(Typography, { variant: "body1" }, daemonset.metadata?.name ?? "unknown object")),
      /* @__PURE__ */ React__default.createElement(Grid, { item: true }, /* @__PURE__ */ React__default.createElement(Typography, { color: "textSecondary", variant: "subtitle1" }, "DaemonSet")),
      namespace && /* @__PURE__ */ React__default.createElement(Grid, { item: true }, /* @__PURE__ */ React__default.createElement(Chip, { size: "small", label: `namespace: ${namespace}` }))
    )
  );
};

export { DaemonSetDrawer };
