// @swagger-api/apidom-ns-openapi-2/src/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard").default;
exports.__esModule = true;
exports.specificationObj = exports.refractorPluginReplaceEmptyElement = exports.refract = exports.mediaTypes = exports.keyMap = exports.isXmlElement = exports.isTagElement = exports.isSwaggerVersionElement = exports.isSwaggerExtension = exports.isSwaggerElement = exports.isStringElement = exports.isSecuritySchemeElement = exports.isSecurityRequirementElement = exports.isSecurityDefinitionsElement = exports.isScopesElement = exports.isSchemaElement = exports.isResponsesElement = exports.isResponsesDefinitionsElement = exports.isResponseElement = exports.isReferenceLikeElement = exports.isReferenceElement = exports.isRefElement = exports.isPathsElement = exports.isPathItemElement = exports.isParametersDefinitionsElement = exports.isParameterElement = exports.isOperationElement = exports.isObjectElement = exports.isNumberElement = exports.isNullElement = exports.isMemberElement = exports.isLinkPrimitiveElement = exports.isLicenseElement = exports.isJSONReferenceLikeElement = exports.isJSONReferenceElement = exports.isItemsElement = exports.isInfoElement = exports.isHeadersElement = exports.isHeaderElement = exports.isExternalDocumentationElement = exports.isExampleElement = exports.isElement = exports.isDefinitionsElement = exports.isContactElement = exports.isBooleanElement = exports.isArrayElement = exports.getNodeType = exports.default = exports.createRefractor = exports.XmlElement = exports.Visitor = exports.TagElement = exports.SwaggerVersionElement = exports.SwaggerTagsElement = exports.SwaggerSecurityElement = exports.SwaggerSchemesElement = exports.SwaggerProducesElement = exports.SwaggerElement = exports.SwaggerConsumesElement = exports.SpecificationVisitor = exports.SecuritySchemeElement = exports.SecurityRequirementElement = exports.SecurityDefinitionsElement = exports.ScopesElement = exports.SchemaElement = exports.ResponsesElement = exports.ResponsesDefinitionsElement = exports.ResponseElement = exports.ReferenceElement = exports.PatternedFieldsVisitor = exports.PathsElement = exports.PathItemParametersElement = exports.PathItemElement = exports.ParametersDefinitionsElement = exports.ParameterElement = exports.OperationTagsElement = exports.OperationSecurityElement = exports.OperationSchemesElement = exports.OperationProducesElement = exports.OperationParametersElement = exports.OperationElement = exports.OperationConsumesElement = exports.OpenAPIMediaTypes = exports.MixedFieldsVisitor = exports.MapVisitor = exports.LicenseElement = exports.JSONReferenceElement = exports.ItemsElement = exports.InfoElement = exports.HeadersElement = exports.HeaderElement = exports.FixedFieldsVisitor = exports.FallbackVisitor = exports.ExternalDocumentationElement = exports.ExampleElement = exports.DefinitionsElement = exports.ContactElement = exports.AlternatingVisitor = void 0;
var _apidomCore = require("@swagger-api/apidom-core");
exports.isRefElement = _apidomCore.isRefElement;
exports.isLinkPrimitiveElement = _apidomCore.isLinkElement;
exports.isMemberElement = _apidomCore.isMemberElement;
exports.isObjectElement = _apidomCore.isObjectElement;
exports.isArrayElement = _apidomCore.isArrayElement;
exports.isBooleanElement = _apidomCore.isBooleanElement;
exports.isNullElement = _apidomCore.isNullElement;
exports.isElement = _apidomCore.isElement;
exports.isNumberElement = _apidomCore.isNumberElement;
exports.isStringElement = _apidomCore.isStringElement;
var _apidomNsJsonSchemaDraft = require("@swagger-api/apidom-ns-json-schema-draft-4");
exports.isJSONReferenceElement = _apidomNsJsonSchemaDraft.isJSONReferenceElement;
exports.isJSONReferenceLikeElement = _apidomNsJsonSchemaDraft.isJSONReferenceLikeElement;
exports.JSONReferenceElement = _apidomNsJsonSchemaDraft.JSONReferenceElement;
var _mediaTypes = _interopRequireWildcard(require("./media-types.cjs"));
exports.mediaTypes = _mediaTypes.default;
exports.OpenAPIMediaTypes = _mediaTypes.OpenAPIMediaTypes;
var _namespace = _interopRequireDefault(require("./namespace.cjs"));
exports.default = _namespace.default;
var _replaceEmptyElement = _interopRequireDefault(require("./refractor/plugins/replace-empty-element.cjs"));
exports.refractorPluginReplaceEmptyElement = _replaceEmptyElement.default;
var _index = _interopRequireWildcard(require("./refractor/index.cjs"));
exports.refract = _index.default;
exports.createRefractor = _index.createRefractor;
var _specification = _interopRequireDefault(require("./refractor/specification.cjs"));
exports.specificationObj = _specification.default;
var _predicates = require("./predicates.cjs");
exports.isSwaggerElement = _predicates.isSwaggerElement;
exports.isSwaggerVersionElement = _predicates.isSwaggerVersionElement;
exports.isInfoElement = _predicates.isInfoElement;
exports.isContactElement = _predicates.isContactElement;
exports.isLicenseElement = _predicates.isLicenseElement;
exports.isPathsElement = _predicates.isPathsElement;
exports.isPathItemElement = _predicates.isPathItemElement;
exports.isOperationElement = _predicates.isOperationElement;
exports.isExternalDocumentationElement = _predicates.isExternalDocumentationElement;
exports.isParameterElement = _predicates.isParameterElement;
exports.isItemsElement = _predicates.isItemsElement;
exports.isResponsesElement = _predicates.isResponsesElement;
exports.isResponseElement = _predicates.isResponseElement;
exports.isHeadersElement = _predicates.isHeadersElement;
exports.isExampleElement = _predicates.isExampleElement;
exports.isHeaderElement = _predicates.isHeaderElement;
exports.isTagElement = _predicates.isTagElement;
exports.isReferenceElement = _predicates.isReferenceElement;
exports.isSchemaElement = _predicates.isSchemaElement;
exports.isXmlElement = _predicates.isXmlElement;
exports.isDefinitionsElement = _predicates.isDefinitionsElement;
exports.isParametersDefinitionsElement = _predicates.isParametersDefinitionsElement;
exports.isResponsesDefinitionsElement = _predicates.isResponsesDefinitionsElement;
exports.isSecurityDefinitionsElement = _predicates.isSecurityDefinitionsElement;
exports.isSecuritySchemeElement = _predicates.isSecuritySchemeElement;
exports.isScopesElement = _predicates.isScopesElement;
exports.isSecurityRequirementElement = _predicates.isSecurityRequirementElement;
var _predicates2 = require("./refractor/predicates.cjs");
exports.isReferenceLikeElement = _predicates2.isReferenceLikeElement;
exports.isSwaggerExtension = _predicates2.isSwaggerExtension;
var _AlternatingVisitor = _interopRequireDefault(require("./refractor/visitors/generics/AlternatingVisitor.cjs"));
exports.AlternatingVisitor = _AlternatingVisitor.default;
var _FixedFieldsVisitor = _interopRequireDefault(require("./refractor/visitors/generics/FixedFieldsVisitor.cjs"));
exports.FixedFieldsVisitor = _FixedFieldsVisitor.default;
var _MapVisitor = _interopRequireDefault(require("./refractor/visitors/generics/MapVisitor.cjs"));
exports.MapVisitor = _MapVisitor.default;
var _MixedFieldsVisitor = _interopRequireDefault(require("./refractor/visitors/generics/MixedFieldsVisitor.cjs"));
exports.MixedFieldsVisitor = _MixedFieldsVisitor.default;
var _PatternedFieldsVisitor = _interopRequireDefault(require("./refractor/visitors/generics/PatternedFieldsVisitor.cjs"));
exports.PatternedFieldsVisitor = _PatternedFieldsVisitor.default;
var _FallbackVisitor = _interopRequireDefault(require("./refractor/visitors/FallbackVisitor.cjs"));
exports.FallbackVisitor = _FallbackVisitor.default;
var _SpecificationVisitor = _interopRequireDefault(require("./refractor/visitors/SpecificationVisitor.cjs"));
exports.SpecificationVisitor = _SpecificationVisitor.default;
var _Visitor = _interopRequireDefault(require("./refractor/visitors/Visitor.cjs"));
exports.Visitor = _Visitor.default;
var _visitor = require("./traversal/visitor.cjs");
exports.keyMap = _visitor.keyMap;
exports.getNodeType = _visitor.getNodeType;
var _registration = require("./refractor/registration.cjs");
exports.SwaggerElement = _registration.SwaggerElement;
exports.SwaggerVersionElement = _registration.SwaggerVersionElement;
exports.InfoElement = _registration.InfoElement;
exports.ContactElement = _registration.ContactElement;
exports.LicenseElement = _registration.LicenseElement;
exports.PathsElement = _registration.PathsElement;
exports.PathItemElement = _registration.PathItemElement;
exports.OperationElement = _registration.OperationElement;
exports.ExternalDocumentationElement = _registration.ExternalDocumentationElement;
exports.ParameterElement = _registration.ParameterElement;
exports.ItemsElement = _registration.ItemsElement;
exports.ResponsesElement = _registration.ResponsesElement;
exports.ResponseElement = _registration.ResponseElement;
exports.HeadersElement = _registration.HeadersElement;
exports.ExampleElement = _registration.ExampleElement;
exports.HeaderElement = _registration.HeaderElement;
exports.TagElement = _registration.TagElement;
exports.ReferenceElement = _registration.ReferenceElement;
exports.SchemaElement = _registration.SchemaElement;
exports.XmlElement = _registration.XmlElement;
exports.DefinitionsElement = _registration.DefinitionsElement;
exports.ParametersDefinitionsElement = _registration.ParametersDefinitionsElement;
exports.ResponsesDefinitionsElement = _registration.ResponsesDefinitionsElement;
exports.SecurityDefinitionsElement = _registration.SecurityDefinitionsElement;
exports.SecuritySchemeElement = _registration.SecuritySchemeElement;
exports.ScopesElement = _registration.ScopesElement;
exports.SecurityRequirementElement = _registration.SecurityRequirementElement;
var _OperationConsumes = _interopRequireDefault(require("./elements/nces/OperationConsumes.cjs"));
exports.OperationConsumesElement = _OperationConsumes.default;
var _OperationParameters = _interopRequireDefault(require("./elements/nces/OperationParameters.cjs"));
exports.OperationParametersElement = _OperationParameters.default;
var _OperationProduces = _interopRequireDefault(require("./elements/nces/OperationProduces.cjs"));
exports.OperationProducesElement = _OperationProduces.default;
var _OperationSchemes = _interopRequireDefault(require("./elements/nces/OperationSchemes.cjs"));
exports.OperationSchemesElement = _OperationSchemes.default;
var _OperationSecurity = _interopRequireDefault(require("./elements/nces/OperationSecurity.cjs"));
exports.OperationSecurityElement = _OperationSecurity.default;
var _OperationTags = _interopRequireDefault(require("./elements/nces/OperationTags.cjs"));
exports.OperationTagsElement = _OperationTags.default;
var _PathItemParameters = _interopRequireDefault(require("./elements/nces/PathItemParameters.cjs"));
exports.PathItemParametersElement = _PathItemParameters.default;
var _SwaggerSchemes = _interopRequireDefault(require("./elements/nces/SwaggerSchemes.cjs"));
exports.SwaggerSchemesElement = _SwaggerSchemes.default;
var _SwaggerConsumes = _interopRequireDefault(require("./elements/nces/SwaggerConsumes.cjs"));
exports.SwaggerConsumesElement = _SwaggerConsumes.default;
var _SwaggerProduces = _interopRequireDefault(require("./elements/nces/SwaggerProduces.cjs"));
exports.SwaggerProducesElement = _SwaggerProduces.default;
var _SwaggerSecurity = _interopRequireDefault(require("./elements/nces/SwaggerSecurity.cjs"));
exports.SwaggerSecurityElement = _SwaggerSecurity.default;
var _SwaggerTags = _interopRequireDefault(require("./elements/nces/SwaggerTags.cjs"));
exports.SwaggerTagsElement = _SwaggerTags.default;
