// @backstage/plugin-scaffolder-backend/dist/scaffolder/tasks/logger.cjs.js
'use strict';

var Transport = require('winston-transport');
var winston = require('winston');
var tripleBeam = require('triple-beam');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var Transport__default = /*#__PURE__*/_interopDefaultCompat(Transport);

const escapeRegExp = (text) => {
  return text.replace(/[.*+?^${}(\)|[\]\\]/g, "\\$&");
};
class BackstageLoggerTransport extends Transport__default.default {
  constructor(backstageLogger, taskContext, stepId, opts) {
    super(opts);
    this.backstageLogger = backstageLogger;
    this.taskContext = taskContext;
    this.stepId = stepId;
  }
  log(info, callback) {
    if (typeof info !== "object" || info === null) {
      callback();
      return;
    }
    const message = info[tripleBeam.MESSAGE];
    const level = info[tripleBeam.LEVEL];
    const splat = info[tripleBeam.SPLAT];
    switch (level) {
      case "error":
        this.backstageLogger.error(String(message), ...splat);
        break;
      case "warn":
        this.backstageLogger.warn(String(message), ...splat);
        break;
      case "info":
        this.backstageLogger.info(String(message), ...splat);
        break;
      case "debug":
        this.backstageLogger.debug(String(message), ...splat);
        break;
      default:
        this.backstageLogger.info(String(message), ...splat);
    }
    this.taskContext.emitLog(message, { stepId: this.stepId });
    callback();
  }
}
class WinstonLogger {
  #winston;
  #addRedactions;
  /**
   * Creates a {@link WinstonLogger} instance.
   */
  static create(options) {
    const redacter = WinstonLogger.redacter();
    let logger = winston.createLogger({
      level: options.level,
      format: winston.format.combine(options.format, redacter.format),
      transports: options.transports ?? new winston.transports.Console()
    });
    if (options.meta) {
      logger = logger.child(options.meta);
    }
    return new WinstonLogger(logger, redacter.add);
  }
  /**
   * Creates a winston log formatter for redacting secrets.
   */
  static redacter() {
    const redactionSet = /* @__PURE__ */ new Set();
    let redactionPattern = void 0;
    return {
      format: winston.format((obj) => {
        if (!redactionPattern || !obj) {
          return obj;
        }
        obj[tripleBeam.MESSAGE] = obj[tripleBeam.MESSAGE]?.replace?.(redactionPattern, "***");
        return obj;
      })(),
      add(newRedactions) {
        let added = 0;
        for (const redactionToTrim of newRedactions) {
          const redaction = redactionToTrim.trim();
          if (redaction.length <= 1) {
            continue;
          }
          if (!redactionSet.has(redaction)) {
            redactionSet.add(redaction);
            added += 1;
          }
        }
        if (added > 0) {
          const redactions = Array.from(redactionSet).map((r) => escapeRegExp(r)).join("|");
          redactionPattern = new RegExp(`(${redactions})`, "g");
        }
      }
    };
  }
  /**
   * Creates a pretty printed winston log formatter.
   */
  static colorFormat() {
    const colorizer = winston.format.colorize();
    return winston.format.combine(
      winston.format.timestamp(),
      winston.format.colorize({
        colors: {
          timestamp: "dim",
          prefix: "blue",
          field: "cyan",
          debug: "grey"
        }
      }),
      winston.format.printf((info) => {
        const { timestamp, plugin, service } = info;
        const message = info[tripleBeam.MESSAGE];
        const level = info[tripleBeam.LEVEL];
        const fields = info[tripleBeam.SPLAT];
        const prefix = plugin || service;
        const timestampColor = colorizer.colorize("timestamp", timestamp);
        const prefixColor = colorizer.colorize("prefix", prefix);
        const extraFields = Object.entries(fields).map(
          ([key, value]) => `${colorizer.colorize("field", `${key}`)}=${value}`
        ).join(" ");
        return `${timestampColor} ${prefixColor} ${level} ${message} ${extraFields}`;
      })
    );
  }
  constructor(winston, addRedactions) {
    this.#winston = winston;
    this.#addRedactions = addRedactions;
  }
  error(message, meta) {
    this.#winston.error(message, meta);
  }
  warn(message, meta) {
    this.#winston.warn(message, meta);
  }
  info(message, meta) {
    this.#winston.info(message, meta);
  }
  debug(message, meta) {
    this.#winston.debug(message, meta);
  }
  child(meta) {
    return new WinstonLogger(this.#winston.child(meta));
  }
  addRedactions(redactions) {
    this.#addRedactions?.(redactions);
  }
}

exports.BackstageLoggerTransport = BackstageLoggerTransport;
exports.WinstonLogger = WinstonLogger;
