// @backstage/plugin-search/dist/components/SidebarSearchModal/SidebarSearchModal.esm.js
import React from 'react';
import SearchIcon from '@material-ui/icons/Search';
import { SidebarItem } from '@backstage/core-components';
import { SearchModal } from '../SearchModal/SearchModal.esm.js';
import { SearchModalProvider, useSearchModal } from '../SearchModal/useSearchModal.esm.js';

const SidebarSearchModalContent = (props) => {
  const { state, toggleModal } = useSearchModal();
  const Icon = props.icon ? props.icon : SearchIcon;
  return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(
    SidebarItem,
    {
      className: "search-icon",
      icon: Icon,
      text: "Search",
      onClick: toggleModal
    }
  ), /* @__PURE__ */ React.createElement(
    SearchModal,
    {
      ...state,
      toggleModal,
      resultItemComponents: props.resultItemComponents,
      children: props.children
    }
  ));
};
const SidebarSearchModal = (props) => {
  return /* @__PURE__ */ React.createElement(SearchModalProvider, null, /* @__PURE__ */ React.createElement(SidebarSearchModalContent, { ...props }));
};

export { SidebarSearchModal };
