// @backstage/plugin-scaffolder/dist/components/fields/EntityNamePicker/EntityNamePicker.esm.js
import React from 'react';
import TextField from '@material-ui/core/TextField';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../../translation.esm.js';
export { EntityNamePickerSchema } from './schema.esm.js';

const EntityNamePicker = (props) => {
  const { t } = useTranslationRef(scaffolderTranslationRef);
  const {
    onChange,
    required,
    schema: {
      title = t("fields.entityNamePicker.title"),
      description = t("fields.entityNamePicker.description")
    },
    rawErrors,
    formData,
    uiSchema: { "ui:autofocus": autoFocus },
    idSchema,
    placeholder
  } = props;
  return /* @__PURE__ */ React.createElement(
    TextField,
    {
      id: idSchema?.$id,
      label: title,
      placeholder,
      helperText: description,
      required,
      value: formData ?? "",
      onChange: ({ target: { value } }) => onChange(value),
      margin: "normal",
      error: rawErrors?.length > 0 && !formData,
      inputProps: { autoFocus },
      FormHelperTextProps: { margin: "dense", style: { marginLeft: 0 } }
    }
  );
};

export { EntityNamePicker };
