// @backstage/plugin-search-backend-module-catalog/dist/collators/config.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var errors = require('@backstage/errors');

const configKey = "search.collators.catalog";
const defaults = {
  schedule: {
    frequency: { minutes: 10 },
    timeout: { minutes: 15 },
    initialDelay: { seconds: 3 }
  },
  collatorOptions: {
    locationTemplate: "/catalog/:namespace/:kind/:name",
    filter: void 0,
    batchSize: 500
  }
};
function readScheduleConfigOptions(configRoot) {
  let schedule = void 0;
  const config = configRoot.getOptionalConfig(configKey);
  if (config) {
    const scheduleConfig = config.getOptionalConfig("schedule");
    if (scheduleConfig) {
      try {
        schedule = backendPluginApi.readSchedulerServiceTaskScheduleDefinitionFromConfig(scheduleConfig);
      } catch (error) {
        throw new errors.InputError(`Invalid schedule at ${configKey}, ${error}`);
      }
    }
  }
  return schedule ?? defaults.schedule;
}
function readCollatorConfigOptions(configRoot) {
  const config = configRoot.getOptionalConfig(configKey);
  if (!config) {
    return defaults.collatorOptions;
  }
  return {
    locationTemplate: config.getOptionalString("locationTemplate") ?? defaults.collatorOptions.locationTemplate,
    filter: config.getOptional("filter") ?? defaults.collatorOptions.filter,
    batchSize: config.getOptionalNumber("batchSize") ?? defaults.collatorOptions.batchSize
  };
}

exports.defaults = defaults;
exports.readCollatorConfigOptions = readCollatorConfigOptions;
exports.readScheduleConfigOptions = readScheduleConfigOptions;
