// @backstage/plugin-search-backend-module-catalog/dist/collators/defaultCatalogCollatorEntityTransformer.cjs.js
'use strict';

var catalogModel = require('@backstage/catalog-model');

const getDocumentText = (entity) => {
  const documentTexts = [];
  if (entity.metadata.description) {
    documentTexts.push(entity.metadata.description);
  }
  if (catalogModel.isUserEntity(entity) || catalogModel.isGroupEntity(entity)) {
    if (entity.spec?.profile?.displayName) {
      documentTexts.push(entity.spec.profile.displayName);
    }
  }
  if (catalogModel.isUserEntity(entity)) {
    if (entity.spec?.profile?.email) {
      documentTexts.push(entity.spec.profile.email);
    }
  }
  return documentTexts.join(" : ");
};
const defaultCatalogCollatorEntityTransformer = (entity) => {
  return {
    title: entity.metadata.title ?? entity.metadata.name,
    text: getDocumentText(entity),
    componentType: entity.spec?.type?.toString() || "other",
    type: entity.spec?.type?.toString() || "other",
    namespace: entity.metadata.namespace || "default",
    kind: entity.kind,
    lifecycle: entity.spec?.lifecycle || "",
    owner: entity.spec?.owner || ""
  };
};

exports.defaultCatalogCollatorEntityTransformer = defaultCatalogCollatorEntityTransformer;
