// @backstage/plugin-search/dist/components/SearchModal/SearchModal.esm.js
import { useContent } from '@backstage/core-components';
import { useRouteRef } from '@backstage/core-plugin-api';
import { SearchContextProvider, SearchBar, SearchResult, SearchResultPager } from '@backstage/plugin-search-react';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import Divider from '@material-ui/core/Divider';
import Grid from '@material-ui/core/Grid';
import { makeStyles, useTheme } from '@material-ui/core/styles';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import ArrowForwardIcon from '@material-ui/icons/ArrowForward';
import CloseIcon from '@material-ui/icons/Close';
import React, { useRef, useEffect, useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import { rootRouteRef } from '../../plugin.esm.js';

const useStyles = makeStyles((theme) => ({
  dialogTitle: {
    gap: theme.spacing(1),
    display: "grid",
    alignItems: "center",
    gridTemplateColumns: "1fr auto",
    "&> button": {
      marginTop: theme.spacing(1)
    }
  },
  input: {
    flex: 1
  },
  button: {
    "&:hover": {
      background: "none"
    }
  },
  // Reduces default height of the modal, keeping a gap of 128px between the top and bottom of the page.
  paperFullWidth: { height: "calc(100% - 128px)" },
  dialogActionsContainer: { padding: theme.spacing(1, 3) },
  viewResultsLink: { verticalAlign: "0.5em" }
}));
const Modal = ({
  toggleModal,
  resultItemComponents
}) => {
  const classes = useStyles();
  const navigate = useNavigate();
  const { transitions } = useTheme();
  const { focusContent } = useContent();
  const searchRootRoute = useRouteRef(rootRouteRef)();
  const searchBarRef = useRef(null);
  useEffect(() => {
    searchBarRef?.current?.focus();
  });
  const handleSearchResultClick = useCallback(() => {
    setTimeout(focusContent, transitions.duration.leavingScreen);
  }, [focusContent, transitions]);
  const handleSearchBarSubmit = useCallback(() => {
    const query = searchBarRef.current?.value ?? "";
    navigate(`${searchRootRoute}?query=${query}`);
    handleSearchResultClick();
  }, [navigate, handleSearchResultClick, searchRootRoute]);
  return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(DialogTitle, null, /* @__PURE__ */ React.createElement(Box, { className: classes.dialogTitle }, /* @__PURE__ */ React.createElement(
    SearchBar,
    {
      className: classes.input,
      inputProps: { ref: searchBarRef },
      onSubmit: handleSearchBarSubmit
    }
  ), /* @__PURE__ */ React.createElement(IconButton, { "aria-label": "close", onClick: toggleModal }, /* @__PURE__ */ React.createElement(CloseIcon, null)))), /* @__PURE__ */ React.createElement(DialogContent, null, /* @__PURE__ */ React.createElement(
    Grid,
    {
      container: true,
      direction: "row-reverse",
      justifyContent: "flex-start",
      alignItems: "center"
    },
    /* @__PURE__ */ React.createElement(Grid, { item: true }, /* @__PURE__ */ React.createElement(
      Button,
      {
        className: classes.button,
        color: "primary",
        endIcon: /* @__PURE__ */ React.createElement(ArrowForwardIcon, null),
        onClick: handleSearchBarSubmit,
        disableRipple: true
      },
      "View Full Results"
    ))
  ), /* @__PURE__ */ React.createElement(Divider, null), /* @__PURE__ */ React.createElement(
    SearchResult,
    {
      onClick: handleSearchResultClick,
      onKeyDown: handleSearchResultClick
    },
    resultItemComponents
  )), /* @__PURE__ */ React.createElement(DialogActions, { className: classes.dialogActionsContainer }, /* @__PURE__ */ React.createElement(Grid, { container: true, direction: "row" }, /* @__PURE__ */ React.createElement(Grid, { item: true, xs: 12 }, /* @__PURE__ */ React.createElement(SearchResultPager, null)))));
};
const SearchModal = (props) => {
  const {
    open = true,
    hidden,
    toggleModal,
    children,
    resultItemComponents
  } = props;
  const classes = useStyles();
  return /* @__PURE__ */ React.createElement(
    Dialog,
    {
      classes: {
        paperFullWidth: classes.paperFullWidth
      },
      onClose: toggleModal,
      "aria-label": "Search Modal",
      "aria-modal": "true",
      fullWidth: true,
      maxWidth: "lg",
      open,
      hidden
    },
    open && /* @__PURE__ */ React.createElement(SearchContextProvider, { inheritParentContextIfAvailable: true }, (children && children({
      toggleModal,
      resultItemComponents: resultItemComponents || []
    })) ?? /* @__PURE__ */ React.createElement(
      Modal,
      {
        toggleModal,
        resultItemComponents
      }
    ))
  );
};

export { Modal, SearchModal };
