// @backstage/plugin-scaffolder-react/dist/next/components/TaskSteps/TaskSteps.esm.js
import React from 'react';
import MuiStepper from '@material-ui/core/Stepper';
import MuiStep from '@material-ui/core/Step';
import MuiStepButton from '@material-ui/core/StepButton';
import MuiStepLabel from '@material-ui/core/StepLabel';
import Box from '@material-ui/core/Box';
import Paper from '@material-ui/core/Paper';
import { StepIcon } from './StepIcon.esm.js';
import { StepTime } from './StepTime.esm.js';
import { TaskBorder } from './TaskBorder.esm.js';

const TaskSteps = (props) => {
  return /* @__PURE__ */ React.createElement(Paper, { style: { position: "relative", overflow: "hidden" } }, /* @__PURE__ */ React.createElement(
    TaskBorder,
    {
      isComplete: props.isComplete ?? false,
      isError: props.isError ?? false
    }
  ), /* @__PURE__ */ React.createElement(Box, { padding: 2 }, /* @__PURE__ */ React.createElement(
    MuiStepper,
    {
      activeStep: props.activeStep,
      alternativeLabel: true,
      variant: "elevation",
      style: { overflowX: "auto" }
    },
    props.steps.map((step) => {
      const isCompleted = step.status === "completed";
      const isFailed = step.status === "failed";
      const isActive = step.status === "processing";
      const isSkipped = step.status === "skipped";
      const stepIconProps = {
        completed: isCompleted,
        error: isFailed,
        active: isActive,
        skipped: isSkipped
      };
      return /* @__PURE__ */ React.createElement(MuiStep, { key: step.id }, /* @__PURE__ */ React.createElement(MuiStepButton, null, /* @__PURE__ */ React.createElement(
        MuiStepLabel,
        {
          StepIconProps: stepIconProps,
          StepIconComponent: StepIcon,
          "data-testid": "step-label"
        },
        /* @__PURE__ */ React.createElement(Box, null, step.name),
        !isSkipped && /* @__PURE__ */ React.createElement(StepTime, { step })
      )));
    })
  )));
};

export { TaskSteps };
