// @backstage/plugin-catalog-import/dist/components/Buttons/index.esm.js
import { LinkButton } from '@backstage/core-components';
import Button from '@material-ui/core/Button';
import CircularProgress from '@material-ui/core/CircularProgress';
import { makeStyles } from '@material-ui/core/styles';
import React from 'react';

const useStyles = makeStyles((theme) => ({
  wrapper: {
    marginTop: theme.spacing(1),
    marginRight: theme.spacing(1),
    position: "relative"
  },
  buttonProgress: {
    position: "absolute",
    top: "50%",
    left: "50%",
    marginTop: -12,
    marginLeft: -12
  },
  button: {
    marginTop: theme.spacing(1),
    marginRight: theme.spacing(1)
  }
}));
const NextButton = (props) => {
  const { loading, ...buttonProps } = props;
  const classes = useStyles();
  return /* @__PURE__ */ React.createElement("div", { className: classes.wrapper }, /* @__PURE__ */ React.createElement(
    Button,
    {
      color: "primary",
      variant: "contained",
      ...buttonProps,
      disabled: props.disabled || props.loading
    }
  ), props.loading && /* @__PURE__ */ React.createElement(CircularProgress, { size: "1.5rem", className: classes.buttonProgress }), props.loading);
};
const BackButton = (props) => {
  const classes = useStyles();
  return /* @__PURE__ */ React.createElement(Button, { variant: "outlined", className: classes.button, ...props }, props.children || "Back");
};
const ViewComponentButton = (props) => {
  const classes = useStyles();
  return /* @__PURE__ */ React.createElement(
    LinkButton,
    {
      color: "primary",
      variant: "contained",
      className: classes.button,
      ...props
    },
    props.children
  );
};

export { BackButton, NextButton, ViewComponentButton };
