// @backstage/core-components/dist/layout/Page/Page.esm.js
import React__default from 'react';
import { makeStyles, ThemeProvider } from '@material-ui/core/styles';

const useStyles = makeStyles(
  (theme) => ({
    root: {
      display: "grid",
      gridTemplateAreas: "'pageHeader pageHeader pageHeader' 'pageSubheader pageSubheader pageSubheader' 'pageNav pageContent pageSidebar'",
      gridTemplateRows: "max-content auto 1fr",
      gridTemplateColumns: "auto 1fr auto",
      overflowY: "auto",
      height: "100vh",
      [theme.breakpoints.down("xs")]: {
        height: "100%"
      },
      "@media print": {
        display: "block",
        height: "auto",
        overflowY: "inherit"
      }
    }
  }),
  { name: "BackstagePage" }
);
function Page(props) {
  const { themeId, children } = props;
  const classes = useStyles();
  return /* @__PURE__ */ React__default.createElement(
    ThemeProvider,
    {
      theme: (baseTheme) => ({
        ...baseTheme,
        page: baseTheme.getPageTheme({ themeId })
      })
    },
    /* @__PURE__ */ React__default.createElement("main", { className: classes.root }, children)
  );
}

export { Page };
