// @backstage/plugin-search-backend-module-catalog/dist/module.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendPluginApi = require('@backstage/backend-plugin-api');
var alpha$1 = require('@backstage/plugin-catalog-node/alpha');
var pluginSearchBackendModuleCatalog = require('@backstage/plugin-search-backend-module-catalog');
var alpha = require('@backstage/plugin-search-backend-node/alpha');
var config = require('./collators/config.cjs.js');

const catalogCollatorExtensionPoint = backendPluginApi.createExtensionPoint({
  id: "search.catalogCollator.extension"
});
var feature = backendPluginApi.createBackendModule({
  pluginId: "search",
  moduleId: "catalog-collator",
  register(env) {
    let entityTransformer;
    env.registerExtensionPoint(catalogCollatorExtensionPoint, {
      setEntityTransformer(transformer) {
        if (entityTransformer) {
          throw new Error("setEntityTransformer can only be called once");
        }
        entityTransformer = transformer;
      }
    });
    env.registerInit({
      deps: {
        auth: backendPluginApi.coreServices.auth,
        config: backendPluginApi.coreServices.rootConfig,
        discovery: backendPluginApi.coreServices.discovery,
        scheduler: backendPluginApi.coreServices.scheduler,
        indexRegistry: alpha.searchIndexRegistryExtensionPoint,
        catalog: alpha$1.catalogServiceRef
      },
      async init({
        auth,
        config: config$1,
        discovery,
        scheduler,
        indexRegistry,
        catalog
      }) {
        indexRegistry.addCollator({
          schedule: scheduler.createScheduledTaskRunner(
            config.readScheduleConfigOptions(config$1)
          ),
          factory: pluginSearchBackendModuleCatalog.DefaultCatalogCollatorFactory.fromConfig(config$1, {
            auth,
            entityTransformer,
            discovery,
            catalogClient: catalog
          })
        });
      }
    });
  }
});

exports.catalogCollatorExtensionPoint = catalogCollatorExtensionPoint;
exports.default = feature;
