// @backstage/plugin-search/dist/components/HomePageComponent/HomePageSearchBar.esm.js
import React, { useState, useRef, useCallback } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { SearchBarBase } from '@backstage/plugin-search-react';
import { useNavigateToQuery } from '../util.esm.js';

const useStyles = makeStyles({
  searchBarRoot: {
    fontSize: "1.5em"
  },
  searchBarOutline: {
    border: "1px solid #555",
    borderRadius: "6px"
  }
});
const HomePageSearchBar = (props) => {
  const classes = useStyles(props);
  const [query, setQuery] = useState("");
  const ref = useRef(null);
  const handleSearch = useNavigateToQuery();
  const handleSubmit = useCallback(() => {
    handleSearch({ query: ref.current?.value ?? "" });
  }, [handleSearch]);
  return /* @__PURE__ */ React.createElement(
    SearchBarBase,
    {
      value: query,
      onSubmit: handleSubmit,
      onChange: setQuery,
      inputProps: { ref },
      InputProps: {
        ...props.InputProps,
        classes: {
          root: classes.searchBarRoot,
          notchedOutline: classes.searchBarOutline,
          ...props.InputProps?.classes
        }
      },
      ...props
    }
  );
};

export { HomePageSearchBar };
