// @swagger-api/apidom-ns-asyncapi-2/src/refractor/registration.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
exports.__esModule = true;
var _index = require("./index.cjs");
var _AsyncApi = _interopRequireDefault(require("../elements/AsyncApi2.cjs"));
exports.AsyncApi2Element = _AsyncApi.default;
var _AsyncApiVersion = _interopRequireDefault(require("../elements/AsyncApiVersion.cjs"));
exports.AsyncApiVersionElement = _AsyncApiVersion.default;
var _ChannelBindings = _interopRequireDefault(require("../elements/ChannelBindings.cjs"));
exports.ChannelBindingsElement = _ChannelBindings.default;
var _ChannelItem = _interopRequireDefault(require("../elements/ChannelItem.cjs"));
exports.ChannelItemElement = _ChannelItem.default;
var _Channels = _interopRequireDefault(require("../elements/Channels.cjs"));
exports.ChannelsElement = _Channels.default;
var _Components = _interopRequireDefault(require("../elements/Components.cjs"));
exports.ComponentsElement = _Components.default;
var _Contact = _interopRequireDefault(require("../elements/Contact.cjs"));
exports.ContactElement = _Contact.default;
var _CorrelationID = _interopRequireDefault(require("../elements/CorrelationID.cjs"));
exports.CorrelationIDElement = _CorrelationID.default;
var _DefaultContentType = _interopRequireDefault(require("../elements/DefaultContentType.cjs"));
exports.DefaultContentTypeElement = _DefaultContentType.default;
var _ExternalDocumentation = _interopRequireDefault(require("../elements/ExternalDocumentation.cjs"));
exports.ExternalDocumentationElement = _ExternalDocumentation.default;
var _Identifier = _interopRequireDefault(require("../elements/Identifier.cjs"));
exports.IdentifierElement = _Identifier.default;
var _Info = _interopRequireDefault(require("../elements/Info.cjs"));
exports.InfoElement = _Info.default;
var _License = _interopRequireDefault(require("../elements/License.cjs"));
exports.LicenseElement = _License.default;
var _Message = _interopRequireDefault(require("../elements/Message.cjs"));
exports.MessageElement = _Message.default;
var _MessageBindings = _interopRequireDefault(require("../elements/MessageBindings.cjs"));
exports.MessageBindingsElement = _MessageBindings.default;
var _MessageExample = _interopRequireDefault(require("../elements/MessageExample.cjs"));
exports.MessageExampleElement = _MessageExample.default;
var _MessageTrait = _interopRequireDefault(require("../elements/MessageTrait.cjs"));
exports.MessageTraitElement = _MessageTrait.default;
var _OAuthFlow = _interopRequireDefault(require("../elements/OAuthFlow.cjs"));
exports.OAuthFlowElement = _OAuthFlow.default;
var _OAuthFlows = _interopRequireDefault(require("../elements/OAuthFlows.cjs"));
exports.OAuthFlowsElement = _OAuthFlows.default;
var _Operation = _interopRequireDefault(require("../elements/Operation.cjs"));
exports.OperationElement = _Operation.default;
var _OperationBindings = _interopRequireDefault(require("../elements/OperationBindings.cjs"));
exports.OperationBindingsElement = _OperationBindings.default;
var _OperationTrait = _interopRequireDefault(require("../elements/OperationTrait.cjs"));
exports.OperationTraitElement = _OperationTrait.default;
var _Parameter = _interopRequireDefault(require("../elements/Parameter.cjs"));
exports.ParameterElement = _Parameter.default;
var _Parameters = _interopRequireDefault(require("../elements/Parameters.cjs"));
exports.ParametersElement = _Parameters.default;
var _Reference = _interopRequireDefault(require("../elements/Reference.cjs"));
exports.ReferenceElement = _Reference.default;
var _Schema = _interopRequireDefault(require("../elements/Schema.cjs"));
exports.SchemaElement = _Schema.default;
var _SecurityRequirement = _interopRequireDefault(require("../elements/SecurityRequirement.cjs"));
exports.SecurityRequirementElement = _SecurityRequirement.default;
var _SecurityScheme = _interopRequireDefault(require("../elements/SecurityScheme.cjs"));
exports.SecuritySchemeElement = _SecurityScheme.default;
var _Server = _interopRequireDefault(require("../elements/Server.cjs"));
exports.ServerElement = _Server.default;
var _ServerBindings = _interopRequireDefault(require("../elements/ServerBindings.cjs"));
exports.ServerBindingsElement = _ServerBindings.default;
var _Servers = _interopRequireDefault(require("../elements/Servers.cjs"));
exports.ServersElement = _Servers.default;
var _ServerVariable = _interopRequireDefault(require("../elements/ServerVariable.cjs"));
exports.ServerVariableElement = _ServerVariable.default;
var _Tag = _interopRequireDefault(require("../elements/Tag.cjs"));
exports.TagElement = _Tag.default;
var _Tags = _interopRequireDefault(require("../elements/Tags.cjs"));
exports.TagsElement = _Tags.default;
var _AmqpChannelBinding = _interopRequireDefault(require("../elements/bindings/amqp/AmqpChannelBinding.cjs"));
exports.AmqpChannelBindingElement = _AmqpChannelBinding.default;
var _AmqpMessageBinding = _interopRequireDefault(require("../elements/bindings/amqp/AmqpMessageBinding.cjs"));
exports.AmqpMessageBindingElement = _AmqpMessageBinding.default;
var _AmqpOperationBinding = _interopRequireDefault(require("../elements/bindings/amqp/AmqpOperationBinding.cjs"));
exports.AmqpOperationBindingElement = _AmqpOperationBinding.default;
var _AmqpServerBinding = _interopRequireDefault(require("../elements/bindings/amqp/AmqpServerBinding.cjs"));
exports.AmqpServerBindingElement = _AmqpServerBinding.default;
var _Amqp1ChannelBinding = _interopRequireDefault(require("../elements/bindings/amqp1/Amqp1ChannelBinding.cjs"));
exports.Amqp1ChannelBindingElement = _Amqp1ChannelBinding.default;
var _Amqp1MessageBinding = _interopRequireDefault(require("../elements/bindings/amqp1/Amqp1MessageBinding.cjs"));
exports.Amqp1MessageBindingElement = _Amqp1MessageBinding.default;
var _Amqp1OperationBinding = _interopRequireDefault(require("../elements/bindings/amqp1/Amqp1OperationBinding.cjs"));
exports.Amqp1OperationBindingElement = _Amqp1OperationBinding.default;
var _Amqp1ServerBinding = _interopRequireDefault(require("../elements/bindings/amqp1/Amqp1ServerBinding.cjs"));
exports.Amqp1ServerBindingElement = _Amqp1ServerBinding.default;
var _AnypointmqChannelBinding = _interopRequireDefault(require("../elements/bindings/anypointmq/AnypointmqChannelBinding.cjs"));
exports.AnypointmqChannelBindingElement = _AnypointmqChannelBinding.default;
var _AnypointmqMessageBinding = _interopRequireDefault(require("../elements/bindings/anypointmq/AnypointmqMessageBinding.cjs"));
exports.AnypointmqMessageBindingElement = _AnypointmqMessageBinding.default;
var _AnypointmqOperationBinding = _interopRequireDefault(require("../elements/bindings/anypointmq/AnypointmqOperationBinding.cjs"));
exports.AnypointmqOperationBindingElement = _AnypointmqOperationBinding.default;
var _AnypointmqServerBinding = _interopRequireDefault(require("../elements/bindings/anypointmq/AnypointmqServerBinding.cjs"));
exports.AnypointmqServerBindingElement = _AnypointmqServerBinding.default;
var _HttpChannelBinding = _interopRequireDefault(require("../elements/bindings/http/HttpChannelBinding.cjs"));
exports.HttpChannelBindingElement = _HttpChannelBinding.default;
var _HttpMessageBinding = _interopRequireDefault(require("../elements/bindings/http/HttpMessageBinding.cjs"));
exports.HttpMessageBindingElement = _HttpMessageBinding.default;
var _HttpOperationBinding = _interopRequireDefault(require("../elements/bindings/http/HttpOperationBinding.cjs"));
exports.HttpOperationBindingElement = _HttpOperationBinding.default;
var _HttpServerBinding = _interopRequireDefault(require("../elements/bindings/http/HttpServerBinding.cjs"));
exports.HttpServerBindingElement = _HttpServerBinding.default;
var _GooglepubsubChannelBinding = _interopRequireDefault(require("../elements/bindings/googlepubsub/GooglepubsubChannelBinding.cjs"));
exports.GooglepubsubChannelBindingElement = _GooglepubsubChannelBinding.default;
var _GooglepubsubMessageBinding = _interopRequireDefault(require("../elements/bindings/googlepubsub/GooglepubsubMessageBinding.cjs"));
exports.GooglepubsubMessageBindingElement = _GooglepubsubMessageBinding.default;
var _GooglepubsubOperationBinding = _interopRequireDefault(require("../elements/bindings/googlepubsub/GooglepubsubOperationBinding.cjs"));
exports.GooglepubsubOperationBindingElement = _GooglepubsubOperationBinding.default;
var _GooglepubsubServerBinding = _interopRequireDefault(require("../elements/bindings/googlepubsub/GooglepubsubServerBinding.cjs"));
exports.GooglepubsubServerBindingElement = _GooglepubsubServerBinding.default;
var _IbmmqChannelBinding = _interopRequireDefault(require("../elements/bindings/ibmmq/IbmmqChannelBinding.cjs"));
exports.IbmmqChannelBindingElement = _IbmmqChannelBinding.default;
var _IbmmqMessageBinding = _interopRequireDefault(require("../elements/bindings/ibmmq/IbmmqMessageBinding.cjs"));
exports.IbmmqMessageBindingElement = _IbmmqMessageBinding.default;
var _IbmmqOperationBinding = _interopRequireDefault(require("../elements/bindings/ibmmq/IbmmqOperationBinding.cjs"));
exports.IbmmqOperationBindingElement = _IbmmqOperationBinding.default;
var _IbmmqServerBinding = _interopRequireDefault(require("../elements/bindings/ibmmq/IbmmqServerBinding.cjs"));
exports.IbmmqServerBindingElement = _IbmmqServerBinding.default;
var _JmsChannelBinding = _interopRequireDefault(require("../elements/bindings/jms/JmsChannelBinding.cjs"));
exports.JmsChannelBindingElement = _JmsChannelBinding.default;
var _JmsMessageBinding = _interopRequireDefault(require("../elements/bindings/jms/JmsMessageBinding.cjs"));
exports.JmsMessageBindingElement = _JmsMessageBinding.default;
var _JmsOperationBinding = _interopRequireDefault(require("../elements/bindings/jms/JmsOperationBinding.cjs"));
exports.JmsOperationBindingElement = _JmsOperationBinding.default;
var _JmsServerBinding = _interopRequireDefault(require("../elements/bindings/jms/JmsServerBinding.cjs"));
exports.JmsServerBindingElement = _JmsServerBinding.default;
var _KafkaChannelBinding = _interopRequireDefault(require("../elements/bindings/kafka/KafkaChannelBinding.cjs"));
exports.KafkaChannelBindingElement = _KafkaChannelBinding.default;
var _KafkaMessageBinding = _interopRequireDefault(require("../elements/bindings/kafka/KafkaMessageBinding.cjs"));
exports.KafkaMessageBindingElement = _KafkaMessageBinding.default;
var _KafkaOperationBinding = _interopRequireDefault(require("../elements/bindings/kafka/KafkaOperationBinding.cjs"));
exports.KafkaOperationBindingElement = _KafkaOperationBinding.default;
var _KafkaServerBinding = _interopRequireDefault(require("../elements/bindings/kafka/KafkaServerBinding.cjs"));
exports.KafkaServerBindingElement = _KafkaServerBinding.default;
var _MercureChannelBinding = _interopRequireDefault(require("../elements/bindings/mercure/MercureChannelBinding.cjs"));
exports.MercureChannelBindingElement = _MercureChannelBinding.default;
var _MercureMessageBinding = _interopRequireDefault(require("../elements/bindings/mercure/MercureMessageBinding.cjs"));
exports.MercureMessageBindingElement = _MercureMessageBinding.default;
var _MercureOperationBinding = _interopRequireDefault(require("../elements/bindings/mercure/MercureOperationBinding.cjs"));
exports.MercureOperationBindingElement = _MercureOperationBinding.default;
var _MercureServerBinding = _interopRequireDefault(require("../elements/bindings/mercure/MercureServerBinding.cjs"));
exports.MercureServerBindingElement = _MercureServerBinding.default;
var _MqttChannelBinding = _interopRequireDefault(require("../elements/bindings/mqtt/MqttChannelBinding.cjs"));
exports.MqttChannelBindingElement = _MqttChannelBinding.default;
var _MqttMessageBinding = _interopRequireDefault(require("../elements/bindings/mqtt/MqttMessageBinding.cjs"));
exports.MqttMessageBindingElement = _MqttMessageBinding.default;
var _MqttOperationBinding = _interopRequireDefault(require("../elements/bindings/mqtt/MqttOperationBinding.cjs"));
exports.MqttOperationBindingElement = _MqttOperationBinding.default;
var _MqttServerBinding = _interopRequireDefault(require("../elements/bindings/mqtt/MqttServerBinding.cjs"));
exports.MqttServerBindingElement = _MqttServerBinding.default;
var _Mqtt5ChannelBinding = _interopRequireDefault(require("../elements/bindings/mqtt5/Mqtt5ChannelBinding.cjs"));
exports.Mqtt5ChannelBindingElement = _Mqtt5ChannelBinding.default;
var _Mqtt5MessageBinding = _interopRequireDefault(require("../elements/bindings/mqtt5/Mqtt5MessageBinding.cjs"));
exports.Mqtt5MessageBindingElement = _Mqtt5MessageBinding.default;
var _Mqtt5OperationBinding = _interopRequireDefault(require("../elements/bindings/mqtt5/Mqtt5OperationBinding.cjs"));
exports.Mqtt5OperationBindingElement = _Mqtt5OperationBinding.default;
var _Mqtt5ServerBinding = _interopRequireDefault(require("../elements/bindings/mqtt5/Mqtt5ServerBinding.cjs"));
exports.Mqtt5ServerBindingElement = _Mqtt5ServerBinding.default;
var _NatsChannelBinding = _interopRequireDefault(require("../elements/bindings/nats/NatsChannelBinding.cjs"));
exports.NatsChannelBindingElement = _NatsChannelBinding.default;
var _NatsMessageBinding = _interopRequireDefault(require("../elements/bindings/nats/NatsMessageBinding.cjs"));
exports.NatsMessageBindingElement = _NatsMessageBinding.default;
var _NatsOperationBinding = _interopRequireDefault(require("../elements/bindings/nats/NatsOperationBinding.cjs"));
exports.NatsOperationBindingElement = _NatsOperationBinding.default;
var _NatsServerBinding = _interopRequireDefault(require("../elements/bindings/nats/NatsServerBinding.cjs"));
exports.NatsServerBindingElement = _NatsServerBinding.default;
var _PulsarChannelBinding = _interopRequireDefault(require("../elements/bindings/pulsar/PulsarChannelBinding.cjs"));
exports.PulsarChannelBindingElement = _PulsarChannelBinding.default;
var _PulsarMessageBinding = _interopRequireDefault(require("../elements/bindings/pulsar/PulsarMessageBinding.cjs"));
exports.PulsarMessageBindingElement = _PulsarMessageBinding.default;
var _PulsarOperationBinding = _interopRequireDefault(require("../elements/bindings/pulsar/PulsarOperationBinding.cjs"));
exports.PulsarOperationBindingElement = _PulsarOperationBinding.default;
var _PulsarServerBinding = _interopRequireDefault(require("../elements/bindings/pulsar/PulsarServerBinding.cjs"));
exports.PulsarServerBindingElement = _PulsarServerBinding.default;
var _RedisChannelBinding = _interopRequireDefault(require("../elements/bindings/redis/RedisChannelBinding.cjs"));
exports.RedisChannelBindingElement = _RedisChannelBinding.default;
var _RedisMessageBinding = _interopRequireDefault(require("../elements/bindings/redis/RedisMessageBinding.cjs"));
exports.RedisMessageBindingElement = _RedisMessageBinding.default;
var _RedisOperationBinding = _interopRequireDefault(require("../elements/bindings/redis/RedisOperationBinding.cjs"));
exports.RedisOperationBindingElement = _RedisOperationBinding.default;
var _RedisServerBinding = _interopRequireDefault(require("../elements/bindings/redis/RedisServerBinding.cjs"));
exports.RedisServerBindingElement = _RedisServerBinding.default;
var _SnsChannelBinding = _interopRequireDefault(require("../elements/bindings/sns/SnsChannelBinding.cjs"));
exports.SnsChannelBindingElement = _SnsChannelBinding.default;
var _SnsMessageBinding = _interopRequireDefault(require("../elements/bindings/sns/SnsMessageBinding.cjs"));
exports.SnsMessageBindingElement = _SnsMessageBinding.default;
var _SnsOperationBinding = _interopRequireDefault(require("../elements/bindings/sns/SnsOperationBinding.cjs"));
exports.SnsOperationBindingElement = _SnsOperationBinding.default;
var _SnsServerBinding = _interopRequireDefault(require("../elements/bindings/sns/SnsServerBinding.cjs"));
exports.SnsServerBindingElement = _SnsServerBinding.default;
var _SolaceChannelBinding = _interopRequireDefault(require("../elements/bindings/solace/SolaceChannelBinding.cjs"));
exports.SolaceChannelBindingElement = _SolaceChannelBinding.default;
var _SolaceMessageBinding = _interopRequireDefault(require("../elements/bindings/solace/SolaceMessageBinding.cjs"));
exports.SolaceMessageBindingElement = _SolaceMessageBinding.default;
var _SolaceOperationBinding = _interopRequireDefault(require("../elements/bindings/solace/SolaceOperationBinding.cjs"));
exports.SolaceOperationBindingElement = _SolaceOperationBinding.default;
var _SolaceServerBinding = _interopRequireDefault(require("../elements/bindings/solace/SolaceServerBinding.cjs"));
exports.SolaceServerBindingElement = _SolaceServerBinding.default;
var _SqsChannelBinding = _interopRequireDefault(require("../elements/bindings/sqs/SqsChannelBinding.cjs"));
exports.SqsChannelBindingElement = _SqsChannelBinding.default;
var _SqsMessageBinding = _interopRequireDefault(require("../elements/bindings/sqs/SqsMessageBinding.cjs"));
exports.SqsMessageBindingElement = _SqsMessageBinding.default;
var _SqsOperationBinding = _interopRequireDefault(require("../elements/bindings/sqs/SqsOperationBinding.cjs"));
exports.SqsOperationBindingElement = _SqsOperationBinding.default;
var _SqsServerBinding = _interopRequireDefault(require("../elements/bindings/sqs/SqsServerBinding.cjs"));
exports.SqsServerBindingElement = _SqsServerBinding.default;
var _StompChannelBinding = _interopRequireDefault(require("../elements/bindings/stomp/StompChannelBinding.cjs"));
exports.StompChannelBindingElement = _StompChannelBinding.default;
var _StompMessageBinding = _interopRequireDefault(require("../elements/bindings/stomp/StompMessageBinding.cjs"));
exports.StompMessageBindingElement = _StompMessageBinding.default;
var _StompOperationBinding = _interopRequireDefault(require("../elements/bindings/stomp/StompOperationBinding.cjs"));
exports.StompOperationBindingElement = _StompOperationBinding.default;
var _StompServerBinding = _interopRequireDefault(require("../elements/bindings/stomp/StompServerBinding.cjs"));
exports.StompServerBindingElement = _StompServerBinding.default;
var _WebSocketChannelBinding = _interopRequireDefault(require("../elements/bindings/ws/WebSocketChannelBinding.cjs"));
exports.WebSocketChannelBindingElement = _WebSocketChannelBinding.default;
var _WebSocketMessageBinding = _interopRequireDefault(require("../elements/bindings/ws/WebSocketMessageBinding.cjs"));
exports.WebSocketMessageBindingElement = _WebSocketMessageBinding.default;
var _WebSocketOperationBinding = _interopRequireDefault(require("../elements/bindings/ws/WebSocketOperationBinding.cjs"));
exports.WebSocketOperationBindingElement = _WebSocketOperationBinding.default;
var _WebSocketServerBinding = _interopRequireDefault(require("../elements/bindings/ws/WebSocketServerBinding.cjs"));
exports.WebSocketServerBindingElement = _WebSocketServerBinding.default;
/**
 * AsyncApi >= 2.0.0 <=2.6.0 specification elements.
 */

/**
 * Binding elements.
 */
// AMQP 0-9-1

// AMQP 1.0

// Anypoint MQ

// HTTP

// Google Cloud Pub/Sub

// IBM MQ

// JMS

// Kafka

// Mercure

// MQTT

// MQTT 5

// NATS

// Pulsar

// Redis

// SNS

// Solace

// SQS

// STOMP

// WebSocket

/**
 * AsyncApi >= 2.0.0 <=2.6.0 specification elements.
 */
_AsyncApi.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'AsyncApi', '$visitor']);
_AsyncApiVersion.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'AsyncApiVersion', '$visitor']);
_ChannelBindings.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'ChannelBindings', '$visitor']);
_ChannelItem.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'ChannelItem', '$visitor']);
_Channels.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Channels', '$visitor']);
_Components.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Components', '$visitor']);
_Contact.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Contact', '$visitor']);
_CorrelationID.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'CorrelationID', '$visitor']);
_DefaultContentType.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'DefaultContentType', '$visitor']);
_ExternalDocumentation.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'ExternalDocumentation', '$visitor']);
_Identifier.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Identifier', '$visitor']);
_Info.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Info', '$visitor']);
_License.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'License', '$visitor']);
_Message.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Message', '$visitor']);
_MessageBindings.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'MessageBindings', '$visitor']);
_MessageExample.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'MessageExample', '$visitor']);
_MessageTrait.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'MessageTrait', '$visitor']);
_OAuthFlow.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'OAuthFlow', '$visitor']);
_OAuthFlows.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'OAuthFlows', '$visitor']);
_Operation.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Operation', '$visitor']);
_OperationBindings.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'OperationBindings', '$visitor']);
_OperationTrait.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'OperationTrait', '$visitor']);
_Parameter.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Parameter', '$visitor']);
_Parameters.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Parameters', '$visitor']);
_Reference.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Reference', '$visitor']);
_Schema.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Schema', '$visitor']);
_SecurityRequirement.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'SecurityRequirement', '$visitor']);
_SecurityScheme.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'SecurityScheme', '$visitor']);
_Server.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Server', '$visitor']);
_ServerBindings.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'ServerBindings', '$visitor']);
_Servers.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Servers', '$visitor']);
_ServerVariable.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'ServerVariable', '$visitor']);
_Tag.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Tag', '$visitor']);
_Tags.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'Tags', '$visitor']);

/**
 * Binding elements.
 */
// AMQP 0-9-1
_AmqpChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'amqp', 'ChannelBinding', '$visitor']);
_AmqpMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'amqp', 'MessageBinding', '$visitor']);
_AmqpOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'amqp', 'OperationBinding', '$visitor']);
_AmqpServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'amqp', 'ServerBinding', '$visitor']);
// AMQP 1.0
_Amqp1ChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'amqp1', 'ChannelBinding', '$visitor']);
_Amqp1MessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'amqp1', 'MessageBinding', '$visitor']);
_Amqp1OperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'amqp1', 'OperationBinding', '$visitor']);
_Amqp1ServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'amqp1', 'ServerBinding', '$visitor']);
// HTTP
_HttpChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'http', 'ChannelBinding', '$visitor']);
_HttpMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'http', 'MessageBinding', '$visitor']);
_HttpOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'http', 'OperationBinding', '$visitor']);
_HttpServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'http', 'ServerBinding', '$visitor']);
// Google Cloud Pub/Sub
_GooglepubsubChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'googlepubsub', 'ChannelBinding', '$visitor']);
_GooglepubsubMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'googlepubsub', 'MessageBinding', '$visitor']);
_GooglepubsubOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'googlepubsub', 'OperationBinding', '$visitor']);
_GooglepubsubServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'googlepubsub', 'ServerBinding', '$visitor']);
// IBM MQ
_IbmmqChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'ibmmq', 'ChannelBinding', '$visitor']);
_IbmmqMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'ibmmq', 'MessageBinding', '$visitor']);
_IbmmqOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'ibmmq', 'OperationBinding', '$visitor']);
_IbmmqServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'ibmmq', 'ServerBinding', '$visitor']);
// JMS
_JmsChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'jms', 'ChannelBinding', '$visitor']);
_JmsMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'jms', 'MessageBinding', '$visitor']);
_JmsOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'jms', 'OperationBinding', '$visitor']);
_JmsServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'jms', 'ServerBinding', '$visitor']);
// Kafka
_KafkaChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'kafka', 'ChannelBinding', '$visitor']);
_KafkaMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'kafka', 'MessageBinding', '$visitor']);
_KafkaOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'kafka', 'OperationBinding', '$visitor']);
_KafkaServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'kafka', 'ServerBinding', '$visitor']);
// Anypoint MQ
_AnypointmqChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'anypointmq', 'ChannelBinding', '$visitor']);
_AnypointmqMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'anypointmq', 'MessageBinding', '$visitor']);
_AnypointmqOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'anypointmq', 'OperationBinding', '$visitor']);
_AnypointmqServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'anypointmq', 'ServerBinding', '$visitor']);
// Mercure
_MercureChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'mercure', 'ChannelBinding', '$visitor']);
_MercureMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'mercure', 'MessageBinding', '$visitor']);
_MercureOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'mercure', 'OperationBinding', '$visitor']);
_MercureServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'mercure', 'ServerBinding', '$visitor']);
// MQTT
_MqttChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'mqtt', 'ChannelBinding', '$visitor']);
_MqttMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'mqtt', 'MessageBinding', '$visitor']);
_MqttOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'mqtt', 'OperationBinding', '$visitor']);
_MqttServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'mqtt', 'ServerBinding', '$visitor']);
// MQTT 5
_Mqtt5ChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'mqtt5', 'ChannelBinding', '$visitor']);
_Mqtt5MessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'mqtt5', 'MessageBinding', '$visitor']);
_Mqtt5OperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'mqtt5', 'OperationBinding', '$visitor']);
_Mqtt5ServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'mqtt5', 'ServerBinding', '$visitor']);
// NATS
_NatsChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'nats', 'ChannelBinding', '$visitor']);
_NatsMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'nats', 'MessageBinding', '$visitor']);
_NatsOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'nats', 'OperationBinding', '$visitor']);
_NatsServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'nats', 'ServerBinding', '$visitor']);
// Pulsar
_PulsarChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'pulsar', 'ChannelBinding', '$visitor']);
_PulsarMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'pulsar', 'MessageBinding', '$visitor']);
_PulsarOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'pulsar', 'OperationBinding', '$visitor']);
_PulsarServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'pulsar', 'ServerBinding', '$visitor']);
// Redis
_RedisChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'redis', 'ChannelBinding', '$visitor']);
_RedisMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'redis', 'MessageBinding', '$visitor']);
_RedisOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'redis', 'OperationBinding', '$visitor']);
_RedisServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'redis', 'ServerBinding', '$visitor']);
// SNS
_SnsChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'sns', 'ChannelBinding', '$visitor']);
_SnsMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'sns', 'MessageBinding', '$visitor']);
_SnsOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'sns', 'OperationBinding', '$visitor']);
_SnsServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'sns', 'ServerBinding', '$visitor']);
// Solace
_SolaceChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'solace', 'ChannelBinding', '$visitor']);
_SolaceMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'solace', 'MessageBinding', '$visitor']);
_SolaceOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'solace', 'OperationBinding', '$visitor']);
_SolaceServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'solace', 'ServerBinding', '$visitor']);
// SQS
_SqsChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'sqs', 'ChannelBinding', '$visitor']);
_SqsMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'sqs', 'MessageBinding', '$visitor']);
_SqsOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'sqs', 'OperationBinding', '$visitor']);
_SqsServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'sqs', 'ServerBinding', '$visitor']);
// STOMP
_StompChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'stomp', 'ChannelBinding', '$visitor']);
_StompMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'stomp', 'MessageBinding', '$visitor']);
_StompOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'stomp', 'OperationBinding', '$visitor']);
_StompServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'stomp', 'ServerBinding', '$visitor']);
// WebSocket
_WebSocketChannelBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'ws', 'ChannelBinding', '$visitor']);
_WebSocketMessageBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'ws', 'MessageBinding', '$visitor']);
_WebSocketOperationBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'ws', 'OperationBinding', '$visitor']);
_WebSocketServerBinding.default.refract = (0, _index.createRefractor)(['visitors', 'document', 'objects', 'bindings', 'ws', 'ServerBinding', '$visitor']);
