// @backstage/test-utils/dist/testUtils/apis/StorageApi/MockStorageApi.esm.js
import ObservableImpl from 'zen-observable';

class MockStorageApi {
  namespace;
  data;
  bucketStorageApis;
  constructor(namespace, bucketStorageApis, data) {
    this.namespace = namespace;
    this.bucketStorageApis = bucketStorageApis;
    this.data = { ...data };
  }
  static create(data) {
    const keyValues = {};
    function put(value, namespace) {
      for (const [key, val] of Object.entries(value)) {
        if (typeof val === "object" && val !== null) {
          put(val, `${namespace}/${key}`);
        } else {
          const namespacedKey = `${namespace}/${key.replace(/^\//, "")}`;
          keyValues[namespacedKey] = val;
        }
      }
    }
    put(data ?? {}, "");
    return new MockStorageApi("", /* @__PURE__ */ new Map(), keyValues);
  }
  forBucket(name) {
    if (!this.bucketStorageApis.has(name)) {
      this.bucketStorageApis.set(
        name,
        new MockStorageApi(
          `${this.namespace}/${name}`,
          this.bucketStorageApis,
          this.data
        )
      );
    }
    return this.bucketStorageApis.get(name);
  }
  snapshot(key) {
    if (this.data.hasOwnProperty(this.getKeyName(key))) {
      const data = this.data[this.getKeyName(key)];
      return {
        key,
        presence: "present",
        value: data
      };
    }
    return {
      key,
      presence: "absent",
      value: void 0
    };
  }
  async set(key, data) {
    const serialized = JSON.parse(JSON.stringify(data), (_key, value) => {
      if (typeof value === "object" && value !== null) {
        Object.freeze(value);
      }
      return value;
    });
    this.data[this.getKeyName(key)] = serialized;
    this.notifyChanges({
      key,
      presence: "present",
      value: serialized
    });
  }
  async remove(key) {
    delete this.data[this.getKeyName(key)];
    this.notifyChanges({
      key,
      presence: "absent",
      value: void 0
    });
  }
  observe$(key) {
    return this.observable.filter(({ key: messageKey }) => messageKey === key);
  }
  getKeyName(key) {
    return `${this.namespace}/${encodeURIComponent(key)}`;
  }
  notifyChanges(message) {
    for (const subscription of this.subscribers) {
      subscription.next(message);
    }
  }
  subscribers = /* @__PURE__ */ new Set();
  observable = new ObservableImpl((subscriber) => {
    this.subscribers.add(subscriber);
    return () => {
      this.subscribers.delete(subscriber);
    };
  });
}

export { MockStorageApi };
