// @swagger-api/apidom-ns-asyncapi-2/src/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard").default;
exports.__esModule = true;
exports.SolaceChannelBindingElement = exports.SnsServerBindingElement = exports.SnsOperationBindingElement = exports.SnsMessageBindingElement = exports.SnsChannelBindingElement = exports.ServersElement = exports.ServerVariableElement = exports.ServerElement = exports.ServerBindingsElement = exports.SecuritySchemeElement = exports.SecurityRequirementElement = exports.SchemaElement = exports.ReferenceElement = exports.RedisServerBindingElement = exports.RedisOperationBindingElement = exports.RedisMessageBindingElement = exports.RedisChannelBindingElement = exports.PulsarServerBindingElement = exports.PulsarOperationBindingElement = exports.PulsarMessageBindingElement = exports.PulsarChannelBindingElement = exports.PatternedFieldsVisitor = exports.ParametersElement = exports.ParameterElement = exports.OperationTraitElement = exports.OperationElement = exports.OperationBindingsElement = exports.OAuthFlowsElement = exports.OAuthFlowElement = exports.NatsServerBindingElement = exports.NatsOperationBindingElement = exports.NatsMessageBindingElement = exports.NatsChannelBindingElement = exports.MqttServerBindingElement = exports.MqttOperationBindingElement = exports.MqttMessageBindingElement = exports.MqttChannelBindingElement = exports.Mqtt5ServerBindingElement = exports.Mqtt5OperationBindingElement = exports.Mqtt5MessageBindingElement = exports.Mqtt5ChannelBindingElement = exports.MixedFieldsVisitor = exports.MessageTraitElement = exports.MessageExampleElement = exports.MessageElement = exports.MessageBindingsElement = exports.MercureServerBindingElement = exports.MercureOperationBindingElement = exports.MercureMessageBindingElement = exports.MercureChannelBindingElement = exports.MapVisitor = exports.LicenseElement = exports.KafkaServerBindingElement = exports.KafkaOperationBindingElement = exports.KafkaMessageBindingElement = exports.KafkaChannelBindingElement = exports.JmsServerBindingElement = exports.JmsOperationBindingElement = exports.JmsMessageBindingElement = exports.JmsChannelBindingElement = exports.InfoElement = exports.IdentifierElement = exports.IbmmqServerBindingElement = exports.IbmmqOperationBindingElement = exports.IbmmqMessageBindingElement = exports.IbmmqChannelBindingElement = exports.HttpServerBindingElement = exports.HttpOperationBindingElement = exports.HttpMessageBindingElement = exports.HttpChannelBindingElement = exports.GooglepubsubServerBindingElement = exports.GooglepubsubOperationBindingElement = exports.GooglepubsubMessageBindingElement = exports.GooglepubsubChannelBindingElement = exports.FixedFieldsVisitor = exports.FallbackVisitor = exports.ExternalDocumentationElement = exports.DefaultContentTypeElement = exports.CorrelationIDElement = exports.ContactElement = exports.ComponentsElement = exports.ChannelsElement = exports.ChannelItemElement = exports.ChannelBindingsElement = exports.AsyncApiVersionElement = exports.AsyncApi2Element = exports.AsyncAPIMediaTypes = exports.AnypointmqServerBindingElement = exports.AnypointmqOperationBindingElement = exports.AnypointmqMessageBindingElement = exports.AnypointmqChannelBindingElement = exports.AmqpServerBindingElement = exports.AmqpOperationBindingElement = exports.AmqpMessageBindingElement = exports.AmqpChannelBindingElement = exports.Amqp1ServerBindingElement = exports.Amqp1OperationBindingElement = exports.Amqp1MessageBindingElement = exports.Amqp1ChannelBindingElement = exports.AlternatingVisitor = void 0;
exports.refractorPluginReplaceEmptyElement = exports.mediaTypes = exports.keyMap = exports.isStringElement = exports.isServersElement = exports.isServerVariableElement = exports.isServerElement = exports.isServerBindingsElement = exports.isSecurityRequirementElement = exports.isSchemaElement = exports.isReferenceLikeElement = exports.isReferenceElement = exports.isRefElement = exports.isParametersElement = exports.isParameterElement = exports.isOperationElement = exports.isObjectElement = exports.isNumberElement = exports.isNullElement = exports.isMemberElement = exports.isLinkElement = exports.isLicenseElement = exports.isInfoElement = exports.isIdentifierElement = exports.isElement = exports.isContactElement = exports.isComponentsElement = exports.isChannelsElement = exports.isChannelItemElement = exports.isChannelBindingsElement = exports.isBooleanJsonSchemaElement = exports.isBooleanElement = exports.isAsyncApiVersionElement = exports.isAsyncApiExtension = exports.isAsyncApi2Element = exports.isArrayElement = exports.getNodeType = exports.default = exports.WebSocketServerBindingElement = exports.WebSocketOperationBindingElement = exports.WebSocketMessageBindingElement = exports.WebSocketChannelBindingElement = exports.Visitor = exports.TagsElement = exports.TagElement = exports.StompServerBindingElement = exports.StompOperationBindingElement = exports.StompMessageBindingElement = exports.StompChannelBindingElement = exports.SqsServerBindingElement = exports.SqsOperationBindingElement = exports.SqsMessageBindingElement = exports.SqsChannelBindingElement = exports.SpecificationVisitor = exports.SpecificationExtensionVisitor = exports.SolaceServerBindingElement = exports.SolaceOperationBindingElement = exports.SolaceMessageBindingElement = void 0;
var _apidomCore = require("@swagger-api/apidom-core");
exports.isRefElement = _apidomCore.isRefElement;
exports.isLinkElement = _apidomCore.isLinkElement;
exports.isMemberElement = _apidomCore.isMemberElement;
exports.isObjectElement = _apidomCore.isObjectElement;
exports.isArrayElement = _apidomCore.isArrayElement;
exports.isBooleanElement = _apidomCore.isBooleanElement;
exports.isNullElement = _apidomCore.isNullElement;
exports.isElement = _apidomCore.isElement;
exports.isNumberElement = _apidomCore.isNumberElement;
exports.isStringElement = _apidomCore.isStringElement;
var _mediaTypes = _interopRequireWildcard(require("./media-types.cjs"));
exports.mediaTypes = _mediaTypes.default;
exports.AsyncAPIMediaTypes = _mediaTypes.AsyncAPIMediaTypes;
var _namespace = _interopRequireDefault(require("./namespace.cjs"));
exports.default = _namespace.default;
var _replaceEmptyElement = _interopRequireDefault(require("./refractor/plugins/replace-empty-element.cjs"));
exports.refractorPluginReplaceEmptyElement = _replaceEmptyElement.default;
var _predicates = require("./predicates.cjs");
exports.isAsyncApi2Element = _predicates.isAsyncApi2Element;
exports.isAsyncApiVersionElement = _predicates.isAsyncApiVersionElement;
exports.isChannelBindingsElement = _predicates.isChannelBindingsElement;
exports.isChannelItemElement = _predicates.isChannelItemElement;
exports.isChannelsElement = _predicates.isChannelsElement;
exports.isComponentsElement = _predicates.isComponentsElement;
exports.isContactElement = _predicates.isContactElement;
exports.isIdentifierElement = _predicates.isIdentifierElement;
exports.isInfoElement = _predicates.isInfoElement;
exports.isLicenseElement = _predicates.isLicenseElement;
exports.isOperationElement = _predicates.isOperationElement;
exports.isParameterElement = _predicates.isParameterElement;
exports.isParametersElement = _predicates.isParametersElement;
exports.isReferenceElement = _predicates.isReferenceElement;
exports.isSchemaElement = _predicates.isSchemaElement;
exports.isBooleanJsonSchemaElement = _predicates.isBooleanJsonSchemaElement;
exports.isSecurityRequirementElement = _predicates.isSecurityRequirementElement;
exports.isServerElement = _predicates.isServerElement;
exports.isServerBindingsElement = _predicates.isServerBindingsElement;
exports.isServersElement = _predicates.isServersElement;
exports.isServerVariableElement = _predicates.isServerVariableElement;
var _predicates2 = require("./refractor/predicates.cjs");
exports.isReferenceLikeElement = _predicates2.isReferenceLikeElement;
exports.isAsyncApiExtension = _predicates2.isAsyncApiExtension;
var _AlternatingVisitor = _interopRequireDefault(require("./refractor/visitors/generics/AlternatingVisitor.cjs"));
exports.AlternatingVisitor = _AlternatingVisitor.default;
var _FixedFieldsVisitor = _interopRequireDefault(require("./refractor/visitors/generics/FixedFieldsVisitor.cjs"));
exports.FixedFieldsVisitor = _FixedFieldsVisitor.default;
var _MapVisitor = _interopRequireDefault(require("./refractor/visitors/generics/MapVisitor.cjs"));
exports.MapVisitor = _MapVisitor.default;
var _MixedFieldsVisitor = _interopRequireDefault(require("./refractor/visitors/generics/MixedFieldsVisitor.cjs"));
exports.MixedFieldsVisitor = _MixedFieldsVisitor.default;
var _PatternedFieldsVisitor = _interopRequireDefault(require("./refractor/visitors/generics/PatternedFieldsVisitor.cjs"));
exports.PatternedFieldsVisitor = _PatternedFieldsVisitor.default;
var _FallbackVisitor = _interopRequireDefault(require("./refractor/visitors/FallbackVisitor.cjs"));
exports.FallbackVisitor = _FallbackVisitor.default;
var _SpecificationExtensionVisitor = _interopRequireDefault(require("./refractor/visitors/SpecificationExtensionVisitor.cjs"));
exports.SpecificationExtensionVisitor = _SpecificationExtensionVisitor.default;
var _SpecificationVisitor = _interopRequireDefault(require("./refractor/visitors/SpecificationVisitor.cjs"));
exports.SpecificationVisitor = _SpecificationVisitor.default;
var _Visitor = _interopRequireDefault(require("./refractor/visitors/Visitor.cjs"));
exports.Visitor = _Visitor.default;
var _visitor = require("./traversal/visitor.cjs");
exports.keyMap = _visitor.keyMap;
exports.getNodeType = _visitor.getNodeType;
var _registration = require("./refractor/registration.cjs");
exports.AsyncApi2Element = _registration.AsyncApi2Element;
exports.AsyncApiVersionElement = _registration.AsyncApiVersionElement;
exports.ChannelBindingsElement = _registration.ChannelBindingsElement;
exports.ChannelItemElement = _registration.ChannelItemElement;
exports.ChannelsElement = _registration.ChannelsElement;
exports.ComponentsElement = _registration.ComponentsElement;
exports.ContactElement = _registration.ContactElement;
exports.CorrelationIDElement = _registration.CorrelationIDElement;
exports.DefaultContentTypeElement = _registration.DefaultContentTypeElement;
exports.ExternalDocumentationElement = _registration.ExternalDocumentationElement;
exports.IdentifierElement = _registration.IdentifierElement;
exports.InfoElement = _registration.InfoElement;
exports.LicenseElement = _registration.LicenseElement;
exports.MessageElement = _registration.MessageElement;
exports.MessageBindingsElement = _registration.MessageBindingsElement;
exports.MessageExampleElement = _registration.MessageExampleElement;
exports.MessageTraitElement = _registration.MessageTraitElement;
exports.OAuthFlowElement = _registration.OAuthFlowElement;
exports.OAuthFlowsElement = _registration.OAuthFlowsElement;
exports.OperationElement = _registration.OperationElement;
exports.OperationBindingsElement = _registration.OperationBindingsElement;
exports.OperationTraitElement = _registration.OperationTraitElement;
exports.ParameterElement = _registration.ParameterElement;
exports.ParametersElement = _registration.ParametersElement;
exports.ReferenceElement = _registration.ReferenceElement;
exports.SchemaElement = _registration.SchemaElement;
exports.SecurityRequirementElement = _registration.SecurityRequirementElement;
exports.SecuritySchemeElement = _registration.SecuritySchemeElement;
exports.ServerElement = _registration.ServerElement;
exports.ServerBindingsElement = _registration.ServerBindingsElement;
exports.ServersElement = _registration.ServersElement;
exports.ServerVariableElement = _registration.ServerVariableElement;
exports.TagElement = _registration.TagElement;
exports.TagsElement = _registration.TagsElement;
exports.AmqpChannelBindingElement = _registration.AmqpChannelBindingElement;
exports.AmqpMessageBindingElement = _registration.AmqpMessageBindingElement;
exports.AmqpOperationBindingElement = _registration.AmqpOperationBindingElement;
exports.AmqpServerBindingElement = _registration.AmqpServerBindingElement;
exports.Amqp1ChannelBindingElement = _registration.Amqp1ChannelBindingElement;
exports.Amqp1MessageBindingElement = _registration.Amqp1MessageBindingElement;
exports.Amqp1OperationBindingElement = _registration.Amqp1OperationBindingElement;
exports.Amqp1ServerBindingElement = _registration.Amqp1ServerBindingElement;
exports.AnypointmqChannelBindingElement = _registration.AnypointmqChannelBindingElement;
exports.AnypointmqMessageBindingElement = _registration.AnypointmqMessageBindingElement;
exports.AnypointmqOperationBindingElement = _registration.AnypointmqOperationBindingElement;
exports.AnypointmqServerBindingElement = _registration.AnypointmqServerBindingElement;
exports.GooglepubsubChannelBindingElement = _registration.GooglepubsubChannelBindingElement;
exports.GooglepubsubMessageBindingElement = _registration.GooglepubsubMessageBindingElement;
exports.GooglepubsubOperationBindingElement = _registration.GooglepubsubOperationBindingElement;
exports.GooglepubsubServerBindingElement = _registration.GooglepubsubServerBindingElement;
exports.HttpChannelBindingElement = _registration.HttpChannelBindingElement;
exports.HttpMessageBindingElement = _registration.HttpMessageBindingElement;
exports.HttpOperationBindingElement = _registration.HttpOperationBindingElement;
exports.HttpServerBindingElement = _registration.HttpServerBindingElement;
exports.IbmmqChannelBindingElement = _registration.IbmmqChannelBindingElement;
exports.IbmmqMessageBindingElement = _registration.IbmmqMessageBindingElement;
exports.IbmmqOperationBindingElement = _registration.IbmmqOperationBindingElement;
exports.IbmmqServerBindingElement = _registration.IbmmqServerBindingElement;
exports.JmsChannelBindingElement = _registration.JmsChannelBindingElement;
exports.JmsMessageBindingElement = _registration.JmsMessageBindingElement;
exports.JmsOperationBindingElement = _registration.JmsOperationBindingElement;
exports.JmsServerBindingElement = _registration.JmsServerBindingElement;
exports.KafkaChannelBindingElement = _registration.KafkaChannelBindingElement;
exports.KafkaMessageBindingElement = _registration.KafkaMessageBindingElement;
exports.KafkaOperationBindingElement = _registration.KafkaOperationBindingElement;
exports.KafkaServerBindingElement = _registration.KafkaServerBindingElement;
exports.MercureChannelBindingElement = _registration.MercureChannelBindingElement;
exports.MercureMessageBindingElement = _registration.MercureMessageBindingElement;
exports.MercureOperationBindingElement = _registration.MercureOperationBindingElement;
exports.MercureServerBindingElement = _registration.MercureServerBindingElement;
exports.MqttChannelBindingElement = _registration.MqttChannelBindingElement;
exports.MqttMessageBindingElement = _registration.MqttMessageBindingElement;
exports.MqttOperationBindingElement = _registration.MqttOperationBindingElement;
exports.MqttServerBindingElement = _registration.MqttServerBindingElement;
exports.Mqtt5ChannelBindingElement = _registration.Mqtt5ChannelBindingElement;
exports.Mqtt5MessageBindingElement = _registration.Mqtt5MessageBindingElement;
exports.Mqtt5OperationBindingElement = _registration.Mqtt5OperationBindingElement;
exports.Mqtt5ServerBindingElement = _registration.Mqtt5ServerBindingElement;
exports.NatsChannelBindingElement = _registration.NatsChannelBindingElement;
exports.NatsMessageBindingElement = _registration.NatsMessageBindingElement;
exports.NatsOperationBindingElement = _registration.NatsOperationBindingElement;
exports.NatsServerBindingElement = _registration.NatsServerBindingElement;
exports.PulsarChannelBindingElement = _registration.PulsarChannelBindingElement;
exports.PulsarMessageBindingElement = _registration.PulsarMessageBindingElement;
exports.PulsarOperationBindingElement = _registration.PulsarOperationBindingElement;
exports.PulsarServerBindingElement = _registration.PulsarServerBindingElement;
exports.RedisChannelBindingElement = _registration.RedisChannelBindingElement;
exports.RedisMessageBindingElement = _registration.RedisMessageBindingElement;
exports.RedisOperationBindingElement = _registration.RedisOperationBindingElement;
exports.RedisServerBindingElement = _registration.RedisServerBindingElement;
exports.SnsChannelBindingElement = _registration.SnsChannelBindingElement;
exports.SnsMessageBindingElement = _registration.SnsMessageBindingElement;
exports.SnsOperationBindingElement = _registration.SnsOperationBindingElement;
exports.SnsServerBindingElement = _registration.SnsServerBindingElement;
exports.SolaceChannelBindingElement = _registration.SolaceChannelBindingElement;
exports.SolaceMessageBindingElement = _registration.SolaceMessageBindingElement;
exports.SolaceOperationBindingElement = _registration.SolaceOperationBindingElement;
exports.SolaceServerBindingElement = _registration.SolaceServerBindingElement;
exports.SqsChannelBindingElement = _registration.SqsChannelBindingElement;
exports.SqsMessageBindingElement = _registration.SqsMessageBindingElement;
exports.SqsOperationBindingElement = _registration.SqsOperationBindingElement;
exports.SqsServerBindingElement = _registration.SqsServerBindingElement;
exports.StompChannelBindingElement = _registration.StompChannelBindingElement;
exports.StompMessageBindingElement = _registration.StompMessageBindingElement;
exports.StompOperationBindingElement = _registration.StompOperationBindingElement;
exports.StompServerBindingElement = _registration.StompServerBindingElement;
exports.WebSocketChannelBindingElement = _registration.WebSocketChannelBindingElement;
exports.WebSocketMessageBindingElement = _registration.WebSocketMessageBindingElement;
exports.WebSocketOperationBindingElement = _registration.WebSocketOperationBindingElement;
exports.WebSocketServerBindingElement = _registration.WebSocketServerBindingElement;
