// @swagger-api/apidom-ns-json-schema-draft-4/src/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime-corejs3/helpers/interopRequireWildcard").default;
exports.__esModule = true;
exports.specificationObj = exports.refractorPluginReplaceEmptyElement = exports.refract = exports.mediaTypes = exports.keyMap = exports.isStringElement = exports.isRefElement = exports.isObjectElement = exports.isNumberElement = exports.isNullElement = exports.isMemberElement = exports.isMediaElement = exports.isLinkElement = exports.isLinkDescriptionElement = exports.isJSONSchemaElement = exports.isJSONReferenceLikeElement = exports.isJSONReferenceElement = exports.isElement = exports.isBooleanElement = exports.isArrayElement = exports.getNodeType = exports.default = exports.createRefractor = exports.Visitor = exports.SpecificationVisitor = exports.PatternedFieldsVisitor = exports.ParentSchemaAwareVisitor = exports.MediaElement = exports.MapVisitor = exports.LinkDescriptionElement = exports.JSONSchemaElement = exports.JSONSchemaDraft4MediaTypes = exports.JSONReferenceElement = exports.FixedFieldsVisitor = exports.FallbackVisitor = exports.AlternatingVisitor = void 0;
var _apidomCore = require("@swagger-api/apidom-core");
exports.isRefElement = _apidomCore.isRefElement;
exports.isLinkElement = _apidomCore.isLinkElement;
exports.isMemberElement = _apidomCore.isMemberElement;
exports.isObjectElement = _apidomCore.isObjectElement;
exports.isArrayElement = _apidomCore.isArrayElement;
exports.isBooleanElement = _apidomCore.isBooleanElement;
exports.isNullElement = _apidomCore.isNullElement;
exports.isElement = _apidomCore.isElement;
exports.isNumberElement = _apidomCore.isNumberElement;
exports.isStringElement = _apidomCore.isStringElement;
var _mediaTypes = _interopRequireWildcard(require("./media-types.cjs"));
exports.mediaTypes = _mediaTypes.default;
exports.JSONSchemaDraft4MediaTypes = _mediaTypes.JSONSchemaDraft4MediaTypes;
var _namespace = _interopRequireDefault(require("./namespace.cjs"));
exports.default = _namespace.default;
var _replaceEmptyElement = _interopRequireDefault(require("./refractor/plugins/replace-empty-element.cjs"));
exports.refractorPluginReplaceEmptyElement = _replaceEmptyElement.default;
var _index = _interopRequireWildcard(require("./refractor/index.cjs"));
exports.refract = _index.default;
exports.createRefractor = _index.createRefractor;
var _specification = _interopRequireDefault(require("./refractor/specification.cjs"));
exports.specificationObj = _specification.default;
var _predicates = require("./predicates.cjs");
exports.isJSONReferenceElement = _predicates.isJSONReferenceElement;
exports.isJSONSchemaElement = _predicates.isJSONSchemaElement;
exports.isLinkDescriptionElement = _predicates.isLinkDescriptionElement;
exports.isMediaElement = _predicates.isMediaElement;
var _predicates2 = require("./refractor/predicates.cjs");
exports.isJSONReferenceLikeElement = _predicates2.isJSONReferenceLikeElement;
var _Visitor = _interopRequireDefault(require("./refractor/visitors/Visitor.cjs"));
exports.Visitor = _Visitor.default;
var _SpecificationVisitor = _interopRequireDefault(require("./refractor/visitors/SpecificationVisitor.cjs"));
exports.SpecificationVisitor = _SpecificationVisitor.default;
var _FallbackVisitor = _interopRequireDefault(require("./refractor/visitors/FallbackVisitor.cjs"));
exports.FallbackVisitor = _FallbackVisitor.default;
var _FixedFieldsVisitor = _interopRequireDefault(require("./refractor/visitors/generics/FixedFieldsVisitor.cjs"));
exports.FixedFieldsVisitor = _FixedFieldsVisitor.default;
var _PatternedFieldsVisitor = _interopRequireDefault(require("./refractor/visitors/generics/PatternedFieldsVisitor.cjs"));
exports.PatternedFieldsVisitor = _PatternedFieldsVisitor.default;
var _MapVisitor = _interopRequireDefault(require("./refractor/visitors/generics/MapVisitor.cjs"));
exports.MapVisitor = _MapVisitor.default;
var _AlternatingVisitor = _interopRequireDefault(require("./refractor/visitors/generics/AlternatingVisitor.cjs"));
exports.AlternatingVisitor = _AlternatingVisitor.default;
var _ParentSchemaAwareVisitor = _interopRequireDefault(require("./refractor/visitors/json-schema/ParentSchemaAwareVisitor.cjs"));
exports.ParentSchemaAwareVisitor = _ParentSchemaAwareVisitor.default;
var _visitor = require("./traversal/visitor.cjs");
exports.keyMap = _visitor.keyMap;
exports.getNodeType = _visitor.getNodeType;
var _registration = require("./refractor/registration.cjs");
exports.JSONSchemaElement = _registration.JSONSchemaElement;
exports.JSONReferenceElement = _registration.JSONReferenceElement;
exports.MediaElement = _registration.MediaElement;
exports.LinkDescriptionElement = _registration.LinkDescriptionElement;
