// @backstage/plugin-catalog/dist/components/DependsOnResourcesCard/DependsOnResourcesCard.esm.js
import { RELATION_DEPENDS_ON } from '@backstage/catalog-model';
import React from 'react';
import { RelatedEntitiesCard } from '../RelatedEntitiesCard/RelatedEntitiesCard.esm.js';
import { resourceEntityColumns, componentEntityHelpLink, asResourceEntities } from '../RelatedEntitiesCard/presets.esm.js';
import { catalogTranslationRef } from '../../alpha/translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

function DependsOnResourcesCard(props) {
  const { t } = useTranslationRef(catalogTranslationRef);
  const {
    variant = "gridItem",
    title = t("dependsOnResourcesCard.title"),
    columns = resourceEntityColumns,
    tableOptions = {}
  } = props;
  return /* @__PURE__ */ React.createElement(
    RelatedEntitiesCard,
    {
      variant,
      title,
      entityKind: "Resource",
      relationType: RELATION_DEPENDS_ON,
      columns,
      emptyMessage: t("dependsOnResourcesCard.emptyMessage"),
      emptyHelpLink: componentEntityHelpLink,
      asRenderableEntities: asResourceEntities,
      tableOptions
    }
  );
}

export { DependsOnResourcesCard };
