// @trendyol-js/openstack-swift-sdk/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UploadResponse = exports.UploadOptions = exports.NotFound = exports.ListObjectResponse = exports.ObjectMetaResponse = exports.ContainerMetaResponse = exports.DownloadResponse = void 0;
var DownloadResponse = /** @class */ (function () {
    function DownloadResponse(headers, data) {
        this.headers = headers;
        this.data = data;
    }
    return DownloadResponse;
}());
exports.DownloadResponse = DownloadResponse;
var ContainerMetaResponse = /** @class */ (function () {
    function ContainerMetaResponse(init) {
        Object.assign(this, init);
    }
    return ContainerMetaResponse;
}());
exports.ContainerMetaResponse = ContainerMetaResponse;
var ObjectMetaResponse = /** @class */ (function () {
    function ObjectMetaResponse(init) {
        Object.assign(this, init);
    }
    return ObjectMetaResponse;
}());
exports.ObjectMetaResponse = ObjectMetaResponse;
var ListObjectResponse = /** @class */ (function () {
    function ListObjectResponse(init) {
        Object.assign(this, init);
    }
    return ListObjectResponse;
}());
exports.ListObjectResponse = ListObjectResponse;
var NotFound = /** @class */ (function () {
    function NotFound() {
    }
    return NotFound;
}());
exports.NotFound = NotFound;
var UploadOptions = /** @class */ (function () {
    function UploadOptions(init) {
        Object.assign(this, init);
    }
    return UploadOptions;
}());
exports.UploadOptions = UploadOptions;
var UploadResponse = /** @class */ (function () {
    function UploadResponse(fullPath) {
        this.fullPath = fullPath;
    }
    return UploadResponse;
}());
exports.UploadResponse = UploadResponse;
