// @backstage/plugin-auth-backend-module-oauth2-proxy-provider/dist/module.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var pluginAuthNode = require('@backstage/plugin-auth-node');
var authenticator = require('./authenticator.cjs.js');
var resolvers = require('./resolvers.cjs.js');

const authModuleOauth2ProxyProvider = backendPluginApi.createBackendModule({
  pluginId: "auth",
  moduleId: "oauth2-proxy-provider",
  register(reg) {
    reg.registerInit({
      deps: {
        providers: pluginAuthNode.authProvidersExtensionPoint
      },
      async init({ providers }) {
        providers.registerProvider({
          providerId: "oauth2Proxy",
          factory: pluginAuthNode.createProxyAuthProviderFactory({
            authenticator: authenticator.oauth2ProxyAuthenticator,
            signInResolverFactories: {
              ...pluginAuthNode.commonSignInResolvers,
              ...resolvers.oauth2ProxySignInResolvers
            }
          })
        });
      }
    });
  }
});

exports.authModuleOauth2ProxyProvider = authModuleOauth2ProxyProvider;
