// @backstage/plugin-auth-backend-module-guest-provider/dist/module.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var pluginAuthNode = require('@backstage/plugin-auth-node');
var authenticator = require('./authenticator.cjs.js');
var resolvers = require('./resolvers.cjs.js');

const authModuleGuestProvider = backendPluginApi.createBackendModule({
  pluginId: "auth",
  moduleId: "guest-provider",
  register(reg) {
    reg.registerInit({
      deps: {
        providers: pluginAuthNode.authProvidersExtensionPoint,
        config: backendPluginApi.coreServices.rootConfig
      },
      async init({ providers, config }) {
        providers.registerProvider({
          providerId: "guest",
          factory: pluginAuthNode.createProxyAuthProviderFactory({
            authenticator: authenticator.guestAuthenticator,
            signInResolver: resolvers.signInAsGuestUser(
              config.getConfig("auth.providers.guest")
            )
          })
        });
      }
    });
  }
});

exports.authModuleGuestProvider = authModuleGuestProvider;
