// @backstage/plugin-auth-backend-module-okta-provider/dist/authenticator.cjs.js
'use strict';

var passportOktaOauth = require('@davidzemon/passport-okta-oauth');
var pluginAuthNode = require('@backstage/plugin-auth-node');

const oktaAuthenticator = pluginAuthNode.createOAuthAuthenticator({
  defaultProfileTransform: pluginAuthNode.PassportOAuthAuthenticatorHelper.defaultProfileTransform,
  scopes: {
    required: ["openid", "email", "profile", "offline_access"]
  },
  initialize({ callbackUrl, config }) {
    const clientId = config.getString("clientId");
    const clientSecret = config.getString("clientSecret");
    const audience = config.getOptionalString("audience") || "https://okta.com";
    const authServerId = config.getOptionalString("authServerId");
    const idp = config.getOptionalString("idp");
    return pluginAuthNode.PassportOAuthAuthenticatorHelper.from(
      new passportOktaOauth.Strategy(
        {
          clientID: clientId,
          clientSecret,
          callbackURL: callbackUrl,
          audience,
          authServerID: authServerId,
          idp,
          passReqToCallback: false,
          response_type: "code"
        },
        (accessToken, refreshToken, params, fullProfile, done) => {
          done(
            void 0,
            { fullProfile, params, accessToken },
            { refreshToken }
          );
        }
      )
    );
  },
  async start(input, helper) {
    return helper.start(input, {
      accessType: "offline",
      prompt: "consent"
    });
  },
  async authenticate(input, helper) {
    return helper.authenticate(input);
  },
  async refresh(input, helper) {
    return helper.refresh(input);
  }
});

exports.oktaAuthenticator = oktaAuthenticator;
