// @backstage/plugin-scaffolder-backend-module-gitlab/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var gitlab = require('./actions/gitlab.cjs.js');
var gitlabGroupEnsureExists = require('./actions/gitlabGroupEnsureExists.cjs.js');
var gitlabIssueCreate = require('./actions/gitlabIssueCreate.cjs.js');
var gitlabIssueEdit = require('./actions/gitlabIssueEdit.cjs.js');
var gitlabMergeRequest = require('./actions/gitlabMergeRequest.cjs.js');
var gitlabPipelineTrigger = require('./actions/gitlabPipelineTrigger.cjs.js');
var gitlabProjectAccessTokenCreate = require('./actions/gitlabProjectAccessTokenCreate.cjs.js');
var gitlabProjectDeployTokenCreate = require('./actions/gitlabProjectDeployTokenCreate.cjs.js');
var gitlabProjectVariableCreate = require('./actions/gitlabProjectVariableCreate.cjs.js');
var gitlabRepoPush = require('./actions/gitlabRepoPush.cjs.js');
var commonGitlabConfig = require('./commonGitlabConfig.cjs.js');
var module$1 = require('./module.cjs.js');



exports.createPublishGitlabAction = gitlab.createPublishGitlabAction;
exports.createGitlabGroupEnsureExistsAction = gitlabGroupEnsureExists.createGitlabGroupEnsureExistsAction;
exports.createGitlabIssueAction = gitlabIssueCreate.createGitlabIssueAction;
exports.editGitlabIssueAction = gitlabIssueEdit.editGitlabIssueAction;
exports.createPublishGitlabMergeRequestAction = gitlabMergeRequest.createPublishGitlabMergeRequestAction;
exports.createTriggerGitlabPipelineAction = gitlabPipelineTrigger.createTriggerGitlabPipelineAction;
exports.createGitlabProjectAccessTokenAction = gitlabProjectAccessTokenCreate.createGitlabProjectAccessTokenAction;
exports.createGitlabProjectDeployTokenAction = gitlabProjectDeployTokenCreate.createGitlabProjectDeployTokenAction;
exports.createGitlabProjectVariableAction = gitlabProjectVariableCreate.createGitlabProjectVariableAction;
exports.createGitlabRepoPushAction = gitlabRepoPush.createGitlabRepoPushAction;
exports.IssueStateEvent = commonGitlabConfig.IssueStateEvent;
exports.IssueType = commonGitlabConfig.IssueType;
exports.default = module$1.gitlabModule;
