// @backstage/plugin-scaffolder-backend/dist/scaffolder/actions/builtin/debug/log.examples.cjs.js
'use strict';

var yaml = require('yaml');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var yaml__default = /*#__PURE__*/_interopDefaultCompat(yaml);

const examples = [
  {
    description: "Write a debug message",
    example: yaml__default.default.stringify({
      steps: [
        {
          action: "debug:log",
          id: "write-debug-line",
          name: 'Write "Hello Backstage!" log line',
          input: {
            message: "Hello Backstage!"
          }
        }
      ]
    })
  },
  {
    description: "List the workspace directory",
    example: yaml__default.default.stringify({
      steps: [
        {
          action: "debug:log",
          id: "write-workspace-directory",
          name: "List the workspace directory",
          input: {
            listWorkspace: true
          }
        }
      ]
    })
  },
  {
    description: "List the workspace directory with file contents",
    example: yaml__default.default.stringify({
      steps: [
        {
          action: "debug:log",
          id: "write-workspace-directory",
          name: "List the workspace directory with file contents",
          input: {
            listWorkspace: "with-contents"
          }
        }
      ]
    })
  }
];

exports.examples = examples;
