// @backstage/plugin-scaffolder-backend/dist/scaffolder/actions/builtin/debug/wait.examples.cjs.js
'use strict';

var yaml = require('yaml');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var yaml__default = /*#__PURE__*/_interopDefaultCompat(yaml);

const examples = [
  {
    description: "Waiting for 50 milliseconds",
    example: yaml__default.default.stringify({
      steps: [
        {
          action: "debug:wait",
          id: "wait-milliseconds",
          name: "Waiting for 50 milliseconds",
          input: {
            milliseconds: 50
          }
        }
      ]
    })
  },
  {
    description: "Waiting for 5 seconds",
    example: yaml__default.default.stringify({
      steps: [
        {
          action: "debug:wait",
          id: "wait-5sec",
          name: "Waiting for 5 seconds",
          input: {
            seconds: 5
          }
        }
      ]
    })
  },
  {
    description: "Waiting for 1 minutes",
    example: yaml__default.default.stringify({
      steps: [
        {
          action: "debug:wait",
          id: "wait-1min",
          name: "Waiting for 1 minutes",
          input: {
            minutes: 1
          }
        }
      ]
    })
  }
];

exports.examples = examples;
