// @backstage/plugin-kubernetes-react/dist/components/CustomResources/ArgoRollouts/RolloutDrawer.esm.js
import React__default from 'react';
import { KubernetesStructuredMetadataTableDrawer } from '../../KubernetesDrawer/KubernetesStructuredMetadataTableDrawer.esm.js';
import '../../KubernetesDrawer/KubernetesDrawer.esm.js';
import '@backstage/core-components';
import '@material-ui/core/FormControlLabel';
import '@material-ui/core/Switch';
import 'js-yaml';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import Chip from '@material-ui/core/Chip';

const RolloutDrawer = ({
  rollout,
  expanded
}) => {
  const namespace = rollout.metadata?.namespace;
  return /* @__PURE__ */ React__default.createElement(
    KubernetesStructuredMetadataTableDrawer,
    {
      object: rollout,
      expanded,
      kind: "Rollout",
      renderObject: () => ({})
    },
    /* @__PURE__ */ React__default.createElement(
      Grid,
      {
        container: true,
        direction: "column",
        justifyContent: "flex-start",
        alignItems: "flex-start",
        spacing: 0
      },
      /* @__PURE__ */ React__default.createElement(Grid, { item: true }, /* @__PURE__ */ React__default.createElement(Typography, { variant: "body1" }, rollout.metadata?.name ?? "unknown object")),
      /* @__PURE__ */ React__default.createElement(Grid, { item: true }, /* @__PURE__ */ React__default.createElement(Typography, { color: "textSecondary", variant: "subtitle1" }, "Rollout")),
      namespace && /* @__PURE__ */ React__default.createElement(Grid, { item: true }, /* @__PURE__ */ React__default.createElement(Chip, { size: "small", label: `namespace: ${namespace}` }))
    )
  );
};

export { RolloutDrawer };
