// @backstage/plugin-user-settings/dist/components/useUserProfileInfo.esm.js
import { useApi, identityApiRef, alertApiRef } from '@backstage/core-plugin-api';
import { useEffect } from 'react';
import useAsync from 'react-use/esm/useAsync';
import { catalogApiRef } from '@backstage/plugin-catalog-react';

const useUserProfile = () => {
  const identityApi = useApi(identityApiRef);
  const alertApi = useApi(alertApiRef);
  const catalogApi = useApi(catalogApiRef);
  const { value, loading, error } = useAsync(async () => {
    let identityProfile = await identityApi.getProfileInfo();
    const backStageIdentity = await identityApi.getBackstageIdentity();
    const catalogProfile = await catalogApi.getEntityByRef(
      backStageIdentity.userEntityRef
    );
    if (identityProfile.picture === void 0 && catalogProfile?.spec?.profile?.picture) {
      identityProfile = {
        ...identityProfile,
        picture: catalogProfile.spec.profile.picture
      };
    }
    return {
      profile: identityProfile,
      identity: backStageIdentity
    };
  }, []);
  useEffect(() => {
    if (error) {
      alertApi.post({
        message: `Failed to load user identity: ${error}`,
        severity: "error"
      });
    }
  }, [error, alertApi]);
  if (loading || error) {
    return {
      profile: {},
      displayName: "",
      loading
    };
  }
  return {
    profile: value.profile,
    backstageIdentity: value.identity,
    displayName: value.profile.displayName ?? value.identity.userEntityRef,
    loading
  };
};

export { useUserProfile };
