// @backstage/core-components/dist/components/ProgressBars/GaugeCard.esm.js
import Box from '@material-ui/core/Box';
import { makeStyles } from '@material-ui/core/styles';
import React__default from 'react';
import { InfoCard } from '../../layout/InfoCard/InfoCard.esm.js';
import { Gauge } from './Gauge.esm.js';

const useStyles = makeStyles(
  {
    root: {
      height: "100%",
      width: 250
    },
    rootSmall: {
      height: "100%",
      width: 160
    }
  },
  { name: "BackstageGaugeCard" }
);
function GaugeCard(props) {
  const classes = useStyles(props);
  const {
    title,
    subheader,
    progress,
    inverse,
    deepLink,
    description,
    icon,
    variant,
    alignGauge = "normal",
    size = "normal",
    getColor
  } = props;
  const gaugeProps = {
    inverse,
    description,
    getColor,
    value: progress
  };
  return /* @__PURE__ */ React__default.createElement(Box, { className: size === "small" ? classes.rootSmall : classes.root }, /* @__PURE__ */ React__default.createElement(
    InfoCard,
    {
      title,
      subheader,
      deepLink,
      variant,
      alignContent: alignGauge,
      icon,
      titleTypographyProps: {
        ...size === "small" ? { variant: "subtitle2" } : void 0
      },
      subheaderTypographyProps: {
        ...size === "small" ? { variant: "body2" } : void 0
      }
    },
    /* @__PURE__ */ React__default.createElement(Gauge, { ...gaugeProps, size })
  ));
}

export { GaugeCard };
