// @backstage/backend-defaults/dist/entrypoints/rootLogger/rootLoggerServiceFactory.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var winston = require('winston');
var WinstonLogger = require('./WinstonLogger.cjs.js');
var createConfigSecretEnumerator = require('../rootConfig/createConfigSecretEnumerator.cjs.js');

const rootLoggerServiceFactory = backendPluginApi.createServiceFactory({
  service: backendPluginApi.coreServices.rootLogger,
  deps: {
    config: backendPluginApi.coreServices.rootConfig
  },
  async factory({ config }) {
    const logger = WinstonLogger.WinstonLogger.create({
      meta: {
        service: "backstage"
      },
      level: process.env.LOG_LEVEL || "info",
      format: process.env.NODE_ENV === "production" ? winston.format.json() : WinstonLogger.WinstonLogger.colorFormat(),
      transports: [new winston.transports.Console()]
    });
    const secretEnumerator = await createConfigSecretEnumerator.createConfigSecretEnumerator({ logger });
    logger.addRedactions(secretEnumerator(config));
    config.subscribe?.(() => logger.addRedactions(secretEnumerator(config)));
    return logger;
  }
});

exports.rootLoggerServiceFactory = rootLoggerServiceFactory;
