// @backstage/plugin-catalog/dist/components/DependencyOfComponentsCard/DependencyOfComponentsCard.esm.js
import { RELATION_DEPENDENCY_OF } from '@backstage/catalog-model';
import React from 'react';
import { RelatedEntitiesCard } from '../RelatedEntitiesCard/RelatedEntitiesCard.esm.js';
import { componentEntityColumns, componentEntityHelpLink, asComponentEntities } from '../RelatedEntitiesCard/presets.esm.js';
import { catalogTranslationRef } from '../../alpha/translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

function DependencyOfComponentsCard(props) {
  const { t } = useTranslationRef(catalogTranslationRef);
  const {
    variant = "gridItem",
    title = t("dependencyOfComponentsCard.title"),
    columns = componentEntityColumns,
    tableOptions = {}
  } = props;
  return /* @__PURE__ */ React.createElement(
    RelatedEntitiesCard,
    {
      variant,
      title,
      entityKind: "Component",
      relationType: RELATION_DEPENDENCY_OF,
      columns,
      emptyMessage: t("dependencyOfComponentsCard.emptyMessage"),
      emptyHelpLink: componentEntityHelpLink,
      asRenderableEntities: asComponentEntities,
      tableOptions
    }
  );
}

export { DependencyOfComponentsCard };
