// @backstage/core-components/dist/components/SimpleStepper/SimpleStepperFooter.esm.js
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import { makeStyles } from '@material-ui/core/styles';
import React__default, { useContext } from 'react';
import { VerticalStepperContext } from './SimpleStepper.esm.js';
import { coreComponentsTranslationRef } from '../../translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

const useStyles = makeStyles(
  (theme) => ({
    root: {
      marginTop: theme.spacing(3),
      "& button": {
        marginRight: theme.spacing(1)
      }
    }
  }),
  { name: "BackstageSimpleStepperFooter" }
);
const RestartBtn = ({ text, handleClick }) => {
  const { t } = useTranslationRef(coreComponentsTranslationRef);
  return /* @__PURE__ */ React__default.createElement(Button, { onClick: handleClick }, text || t("simpleStepper.reset"));
};
const NextBtn = ({
  text,
  handleClick,
  disabled,
  last,
  stepIndex
}) => {
  const { t } = useTranslationRef(coreComponentsTranslationRef);
  return /* @__PURE__ */ React__default.createElement(
    Button,
    {
      variant: "contained",
      color: "primary",
      disabled,
      "data-testid": `nextButton-${stepIndex}`,
      onClick: handleClick
    },
    text || (last ? t("simpleStepper.finish") : t("simpleStepper.next"))
  );
};
const SkipBtn = ({ text, handleClick, disabled, stepIndex }) => {
  const { t } = useTranslationRef(coreComponentsTranslationRef);
  return /* @__PURE__ */ React__default.createElement(
    Button,
    {
      variant: "outlined",
      color: "primary",
      disabled,
      "data-testid": `skipButton-${stepIndex}`,
      onClick: handleClick
    },
    text || t("simpleStepper.skip")
  );
};
const BackBtn = ({ text, handleClick, disabled, stepIndex }) => {
  const { t } = useTranslationRef(coreComponentsTranslationRef);
  return /* @__PURE__ */ React__default.createElement(
    Button,
    {
      onClick: handleClick,
      "data-testid": `backButton-${stepIndex}`,
      disabled
    },
    text || t("simpleStepper.back")
  );
};
const SimpleStepperFooter = ({
  actions = {},
  children
}) => {
  const classes = useStyles();
  const {
    stepperLength,
    stepIndex,
    setStepIndex,
    stepHistory,
    setStepHistory,
    onStepChange
  } = useContext(VerticalStepperContext);
  const onChange = (newIndex, callback) => {
    if (callback) {
      callback();
    }
    if (onStepChange) {
      onStepChange(stepIndex, newIndex);
    }
    setStepIndex(newIndex);
  };
  const handleNext = () => {
    const newIndex = actions.nextStep ? actions.nextStep(stepIndex, stepperLength - 1) : stepIndex + 1;
    onChange(newIndex, actions.onNext);
    setStepHistory([...stepHistory, newIndex]);
  };
  const handleBack = () => {
    stepHistory.pop();
    onChange(stepHistory[stepHistory.length - 1], actions.onBack);
    setStepHistory([...stepHistory]);
  };
  const handleRestart = () => {
    onChange(0, actions.onRestart);
    setStepHistory([0]);
  };
  return /* @__PURE__ */ React__default.createElement(Box, { className: classes.root }, [void 0, true].includes(actions.showBack) && stepIndex !== 0 && /* @__PURE__ */ React__default.createElement(
    BackBtn,
    {
      text: actions.backText,
      handleClick: handleBack,
      disabled: stepIndex === 0,
      stepIndex
    }
  ), actions.showSkip && /* @__PURE__ */ React__default.createElement(
    SkipBtn,
    {
      text: actions.skipText,
      handleClick: handleNext,
      disabled: !!stepperLength && stepIndex >= stepperLength || !!actions.canSkip && !actions.canSkip(),
      stepIndex
    }
  ), [void 0, true].includes(actions.showNext) && /* @__PURE__ */ React__default.createElement(
    NextBtn,
    {
      text: actions.nextText,
      handleClick: handleNext,
      disabled: !!stepperLength && stepIndex >= stepperLength || !!actions.canNext && !actions.canNext(),
      stepIndex
    }
  ), actions.showRestart && stepIndex !== 0 && /* @__PURE__ */ React__default.createElement(
    RestartBtn,
    {
      text: actions.restartText,
      handleClick: handleRestart,
      stepIndex
    }
  ), children);
};

export { RestartBtn, SimpleStepperFooter };
