// @backstage/plugin-kubernetes-react/dist/components/JobsAccordions/JobsDrawer.esm.js
import React__default from 'react';
import { KubernetesStructuredMetadataTableDrawer } from '../KubernetesDrawer/KubernetesStructuredMetadataTableDrawer.esm.js';
import '../KubernetesDrawer/KubernetesDrawer.esm.js';
import '@backstage/core-components';
import '@material-ui/core/FormControlLabel';
import '@material-ui/core/Switch';
import 'js-yaml';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';

const JobDrawer = ({
  job,
  expanded
}) => {
  return /* @__PURE__ */ React__default.createElement(
    KubernetesStructuredMetadataTableDrawer,
    {
      object: job,
      expanded,
      kind: "Job",
      renderObject: (jobObj) => {
        return {
          parallelism: jobObj.spec?.parallelism ?? "???",
          completions: jobObj.spec?.completions ?? "???",
          backoffLimit: jobObj.spec?.backoffLimit ?? "???",
          startTime: jobObj.status?.startTime ?? "???"
        };
      }
    },
    /* @__PURE__ */ React__default.createElement(
      Grid,
      {
        container: true,
        direction: "column",
        justifyContent: "flex-start",
        alignItems: "flex-start",
        spacing: 0
      },
      /* @__PURE__ */ React__default.createElement(Grid, { item: true }, /* @__PURE__ */ React__default.createElement(Typography, { variant: "body1" }, job.metadata?.name ?? "unknown object")),
      /* @__PURE__ */ React__default.createElement(Grid, { item: true }, /* @__PURE__ */ React__default.createElement(Typography, { color: "textSecondary", variant: "subtitle1" }, "Job"))
    )
  );
};

export { JobDrawer };
