// @backstage/plugin-kubernetes-common/dist/catalog-entity-constants.cjs.js
'use strict';

const ANNOTATION_KUBERNETES_API_SERVER = "kubernetes.io/api-server";
const ANNOTATION_KUBERNETES_API_SERVER_CA = "kubernetes.io/api-server-certificate-authority";
const ANNOTATION_KUBERNETES_AUTH_PROVIDER = "kubernetes.io/auth-provider";
const ANNOTATION_KUBERNETES_OIDC_TOKEN_PROVIDER = "kubernetes.io/oidc-token-provider";
const ANNOTATION_KUBERNETES_SKIP_METRICS_LOOKUP = "kubernetes.io/skip-metrics-lookup";
const ANNOTATION_KUBERNETES_SKIP_TLS_VERIFY = "kubernetes.io/skip-tls-verify";
const ANNOTATION_KUBERNETES_DASHBOARD_URL = "kubernetes.io/dashboard-url";
const ANNOTATION_KUBERNETES_DASHBOARD_APP = "kubernetes.io/dashboard-app";
const ANNOTATION_KUBERNETES_DASHBOARD_PARAMETERS = "kubernetes.io/dashboard-parameters";
const ANNOTATION_KUBERNETES_AWS_ASSUME_ROLE = "kubernetes.io/aws-assume-role";
const ANNOTATION_KUBERNETES_AWS_CLUSTER_ID = "kubernetes.io/x-k8s-aws-id";
const ANNOTATION_KUBERNETES_AWS_EXTERNAL_ID = "kubernetes.io/aws-external-id";

exports.ANNOTATION_KUBERNETES_API_SERVER = ANNOTATION_KUBERNETES_API_SERVER;
exports.ANNOTATION_KUBERNETES_API_SERVER_CA = ANNOTATION_KUBERNETES_API_SERVER_CA;
exports.ANNOTATION_KUBERNETES_AUTH_PROVIDER = ANNOTATION_KUBERNETES_AUTH_PROVIDER;
exports.ANNOTATION_KUBERNETES_AWS_ASSUME_ROLE = ANNOTATION_KUBERNETES_AWS_ASSUME_ROLE;
exports.ANNOTATION_KUBERNETES_AWS_CLUSTER_ID = ANNOTATION_KUBERNETES_AWS_CLUSTER_ID;
exports.ANNOTATION_KUBERNETES_AWS_EXTERNAL_ID = ANNOTATION_KUBERNETES_AWS_EXTERNAL_ID;
exports.ANNOTATION_KUBERNETES_DASHBOARD_APP = ANNOTATION_KUBERNETES_DASHBOARD_APP;
exports.ANNOTATION_KUBERNETES_DASHBOARD_PARAMETERS = ANNOTATION_KUBERNETES_DASHBOARD_PARAMETERS;
exports.ANNOTATION_KUBERNETES_DASHBOARD_URL = ANNOTATION_KUBERNETES_DASHBOARD_URL;
exports.ANNOTATION_KUBERNETES_OIDC_TOKEN_PROVIDER = ANNOTATION_KUBERNETES_OIDC_TOKEN_PROVIDER;
exports.ANNOTATION_KUBERNETES_SKIP_METRICS_LOOKUP = ANNOTATION_KUBERNETES_SKIP_METRICS_LOOKUP;
exports.ANNOTATION_KUBERNETES_SKIP_TLS_VERIFY = ANNOTATION_KUBERNETES_SKIP_TLS_VERIFY;
