// @backstage/plugin-auth-backend-module-bitbucket-provider/dist/authenticator.cjs.js
'use strict';

var passportBitbucketOauth2 = require('passport-bitbucket-oauth2');
var pluginAuthNode = require('@backstage/plugin-auth-node');

const bitbucketAuthenticator = pluginAuthNode.createOAuthAuthenticator({
  defaultProfileTransform: pluginAuthNode.PassportOAuthAuthenticatorHelper.defaultProfileTransform,
  scopes: {
    required: ["account"]
  },
  initialize({ callbackUrl, config }) {
    const clientID = config.getString("clientId");
    const clientSecret = config.getString("clientSecret");
    const baseURL = "https://bitbucket.org/site/oauth2";
    return pluginAuthNode.PassportOAuthAuthenticatorHelper.from(
      new passportBitbucketOauth2.Strategy(
        {
          clientID,
          clientSecret,
          callbackURL: callbackUrl,
          passReqToCallback: false,
          baseURL,
          authorizationURL: `${baseURL}/authorize`,
          tokenURL: `${baseURL}/access_token`
        },
        (accessToken, refreshToken, params, fullProfile, done) => {
          done(
            void 0,
            { fullProfile, params, accessToken },
            { refreshToken }
          );
        }
      )
    );
  },
  async start(input, helper) {
    return helper.start(input, {
      accessType: "offline",
      prompt: "consent"
    });
  },
  async authenticate(input, helper) {
    return helper.authenticate(input);
  },
  async refresh(input, helper) {
    return helper.refresh(input);
  }
});

exports.bitbucketAuthenticator = bitbucketAuthenticator;
