// @backstage/plugin-scaffolder/dist/alpha/components/TemplateEditorPage/CustomFieldPlaygroud.esm.js
import React, { useState, useMemo, useCallback } from 'react';
import yaml from 'yaml';
import validator from '@rjsf/validator-ajv8';
import CodeMirror from '@uiw/react-codemirror';
import { StreamLanguage } from '@codemirror/language';
import { yaml as yaml$1 } from '@codemirror/legacy-modes/mode/yaml';
import { makeStyles } from '@material-ui/core/styles';
import Accordion from '@material-ui/core/Accordion';
import AccordionSummary from '@material-ui/core/AccordionSummary';
import AccordionDetails from '@material-ui/core/AccordionDetails';
import Autocomplete from '@material-ui/lab/Autocomplete';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import InputAdornment from '@material-ui/core/InputAdornment';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import SearchIcon from '@material-ui/icons/Search';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { Form } from '@backstage/plugin-scaffolder-react/alpha';
import { scaffolderTranslationRef } from '../../../translation.esm.js';
import { TemplateEditorForm } from './TemplateEditorForm.esm.js';

const useStyles = makeStyles(
  (theme) => ({
    root: {
      gridArea: "pageContent",
      display: "grid",
      gridTemplateRows: "auto 1fr"
    },
    controls: {
      marginBottom: theme.spacing(3)
    },
    code: {
      width: "100%"
    }
  }),
  { name: "ScaffolderCustomFieldExtensionsPlaygroud" }
);
const CustomFieldPlaygroud = ({
  fieldExtensions = []
}) => {
  const classes = useStyles();
  const { t } = useTranslationRef(scaffolderTranslationRef);
  const fieldOptions = fieldExtensions.filter((field) => !!field.schema);
  const [refreshKey, setRefreshKey] = useState(Date.now());
  const [fieldFormState, setFieldFormState] = useState({});
  const [selectedField, setSelectedField] = useState(fieldOptions[0]);
  const sampleFieldTemplate = useMemo(
    () => yaml.stringify({
      parameters: [
        {
          title: `${selectedField.name} Example`,
          properties: {
            [selectedField.name]: {
              type: selectedField.schema?.returnValue?.type,
              "ui:field": selectedField.name,
              "ui:options": fieldFormState
            }
          }
        }
      ]
    }),
    [fieldFormState, selectedField]
  );
  const fieldComponents = useMemo(() => {
    return Object.fromEntries(
      fieldExtensions.map(({ name, component }) => [name, component])
    );
  }, [fieldExtensions]);
  const handleSelectionChange = useCallback(
    (selection) => {
      setSelectedField(selection);
      setFieldFormState({});
    },
    [setFieldFormState, setSelectedField]
  );
  const handleFieldConfigChange = useCallback(
    (state) => {
      setFieldFormState(state);
      setRefreshKey(Date.now());
    },
    [setFieldFormState, setRefreshKey]
  );
  return /* @__PURE__ */ React.createElement("main", { className: classes.root }, /* @__PURE__ */ React.createElement("div", { className: classes.controls }, /* @__PURE__ */ React.createElement(
    Autocomplete,
    {
      id: "custom-fields-autocomplete",
      value: selectedField,
      options: fieldOptions,
      getOptionLabel: (option) => option.name,
      renderInput: (params) => /* @__PURE__ */ React.createElement(
        TextField,
        {
          ...params,
          "aria-label": t(
            "templateEditorPage.customFieldExplorer.selectFieldLabel"
          ),
          placeholder: t(
            "templateEditorPage.customFieldExplorer.selectFieldLabel"
          ),
          variant: "outlined",
          InputProps: {
            ...params.InputProps,
            startAdornment: /* @__PURE__ */ React.createElement(InputAdornment, { position: "start" }, /* @__PURE__ */ React.createElement(SearchIcon, null))
          }
        }
      ),
      onChange: (_event, option) => {
        if (option) {
          handleSelectionChange(option);
        }
      },
      disableClearable: true,
      fullWidth: true
    }
  )), /* @__PURE__ */ React.createElement("div", null, /* @__PURE__ */ React.createElement(Accordion, { defaultExpanded: true }, /* @__PURE__ */ React.createElement(
    AccordionSummary,
    {
      expandIcon: /* @__PURE__ */ React.createElement(ExpandMoreIcon, null),
      "aria-controls": "panel-code-content",
      id: "panel-code-header"
    },
    /* @__PURE__ */ React.createElement(Typography, { variant: "h6" }, t("templateEditorPage.customFieldExplorer.preview.title"))
  ), /* @__PURE__ */ React.createElement(AccordionDetails, null, /* @__PURE__ */ React.createElement("div", { className: classes.code }, /* @__PURE__ */ React.createElement(
    CodeMirror,
    {
      readOnly: true,
      theme: "dark",
      height: "100%",
      width: "100%",
      extensions: [StreamLanguage.define(yaml$1)],
      value: sampleFieldTemplate
    }
  )))), /* @__PURE__ */ React.createElement(Accordion, { defaultExpanded: true }, /* @__PURE__ */ React.createElement(
    AccordionSummary,
    {
      expandIcon: /* @__PURE__ */ React.createElement(ExpandMoreIcon, null),
      "aria-controls": "panel-preview-content",
      id: "panel-preview-header"
    },
    /* @__PURE__ */ React.createElement(Typography, { variant: "h6" }, t("templateEditorPage.customFieldExplorer.fieldPreview.title"))
  ), /* @__PURE__ */ React.createElement(AccordionDetails, null, /* @__PURE__ */ React.createElement(
    TemplateEditorForm,
    {
      key: refreshKey,
      content: sampleFieldTemplate,
      contentIsSpec: true,
      fieldExtensions,
      setErrorText: () => null
    }
  ))), /* @__PURE__ */ React.createElement(Accordion, { defaultExpanded: true }, /* @__PURE__ */ React.createElement(
    AccordionSummary,
    {
      expandIcon: /* @__PURE__ */ React.createElement(ExpandMoreIcon, null),
      "aria-controls": "panel-options-content",
      id: "panel-options-header"
    },
    /* @__PURE__ */ React.createElement(Typography, { variant: "h6" }, t("templateEditorPage.customFieldExplorer.fieldForm.title"))
  ), /* @__PURE__ */ React.createElement(AccordionDetails, null, /* @__PURE__ */ React.createElement(
    Form,
    {
      showErrorList: false,
      fields: { ...fieldComponents },
      noHtml5Validate: true,
      formData: fieldFormState,
      formContext: { fieldFormState },
      onSubmit: (e) => handleFieldConfigChange(e.formData),
      validator,
      schema: selectedField.schema?.uiOptions || {},
      experimental_defaultFormStateBehavior: {
        allOf: "populateDefaults"
      }
    },
    /* @__PURE__ */ React.createElement(
      Button,
      {
        variant: "contained",
        color: "primary",
        type: "submit",
        disabled: !selectedField.schema?.uiOptions
      },
      t(
        "templateEditorPage.customFieldExplorer.fieldForm.applyButtonTitle"
      )
    )
  )))));
};

export { CustomFieldPlaygroud };
