// @backstage/plugin-scaffolder/dist/alpha/components/TemplateEditorPage/TemplateEditorIntro.esm.js
import React from 'react';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import 'idb-keyval';
import { WebFileSystemAccess } from '../../../lib/filesystem/WebFileSystemAccess.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../../translation.esm.js';
import CreateNewFolderIcon from '@material-ui/icons/CreateNewFolder';
import ListAltIcon from '@material-ui/icons/ListAlt';
import FormatListBulletedIcon from '@material-ui/icons/FormatListBulleted';
import InfoOutlinedIcon from '@material-ui/icons/InfoOutlined';
import CardMedia from '@material-ui/core/CardMedia';
import PublishIcon from '@material-ui/icons/Publish';
import Tooltip from '@material-ui/core/Tooltip';

const useStyles = makeStyles((theme) => ({
  gridRoot: {
    display: "flex",
    flex: 1,
    alignItems: "center",
    justifyContent: "center"
  },
  cardGrid: {
    maxWidth: 1e3,
    display: "grid",
    gridGap: theme.spacing(2),
    gridAutoFlow: "row",
    [theme.breakpoints.up("md")]: {
      gridTemplateRows: "1fr 1fr",
      gridTemplateColumns: "1fr 1fr"
    }
  },
  card: {
    display: "grid",
    gridTemplateColumns: "auto 1fr",
    gridTemplateRows: "1fr",
    alignItems: "center",
    margin: theme.spacing(0, 1),
    marginTop: theme.spacing(2),
    padding: theme.spacing(2)
  },
  icon: {
    justifySelf: "center",
    paddingTop: theme.spacing(1),
    fontSize: 48
  },
  introText: {
    textAlign: "center",
    marginTop: theme.spacing(2)
  },
  infoIcon: {
    position: "absolute",
    top: theme.spacing(1),
    right: theme.spacing(1)
  },
  cardContent: {
    padding: theme.spacing(1)
  }
}));
function ActionCard(props) {
  const supportsLoad = props.requireLoad ? WebFileSystemAccess.isSupported() : true;
  const { t } = useTranslationRef(scaffolderTranslationRef);
  const classes = useStyles();
  const { Icon, title, description, action } = props;
  return /* @__PURE__ */ React.createElement(Card, { className: classes.card }, !supportsLoad && /* @__PURE__ */ React.createElement(
    Tooltip,
    {
      placement: "top",
      title: t(
        "templateEditorPage.templateEditorIntro.loadLocal.unsupportedTooltip"
      )
    },
    /* @__PURE__ */ React.createElement(InfoOutlinedIcon, null)
  ), /* @__PURE__ */ React.createElement(CardActionArea, { onClick: action }, /* @__PURE__ */ React.createElement(CardMedia, null, /* @__PURE__ */ React.createElement(
    Icon,
    {
      className: classes.icon,
      color: supportsLoad ? void 0 : "disabled"
    }
  )), /* @__PURE__ */ React.createElement(CardContent, { className: classes.cardContent }, /* @__PURE__ */ React.createElement(
    Typography,
    {
      gutterBottom: true,
      variant: "h5",
      component: "h2",
      color: supportsLoad ? void 0 : "textSecondary"
    },
    title
  ), /* @__PURE__ */ React.createElement(Typography, { variant: "body2", color: "textSecondary", component: "p" }, description))));
}
function TemplateEditorIntro(props) {
  const classes = useStyles();
  const { t } = useTranslationRef(scaffolderTranslationRef);
  return /* @__PURE__ */ React.createElement("div", { style: props.style }, /* @__PURE__ */ React.createElement(Typography, { variant: "h4", component: "h2", className: classes.introText }, t("templateEditorPage.templateEditorIntro.title")), /* @__PURE__ */ React.createElement("div", { className: classes.gridRoot }, /* @__PURE__ */ React.createElement("div", { className: classes.cardGrid }, /* @__PURE__ */ React.createElement(
    ActionCard,
    {
      title: t("templateEditorPage.templateEditorIntro.loadLocal.title"),
      description: t(
        "templateEditorPage.templateEditorIntro.loadLocal.description"
      ),
      requireLoad: true,
      Icon: PublishIcon,
      action: () => props.onSelect?.("local")
    }
  ), /* @__PURE__ */ React.createElement(
    ActionCard,
    {
      title: t(
        "templateEditorPage.templateEditorIntro.createLocal.title"
      ),
      description: t(
        "templateEditorPage.templateEditorIntro.createLocal.description"
      ),
      requireLoad: true,
      action: () => props.onSelect?.("create-template"),
      Icon: CreateNewFolderIcon
    }
  ), /* @__PURE__ */ React.createElement(
    ActionCard,
    {
      title: t("templateEditorPage.templateEditorIntro.formEditor.title"),
      description: t(
        "templateEditorPage.templateEditorIntro.formEditor.description"
      ),
      Icon: ListAltIcon,
      action: () => props.onSelect?.("form")
    }
  ), /* @__PURE__ */ React.createElement(
    ActionCard,
    {
      title: t(
        "templateEditorPage.templateEditorIntro.fieldExplorer.title"
      ),
      description: t(
        "templateEditorPage.templateEditorIntro.fieldExplorer.description"
      ),
      Icon: FormatListBulletedIcon,
      action: () => props.onSelect?.("field-explorer")
    }
  ))));
}

export { TemplateEditorIntro };
