// @backstage/plugin-scaffolder/dist/alpha/components/TemplateEditorPage/TemplateEditorLayout.esm.js
import React from 'react';
import { withStyles } from '@material-ui/core/styles';

const TemplateEditorLayout = withStyles(
  (theme) => ({
    root: {
      height: "100%",
      gridArea: "pageContent",
      display: "grid",
      gridTemplateAreas: `
      "toolbar"
      "browser"
      "editor"
      "preview"
      "results"
    `,
      [theme.breakpoints.up("md")]: {
        gridTemplateAreas: `
      "toolbar toolbar toolbar"
      "browser editor preview"
      "results results results"
    `,
        gridTemplateColumns: "1fr 3fr 2fr",
        gridTemplateRows: "auto 1fr auto"
      }
    }
  }),
  { name: "ScaffolderTemplateEditorLayout" }
)(({ children, classes }) => /* @__PURE__ */ React.createElement("main", { className: classes.root }, children));
const TemplateEditorLayoutToolbar = withStyles(
  {
    root: {
      gridArea: "toolbar"
    }
  },
  { name: "ScaffolderTemplateEditorLayoutToolbar" }
)(({ children, classes }) => /* @__PURE__ */ React.createElement("section", { className: classes.root }, children));
const TemplateEditorLayoutBrowser = withStyles(
  (theme) => ({
    root: {
      gridArea: "browser",
      overflow: "auto",
      [theme.breakpoints.up("md")]: {
        borderRight: `1px solid ${theme.palette.divider}`
      }
    }
  }),
  { name: "ScaffolderTemplateEditorLayoutBrowser" }
)(({ children, classes }) => /* @__PURE__ */ React.createElement("section", { className: classes.root }, children));
const TemplateEditorLayoutFiles = withStyles(
  {
    root: {
      gridArea: "editor",
      overflow: "auto"
    }
  },
  { name: "ScaffolderTemplateEditorLayoutFiles" }
)(({ children, classes }) => /* @__PURE__ */ React.createElement("section", { className: classes.root }, children));
const TemplateEditorLayoutPreview = withStyles(
  (theme) => ({
    root: {
      gridArea: "preview",
      position: "relative",
      backgroundColor: theme.palette.background.default,
      [theme.breakpoints.up("md")]: {
        borderLeft: `1px solid ${theme.palette.divider}`
      }
    },
    scroll: {
      height: "100%",
      padding: theme.spacing(1),
      [theme.breakpoints.up("md")]: {
        position: "absolute",
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        overflow: "auto"
      }
    }
  }),
  { name: "ScaffolderTemplateEditorLayoutPreview" }
)(({ children, classes }) => /* @__PURE__ */ React.createElement("section", { className: classes.root }, /* @__PURE__ */ React.createElement("div", { className: classes.scroll }, children)));
const TemplateEditorLayoutConsole = withStyles(
  {
    root: {
      gridArea: "results"
    }
  },
  { name: "ScaffolderTemplateEditorLayoutConsole" }
)(({ children, classes }) => /* @__PURE__ */ React.createElement("section", { className: classes.root }, children));

export { TemplateEditorLayout, TemplateEditorLayoutBrowser, TemplateEditorLayoutConsole, TemplateEditorLayoutFiles, TemplateEditorLayoutPreview, TemplateEditorLayoutToolbar };
