// @backstage/plugin-scaffolder/dist/alpha/components/TemplateEditorPage/TemplateEditorPage.esm.js
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { Page, Header, Content } from '@backstage/core-components';
import { useRouteRef } from '@backstage/core-plugin-api';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../../translation.esm.js';
import { editRouteRef } from '../../../routes.esm.js';
import { TemplateEditor } from './TemplateEditor.esm.js';

const useStyles = makeStyles(
  {
    content: {
      padding: 0
    }
  },
  { name: "ScaffolderTemplateEditorToolbar" }
);
function TemplateEditorPage(props) {
  const classes = useStyles();
  const editLink = useRouteRef(editRouteRef);
  const { t } = useTranslationRef(scaffolderTranslationRef);
  return /* @__PURE__ */ React.createElement(Page, { themeId: "home" }, /* @__PURE__ */ React.createElement(
    Header,
    {
      title: t("templateEditorPage.title"),
      subtitle: t("templateEditorPage.subtitle"),
      type: t("templateIntroPage.title"),
      typeLink: editLink()
    }
  ), /* @__PURE__ */ React.createElement(Content, { className: classes.content }, /* @__PURE__ */ React.createElement(
    TemplateEditor,
    {
      layouts: props.layouts,
      formProps: props.formProps,
      fieldExtensions: props.fieldExtensions
    }
  )));
}

export { TemplateEditorPage };
