// @backstage/plugin-scaffolder/dist/alpha/components/TemplateEditorPage/TemplateEditorTextArea.esm.js
import { StreamLanguage } from '@codemirror/language';
import { yaml } from '@codemirror/legacy-modes/mode/yaml';
import { showPanel } from '@codemirror/view';
import IconButton from '@material-ui/core/IconButton';
import Paper from '@material-ui/core/Paper';
import Tooltip from '@material-ui/core/Tooltip';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import RefreshIcon from '@material-ui/icons/Refresh';
import SaveIcon from '@material-ui/icons/Save';
import { useKeyboardEvent } from '@react-hookz/web';
import CodeMirror from '@uiw/react-codemirror';
import React, { useMemo } from 'react';
import { useDirectoryEditor } from './DirectoryEditorContext.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../../translation.esm.js';

const useStyles = makeStyles((theme) => ({
  container: {
    position: "relative",
    width: "100%",
    height: "100%"
  },
  typography: {
    padding: theme.spacing(1.5)
  },
  button: {
    verticalAlign: "top"
  },
  codeMirror: {
    height: "100%",
    [theme.breakpoints.up("md")]: {
      position: "absolute",
      top: 0,
      bottom: 0,
      left: 0,
      right: 0
    }
  },
  errorPanel: {
    color: theme.palette.error.main,
    lineHeight: 2,
    margin: theme.spacing(0, 1)
  },
  floatingButtons: {
    position: "absolute",
    top: theme.spacing(1),
    right: theme.spacing(3)
  },
  floatingButton: {
    padding: theme.spacing(1)
  }
}));
function TemplateEditorTextArea(props) {
  const { errorText } = props;
  const classes = useStyles();
  const { t } = useTranslationRef(scaffolderTranslationRef);
  const panelExtension = useMemo(() => {
    if (!errorText) {
      return showPanel.of(null);
    }
    const dom = document.createElement("div");
    dom.classList.add(classes.errorPanel);
    dom.textContent = errorText;
    return showPanel.of(() => ({ dom, bottom: true }));
  }, [classes, errorText]);
  useKeyboardEvent(
    (e) => e.key === "s" && (e.ctrlKey || e.metaKey),
    (e) => {
      e.preventDefault();
      if (props.onSave) {
        props.onSave();
      }
    }
  );
  return /* @__PURE__ */ React.createElement("div", { className: classes.container }, /* @__PURE__ */ React.createElement(
    CodeMirror,
    {
      className: classes.codeMirror,
      theme: "dark",
      height: "100%",
      extensions: [StreamLanguage.define(yaml), panelExtension],
      value: props.content,
      onChange: props.onUpdate
    }
  ), (props.onSave || props.onReload) && /* @__PURE__ */ React.createElement("div", { className: classes.floatingButtons }, /* @__PURE__ */ React.createElement(Paper, null, props.onSave && /* @__PURE__ */ React.createElement(
    Tooltip,
    {
      title: t(
        "templateEditorPage.templateEditorTextArea.saveIconTooltip"
      )
    },
    /* @__PURE__ */ React.createElement(
      IconButton,
      {
        className: classes.floatingButton,
        onClick: () => props.onSave?.()
      },
      /* @__PURE__ */ React.createElement(SaveIcon, null)
    )
  ), props.onReload && /* @__PURE__ */ React.createElement(
    Tooltip,
    {
      title: t(
        "templateEditorPage.templateEditorTextArea.refreshIconTooltip"
      )
    },
    /* @__PURE__ */ React.createElement(
      IconButton,
      {
        className: classes.floatingButton,
        onClick: () => props.onReload?.()
      },
      /* @__PURE__ */ React.createElement(RefreshIcon, null)
    )
  ))));
}
function TemplateEditorDirectoryEditorTextArea(props) {
  const classes = useStyles();
  const directoryEditor = useDirectoryEditor();
  const { t } = useTranslationRef(scaffolderTranslationRef);
  if (!directoryEditor) {
    return /* @__PURE__ */ React.createElement(
      Typography,
      {
        className: classes.typography,
        color: "textSecondary",
        align: "center"
      },
      t("templateEditorPage.templateEditorTextArea.emptyStateParagraph")
    );
  }
  const actions = directoryEditor?.selectedFile?.dirty ? {
    onSave: () => directoryEditor.save(),
    onReload: () => directoryEditor.reload()
  } : {
    onReload: () => directoryEditor.reload()
  };
  return /* @__PURE__ */ React.createElement(
    TemplateEditorTextArea,
    {
      errorText: props.errorText,
      content: directoryEditor.selectedFile?.content,
      onUpdate: (content) => directoryEditor?.selectedFile?.updateContent(content),
      ...actions
    }
  );
}
TemplateEditorTextArea.DirectoryEditor = TemplateEditorDirectoryEditorTextArea;

export { TemplateEditorDirectoryEditorTextArea, TemplateEditorTextArea };
