// @backstage/config-loader/dist/index.cjs.js
'use strict';

var compile = require('./schema/compile.cjs.js');
var load = require('./schema/load.cjs.js');
var loader = require('./loader.cjs.js');
var ConfigSources = require('./sources/ConfigSources.cjs.js');
var EnvConfigSource = require('./sources/EnvConfigSource.cjs.js');
var FileConfigSource = require('./sources/FileConfigSource.cjs.js');
var MutableConfigSource = require('./sources/MutableConfigSource.cjs.js');
var RemoteConfigSource = require('./sources/RemoteConfigSource.cjs.js');
var StaticConfigSource = require('./sources/StaticConfigSource.cjs.js');



exports.mergeConfigSchemas = compile.mergeConfigSchemas;
exports.loadConfigSchema = load.loadConfigSchema;
exports.loadConfig = loader.loadConfig;
exports.ConfigSources = ConfigSources.ConfigSources;
exports.EnvConfigSource = EnvConfigSource.EnvConfigSource;
exports.readEnvConfig = EnvConfigSource.readEnvConfig;
exports.FileConfigSource = FileConfigSource.FileConfigSource;
exports.MutableConfigSource = MutableConfigSource.MutableConfigSource;
exports.RemoteConfigSource = RemoteConfigSource.RemoteConfigSource;
exports.StaticConfigSource = StaticConfigSource.StaticConfigSource;
