// @backstage/plugin-scaffolder/dist/lib/filesystem/WebFileSystemAccess.esm.js
const showDirectoryPicker = window.showDirectoryPicker;
class WebFileAccess {
  constructor(path, handle) {
    this.path = path;
    this.handle = handle;
  }
  file() {
    return this.handle.getFile();
  }
  async save(data) {
    const writable = await this.handle.createWritable();
    await writable.write(data);
    await writable.close();
  }
}
class WebDirectoryAccess {
  constructor(handle) {
    this.handle = handle;
  }
  async listFiles() {
    const content = [];
    for await (const entry of this.listDirectoryContents(this.handle)) {
      content.push(entry);
    }
    return content;
  }
  async *listDirectoryContents(dirHandle, basePath = []) {
    for await (const handle of dirHandle.values()) {
      if (handle.kind === "file") {
        yield new WebFileAccess([...basePath, handle.name].join("/"), handle);
      } else if (handle.kind === "directory") {
        if (handle.name === ".git") {
          continue;
        }
        yield* this.listDirectoryContents(handle, [...basePath, handle.name]);
      }
    }
  }
  async createFile(options) {
    const { name, data } = options;
    let file;
    if (name.includes("/")) {
      const [dir, path] = name.split("/");
      const handle = await this.handle.getDirectoryHandle(dir, {
        create: true
      });
      file = await handle.getFileHandle(path, { create: true });
    } else {
      file = await this.handle.getFileHandle(name, {
        create: true
      });
    }
    const writable = await file.createWritable();
    await writable.write(data);
    await writable.close();
  }
}
class WebFileSystemAccess {
  static isSupported() {
    return Boolean(showDirectoryPicker);
  }
  static fromHandle(handle) {
    return new WebDirectoryAccess(handle);
  }
  static async requestDirectoryAccess() {
    if (!showDirectoryPicker) {
      throw new Error("File system access is not supported");
    }
    const handle = await showDirectoryPicker();
    return new WebDirectoryAccess(handle);
  }
  constructor() {
  }
}

export { WebDirectoryAccess, WebFileSystemAccess };
