// @backstage/test-utils/dist/testUtils/apis/ConfigApi/MockConfigApi.esm.js
import { ConfigReader } from '@backstage/config';

class MockConfigApi {
  config;
  // NOTE: not extending in order to avoid inheriting the static `.fromConfigs`
  constructor(data) {
    this.config = new ConfigReader(data);
  }
  /** {@inheritdoc @backstage/config#Config.has} */
  has(key) {
    return this.config.has(key);
  }
  /** {@inheritdoc @backstage/config#Config.keys} */
  keys() {
    return this.config.keys();
  }
  /** {@inheritdoc @backstage/config#Config.get} */
  get(key) {
    return this.config.get(key);
  }
  /** {@inheritdoc @backstage/config#Config.getOptional} */
  getOptional(key) {
    return this.config.getOptional(key);
  }
  /** {@inheritdoc @backstage/config#Config.getConfig} */
  getConfig(key) {
    return this.config.getConfig(key);
  }
  /** {@inheritdoc @backstage/config#Config.getOptionalConfig} */
  getOptionalConfig(key) {
    return this.config.getOptionalConfig(key);
  }
  /** {@inheritdoc @backstage/config#Config.getConfigArray} */
  getConfigArray(key) {
    return this.config.getConfigArray(key);
  }
  /** {@inheritdoc @backstage/config#Config.getOptionalConfigArray} */
  getOptionalConfigArray(key) {
    return this.config.getOptionalConfigArray(key);
  }
  /** {@inheritdoc @backstage/config#Config.getNumber} */
  getNumber(key) {
    return this.config.getNumber(key);
  }
  /** {@inheritdoc @backstage/config#Config.getOptionalNumber} */
  getOptionalNumber(key) {
    return this.config.getOptionalNumber(key);
  }
  /** {@inheritdoc @backstage/config#Config.getBoolean} */
  getBoolean(key) {
    return this.config.getBoolean(key);
  }
  /** {@inheritdoc @backstage/config#Config.getOptionalBoolean} */
  getOptionalBoolean(key) {
    return this.config.getOptionalBoolean(key);
  }
  /** {@inheritdoc @backstage/config#Config.getString} */
  getString(key) {
    return this.config.getString(key);
  }
  /** {@inheritdoc @backstage/config#Config.getOptionalString} */
  getOptionalString(key) {
    return this.config.getOptionalString(key);
  }
  /** {@inheritdoc @backstage/config#Config.getStringArray} */
  getStringArray(key) {
    return this.config.getStringArray(key);
  }
  /** {@inheritdoc @backstage/config#Config.getOptionalStringArray} */
  getOptionalStringArray(key) {
    return this.config.getOptionalStringArray(key);
  }
}

export { MockConfigApi };
