// @backstage/cli/dist/lib/builder/config.cjs.js
'use strict';

var chalk = require('chalk');
var fs = require('fs-extra');
var path = require('path');
var commonjs = require('@rollup/plugin-commonjs');
var resolve = require('@rollup/plugin-node-resolve');
var postcss = require('rollup-plugin-postcss');
var esbuild = require('rollup-plugin-esbuild');
var svgr = require('@svgr/rollup');
var dts = require('rollup-plugin-dts');
var json = require('@rollup/plugin-json');
var yaml = require('@rollup/plugin-yaml');
var plugins = require('./plugins.cjs.js');
var types = require('./types.cjs.js');
var paths = require('../paths.cjs.js');
var svgrTemplate = require('../svgrTemplate.cjs.js');
var entryPoints = require('../entryPoints.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var chalk__default = /*#__PURE__*/_interopDefaultCompat(chalk);
var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);
var commonjs__default = /*#__PURE__*/_interopDefaultCompat(commonjs);
var resolve__default = /*#__PURE__*/_interopDefaultCompat(resolve);
var postcss__default = /*#__PURE__*/_interopDefaultCompat(postcss);
var esbuild__default = /*#__PURE__*/_interopDefaultCompat(esbuild);
var svgr__default = /*#__PURE__*/_interopDefaultCompat(svgr);
var dts__default = /*#__PURE__*/_interopDefaultCompat(dts);
var json__default = /*#__PURE__*/_interopDefaultCompat(json);
var yaml__default = /*#__PURE__*/_interopDefaultCompat(yaml);

const SCRIPT_EXTS = [".js", ".jsx", ".ts", ".tsx"];
function isFileImport(source) {
  if (source.startsWith(".")) {
    return true;
  }
  if (source.startsWith("/")) {
    return true;
  }
  if (source.match(/[a-z]:/i)) {
    return true;
  }
  return false;
}
function buildInternalImportPattern(options) {
  const inlinedPackages = options.workspacePackages.filter(
    (pkg) => pkg.packageJson.backstage?.inline
  );
  for (const { packageJson } of inlinedPackages) {
    if (!packageJson.private) {
      throw new Error(
        `Inlined package ${packageJson.name} must be marked as private`
      );
    }
  }
  const names = inlinedPackages.map((pkg) => pkg.packageJson.name);
  return new RegExp(`^(?:${names.join("|")})(?:$|/)`);
}
async function makeRollupConfigs(options) {
  const configs = new Array();
  const targetDir = options.targetDir ?? paths.paths.targetDir;
  let targetPkg = options.packageJson;
  if (!targetPkg) {
    const packagePath = path.resolve(targetDir, "package.json");
    targetPkg = await fs__default.default.readJson(packagePath);
  }
  const onwarn = ({ code, message }) => {
    if (code === "EMPTY_BUNDLE") {
      return;
    }
    if (options.logPrefix) {
      console.log(options.logPrefix + message);
    } else {
      console.log(message);
    }
  };
  const distDir = path.resolve(targetDir, "dist");
  const entryPoints$1 = entryPoints.readEntryPoints(targetPkg);
  const scriptEntryPoints = entryPoints$1.filter(
    (e) => SCRIPT_EXTS.includes(e.ext)
  );
  const internalImportPattern = buildInternalImportPattern(options);
  const external = (source, importer, isResolved) => Boolean(
    importer && !isResolved && !internalImportPattern.test(source) && !isFileImport(source)
  );
  if (options.outputs.has(types.Output.cjs) || options.outputs.has(types.Output.esm)) {
    const output = new Array();
    const mainFields = ["module", "main"];
    const rewriteNodeModules = (name) => name.replaceAll("node_modules", "node_modules_dist");
    if (options.outputs.has(types.Output.cjs)) {
      output.push({
        dir: distDir,
        entryFileNames: (chunkInfo) => `${rewriteNodeModules(chunkInfo.name)}.cjs.js`,
        chunkFileNames: `cjs/[name]-[hash].cjs.js`,
        format: "commonjs",
        interop: "compat",
        sourcemap: true,
        preserveModules: true,
        preserveModulesRoot: `${targetDir}/src`,
        exports: "named"
      });
    }
    if (options.outputs.has(types.Output.esm)) {
      output.push({
        dir: distDir,
        entryFileNames: (chunkInfo) => `${rewriteNodeModules(chunkInfo.name)}.esm.js`,
        chunkFileNames: `esm/[name]-[hash].esm.js`,
        format: "module",
        sourcemap: true,
        preserveModules: true,
        preserveModulesRoot: `${targetDir}/src`
      });
      mainFields.unshift("browser");
    }
    configs.push({
      input: Object.fromEntries(
        scriptEntryPoints.map((e) => [e.name, path.resolve(targetDir, e.path)])
      ),
      output,
      onwarn,
      makeAbsoluteExternalsRelative: false,
      preserveEntrySignatures: "strict",
      // All module imports are always marked as external
      external,
      plugins: [
        resolve__default.default({ mainFields }),
        commonjs__default.default({
          include: /node_modules/,
          exclude: [/\/[^/]+\.(?:stories|test)\.[^/]+$/]
        }),
        postcss__default.default(),
        plugins.forwardFileImports({
          exclude: /\.icon\.svg$/,
          include: [
            /\.svg$/,
            /\.png$/,
            /\.gif$/,
            /\.jpg$/,
            /\.jpeg$/,
            /\.webp$/,
            /\.eot$/,
            /\.woff$/,
            /\.woff2$/,
            /\.ttf$/,
            /\.md$/
          ]
        }),
        json__default.default(),
        yaml__default.default(),
        svgr__default.default({
          include: /\.icon\.svg$/,
          template: svgrTemplate.svgrTemplate
        }),
        esbuild__default.default({
          target: "ES2022",
          minify: options.minify
        })
      ]
    });
  }
  if (options.outputs.has(types.Output.types)) {
    const input = Object.fromEntries(
      scriptEntryPoints.map((e) => [
        e.name,
        paths.paths.resolveTargetRoot(
          "dist-types",
          path.relative(paths.paths.targetRoot, targetDir),
          e.path.replace(/\.(?:ts|tsx)$/, ".d.ts")
        )
      ])
    );
    for (const path$1 of Object.values(input)) {
      const declarationsExist = await fs__default.default.pathExists(path$1);
      if (!declarationsExist) {
        const declarationPath = path.relative(targetDir, path$1);
        throw new Error(
          `No declaration files found at ${declarationPath}, be sure to run ${chalk__default.default.bgRed.white(
            "yarn tsc"
          )} to generate .d.ts files before packaging`
        );
      }
    }
    configs.push({
      input,
      output: {
        dir: distDir,
        entryFileNames: `[name].d.ts`,
        chunkFileNames: `types/[name]-[hash].d.ts`,
        format: "es"
      },
      external: (source, importer, isResolved) => /\.css|scss|sass|svg|eot|woff|woff2|ttf$/.test(source) || external(source, importer, isResolved),
      onwarn,
      plugins: [dts__default.default({ respectExternal: true })]
    });
  }
  return configs;
}

exports.makeRollupConfigs = makeRollupConfigs;
