// @backstage/plugin-techdocs/dist/alpha.esm.js
import React from 'react';
import LibraryBooks from '@material-ui/icons/LibraryBooks';
import { ApiBlueprint, PageBlueprint, createExtensionInput, coreExtensionData, createExtension, NavItemBlueprint, createFrontendPlugin } from '@backstage/frontend-plugin-api';
import { createApiFactory, configApiRef, discoveryApiRef, fetchApiRef } from '@backstage/core-plugin-api';
import { compatWrapper, convertLegacyRouteRef, convertLegacyRouteRefs } from '@backstage/core-compat-api';
import { EntityContentBlueprint } from '@backstage/plugin-catalog-react/alpha';
import { SearchResultListItemBlueprint } from '@backstage/plugin-search-react/alpha';
import { techdocsStorageApiRef, techdocsApiRef } from '@backstage/plugin-techdocs-react';
import { TechDocsStorageClient, TechDocsClient } from './client.esm.js';
import { rootRouteRef, rootDocsRouteRef, rootCatalogDocsRouteRef } from './routes.esm.js';

const techDocsStorageApi = ApiBlueprint.make({
  name: "storage",
  params: {
    factory: createApiFactory({
      api: techdocsStorageApiRef,
      deps: {
        configApi: configApiRef,
        discoveryApi: discoveryApiRef,
        fetchApi: fetchApiRef
      },
      factory: ({ configApi, discoveryApi, fetchApi }) => new TechDocsStorageClient({
        configApi,
        discoveryApi,
        fetchApi
      })
    })
  }
});
const techDocsClientApi = ApiBlueprint.make({
  params: {
    factory: createApiFactory({
      api: techdocsApiRef,
      deps: {
        configApi: configApiRef,
        discoveryApi: discoveryApiRef,
        fetchApi: fetchApiRef
      },
      factory: ({ configApi, discoveryApi, fetchApi }) => new TechDocsClient({
        configApi,
        discoveryApi,
        fetchApi
      })
    })
  }
});
const techDocsSearchResultListItemExtension = SearchResultListItemBlueprint.makeWithOverrides({
  config: {
    schema: {
      title: (z) => z.string().optional(),
      lineClamp: (z) => z.number().default(5),
      asLink: (z) => z.boolean().default(true),
      asListItem: (z) => z.boolean().default(true)
    }
  },
  factory(originalFactory, { config }) {
    return originalFactory({
      predicate: (result) => result.type === "techdocs",
      component: async () => {
        const { TechDocsSearchResultListItem } = await import('./search/components/TechDocsSearchResultListItem.esm.js');
        return (props) => compatWrapper(
          /* @__PURE__ */ React.createElement(TechDocsSearchResultListItem, { ...props, ...config })
        );
      }
    });
  }
});
const techDocsPage = PageBlueprint.make({
  params: {
    defaultPath: "/docs",
    routeRef: convertLegacyRouteRef(rootRouteRef),
    loader: () => import('./home/components/TechDocsIndexPage.esm.js').then(
      (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.TechDocsIndexPage, null))
    )
  }
});
const techDocsReaderPage = PageBlueprint.make({
  name: "reader",
  params: {
    defaultPath: "/docs/:namespace/:kind/:name",
    routeRef: convertLegacyRouteRef(rootDocsRouteRef),
    loader: () => import('./reader/components/TechDocsReaderPage/index.esm.js').then(
      (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.TechDocsReaderPage, null))
    )
  }
});
const techDocsEntityContent = EntityContentBlueprint.makeWithOverrides({
  inputs: {
    emptyState: createExtensionInput(
      [coreExtensionData.reactElement.optional()],
      {
        singleton: true,
        optional: true
      }
    )
  },
  factory(originalFactory, context) {
    return originalFactory(
      {
        defaultPath: "docs",
        defaultTitle: "TechDocs",
        loader: () => import('./Router.esm.js').then(
          ({ EmbeddedDocsRouter }) => compatWrapper(
            /* @__PURE__ */ React.createElement(
              EmbeddedDocsRouter,
              {
                emptyState: context.inputs.emptyState?.get(
                  coreExtensionData.reactElement
                )
              }
            )
          )
        )
      },
      context
    );
  }
});
const techDocsEntityContentEmptyState = createExtension({
  kind: "empty-state",
  name: "entity-content",
  attachTo: { id: "entity-content:techdocs", input: "emptyState" },
  output: [coreExtensionData.reactElement.optional()],
  factory: () => []
});
const techDocsNavItem = NavItemBlueprint.make({
  params: {
    icon: LibraryBooks,
    title: "Docs",
    routeRef: convertLegacyRouteRef(rootRouteRef)
  }
});
var alpha = createFrontendPlugin({
  id: "techdocs",
  extensions: [
    techDocsClientApi,
    techDocsStorageApi,
    techDocsNavItem,
    techDocsPage,
    techDocsReaderPage,
    techDocsEntityContent,
    techDocsEntityContentEmptyState,
    techDocsSearchResultListItemExtension
  ],
  routes: convertLegacyRouteRefs({
    root: rootRouteRef,
    docRoot: rootDocsRouteRef,
    entityContent: rootCatalogDocsRouteRef
  })
});

export { alpha as default, techDocsSearchResultListItemExtension };
