// @backstage/plugin-scaffolder/dist/alpha/extensions.esm.js
import { convertLegacyRouteRef, compatWrapper } from '@backstage/core-compat-api';
import { PageBlueprint, NavItemBlueprint, ApiBlueprint, createApiFactory, discoveryApiRef, fetchApiRef, identityApiRef } from '@backstage/frontend-plugin-api';
import React from 'react';
import { rootRouteRef } from '../routes.esm.js';
import CreateComponentIcon from '@material-ui/icons/AddCircleOutline';
import { FormFieldBlueprint } from '@backstage/plugin-scaffolder-react/alpha';
import { scmIntegrationsApiRef } from '@backstage/integration-react';
import { scaffolderApiRef } from '@backstage/plugin-scaffolder-react';
import { ScaffolderClient } from '../api.esm.js';

const scaffolderPage = PageBlueprint.make({
  params: {
    routeRef: convertLegacyRouteRef(rootRouteRef),
    defaultPath: "/create",
    loader: () => import('../components/Router/index.esm.js').then((m) => compatWrapper(/* @__PURE__ */ React.createElement(m.Router, null)))
  }
});
const scaffolderNavItem = NavItemBlueprint.make({
  params: {
    routeRef: convertLegacyRouteRef(rootRouteRef),
    title: "Create...",
    icon: CreateComponentIcon
  }
});
const repoUrlPickerFormField = FormFieldBlueprint.make({
  name: "repo-url-picker",
  params: {
    field: () => import('./fields/RepoUrlPicker.esm.js').then((m) => m.RepoUrlPicker)
  }
});
const scaffolderApi = ApiBlueprint.make({
  params: {
    factory: createApiFactory({
      api: scaffolderApiRef,
      deps: {
        discoveryApi: discoveryApiRef,
        scmIntegrationsApi: scmIntegrationsApiRef,
        fetchApi: fetchApiRef,
        identityApi: identityApiRef
      },
      factory: ({ discoveryApi, scmIntegrationsApi, fetchApi, identityApi }) => new ScaffolderClient({
        discoveryApi,
        scmIntegrationsApi,
        fetchApi,
        identityApi
      })
    })
  }
});

export { repoUrlPickerFormField, scaffolderApi, scaffolderNavItem, scaffolderPage };
